{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "47d63650",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(infercnv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6577d37f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 15:12:45] Parsing matrix: /home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_input/exp_arch_prolif_MARS.tsv\n",
      "INFO [2025-10-26 15:13:02] Parsing gene order file: /home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_input/ann_gene.txt\n",
      "INFO [2025-10-26 15:13:02] Parsing cell annotations file: /home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_input/ann_arch_prolif_MARS.txt\n",
      "INFO [2025-10-26 15:13:02] ::order_reduce:Start.\n",
      "INFO [2025-10-26 15:13:02] .order_reduce(): expr and order match.\n",
      "INFO [2025-10-26 15:13:03] ::process_data:order_reduce:Reduction from positional data, new dimensions (r,c) = 52317,1875 Total=148999098 Min=0 Max=5755.\n",
      "INFO [2025-10-26 15:13:04] num genes removed taking into account provided gene ordering list: 2340 = 4.47273352829864% removed.\n",
      "INFO [2025-10-26 15:13:04] -filtering out cells < 100 or > Inf, removing 0 % of cells\n",
      "WARN [2025-10-26 15:13:05] Please use \"options(scipen = 100)\" before running infercnv if you are using the analysis_mode=\"subclusters\" option or you may encounter an error while the hclust is being generated.\n",
      "INFO [2025-10-26 15:13:07] validating infercnv_obj\n",
      "INFO [2025-10-26 15:13:07] ::process_data:Start\n",
      "INFO [2025-10-26 15:13:07] Checking for saved results.\n",
      "INFO [2025-10-26 15:13:07] Trying to reload from step 2\n",
      "INFO [2025-10-26 15:13:09] Using backup from step 2\n",
      "INFO [2025-10-26 15:13:09] \n",
      "\n",
      "\tSTEP 1: incoming data\n",
      "\n",
      "INFO [2025-10-26 15:13:09] \n",
      "\n",
      "\tSTEP 02: Removing lowly expressed genes\n",
      "\n",
      "INFO [2025-10-26 15:13:09] \n",
      "\n",
      "\tSTEP 03: normalization by sequencing depth\n",
      "\n",
      "INFO [2025-10-26 15:13:09] normalizing counts matrix by depth\n",
      "INFO [2025-10-26 15:13:10] Computed total sum normalization factor as median libsize: 70203.000000\n",
      "INFO [2025-10-26 15:13:10] Adding h-spike\n",
      "INFO [2025-10-26 15:13:10] -hspike modeling of Normal_PC\n",
      "INFO [2025-10-26 15:14:57] validating infercnv_obj\n",
      "INFO [2025-10-26 15:14:57] normalizing counts matrix by depth\n",
      "INFO [2025-10-26 15:14:57] Using specified normalization factor: 70203.000000\n",
      "INFO [2025-10-26 15:15:14] \n",
      "\n",
      "\tSTEP 04: log transformation of data\n",
      "\n",
      "INFO [2025-10-26 15:15:14] transforming log2xplus1()\n",
      "INFO [2025-10-26 15:15:14] -mirroring for hspike\n",
      "INFO [2025-10-26 15:15:14] transforming log2xplus1()\n",
      "INFO [2025-10-26 15:15:32] \n",
      "\n",
      "\tSTEP 08: removing average of reference data (before smoothing)\n",
      "\n",
      "INFO [2025-10-26 15:15:32] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
      "INFO [2025-10-26 15:15:32] subtracting mean(normal) per gene per cell across all data\n",
      "INFO [2025-10-26 15:15:34] -subtracting expr per gene, use_bounds=TRUE\n",
      "INFO [2025-10-26 15:15:36] -mirroring for hspike\n",
      "INFO [2025-10-26 15:15:36] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
      "INFO [2025-10-26 15:15:36] subtracting mean(normal) per gene per cell across all data\n",
      "INFO [2025-10-26 15:15:39] -subtracting expr per gene, use_bounds=TRUE\n",
      "INFO [2025-10-26 15:15:59] \n",
      "\n",
      "\tSTEP 09: apply max centered expression threshold: 3\n",
      "\n",
      "INFO [2025-10-26 15:15:59] ::process_data:setting max centered expr, threshold set to: +/-:  3\n",
      "INFO [2025-10-26 15:15:59] -mirroring for hspike\n",
      "INFO [2025-10-26 15:15:59] ::process_data:setting max centered expr, threshold set to: +/-:  3\n",
      "INFO [2025-10-26 15:16:19] \n",
      "\n",
      "\tSTEP 10: Smoothing data per cell by chromosome\n",
      "\n",
      "INFO [2025-10-26 15:16:19] smooth_by_chromosome: chr: chr1\n",
      "INFO [2025-10-26 15:16:22] smooth_by_chromosome: chr: chr14\n",
      "INFO [2025-10-26 15:16:24] smooth_by_chromosome: chr: chr19\n",
      "INFO [2025-10-26 15:16:26] smooth_by_chromosome: chr: chr10\n",
      "INFO [2025-10-26 15:16:27] smooth_by_chromosome: chr: chr12\n",
      "INFO [2025-10-26 15:16:30] smooth_by_chromosome: chr: chr22\n",
      "INFO [2025-10-26 15:16:32] smooth_by_chromosome: chr: chr3\n",
      "INFO [2025-10-26 15:16:33] smooth_by_chromosome: chr: chr5\n",
      "INFO [2025-10-26 15:16:35] smooth_by_chromosome: chr: chr4\n",
      "INFO [2025-10-26 15:16:37] smooth_by_chromosome: chr: chr9\n",
      "INFO [2025-10-26 15:16:38] smooth_by_chromosome: chr: chr15\n",
      "INFO [2025-10-26 15:16:40] smooth_by_chromosome: chr: chr2\n",
      "INFO [2025-10-26 15:16:42] smooth_by_chromosome: chr: chr11\n",
      "INFO [2025-10-26 15:16:44] smooth_by_chromosome: chr: chr17\n",
      "INFO [2025-10-26 15:16:46] smooth_by_chromosome: chr: chr20\n",
      "INFO [2025-10-26 15:16:48] smooth_by_chromosome: chr: chr8\n",
      "INFO [2025-10-26 15:16:50] smooth_by_chromosome: chr: chr16\n",
      "INFO [2025-10-26 15:16:51] smooth_by_chromosome: chr: chr6\n",
      "INFO [2025-10-26 15:16:53] smooth_by_chromosome: chr: chr7\n",
      "INFO [2025-10-26 15:16:55] smooth_by_chromosome: chr: chr21\n",
      "INFO [2025-10-26 15:16:56] smooth_by_chromosome: chr: chr13\n",
      "INFO [2025-10-26 15:16:58] smooth_by_chromosome: chr: chr18\n",
      "INFO [2025-10-26 15:17:00] -mirroring for hspike\n",
      "INFO [2025-10-26 15:17:00] smooth_by_chromosome: chr: chrA\n",
      "INFO [2025-10-26 15:17:00] smooth_by_chromosome: chr: chr_0\n",
      "INFO [2025-10-26 15:17:00] smooth_by_chromosome: chr: chr_B\n",
      "INFO [2025-10-26 15:17:00] smooth_by_chromosome: chr: chr_0pt5\n",
      "INFO [2025-10-26 15:17:00] smooth_by_chromosome: chr: chr_C\n",
      "INFO [2025-10-26 15:17:00] smooth_by_chromosome: chr: chr_1pt5\n",
      "INFO [2025-10-26 15:17:00] smooth_by_chromosome: chr: chr_D\n",
      "INFO [2025-10-26 15:17:01] smooth_by_chromosome: chr: chr_2pt0\n",
      "INFO [2025-10-26 15:17:01] smooth_by_chromosome: chr: chr_E\n",
      "INFO [2025-10-26 15:17:01] smooth_by_chromosome: chr: chr_3pt0\n",
      "INFO [2025-10-26 15:17:01] smooth_by_chromosome: chr: chr_F\n",
      "INFO [2025-10-26 15:17:22] \n",
      "\n",
      "\tSTEP 11: re-centering data across chromosome after smoothing\n",
      "\n",
      "INFO [2025-10-26 15:17:22] ::center_smooth across chromosomes per cell\n",
      "INFO [2025-10-26 15:17:24] -mirroring for hspike\n",
      "INFO [2025-10-26 15:17:24] ::center_smooth across chromosomes per cell\n",
      "INFO [2025-10-26 15:17:44] \n",
      "\n",
      "\tSTEP 12: removing average of reference data (after smoothing)\n",
      "\n",
      "INFO [2025-10-26 15:17:44] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
      "INFO [2025-10-26 15:17:44] subtracting mean(normal) per gene per cell across all data\n",
      "INFO [2025-10-26 15:17:47] -subtracting expr per gene, use_bounds=TRUE\n",
      "INFO [2025-10-26 15:17:48] -mirroring for hspike\n",
      "INFO [2025-10-26 15:17:48] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
      "INFO [2025-10-26 15:17:48] subtracting mean(normal) per gene per cell across all data\n",
      "INFO [2025-10-26 15:17:51] -subtracting expr per gene, use_bounds=TRUE\n",
      "INFO [2025-10-26 15:18:11] \n",
      "\n",
      "\tSTEP 14: invert log2(FC) to FC\n",
      "\n",
      "INFO [2025-10-26 15:18:11] invert_log2(), computing 2^x\n",
      "INFO [2025-10-26 15:18:11] -mirroring for hspike\n",
      "INFO [2025-10-26 15:18:11] invert_log2(), computing 2^x\n",
      "INFO [2025-10-26 15:18:34] \n",
      "\n",
      "\tSTEP 15: computing tumor subclusters via leiden\n",
      "\n",
      "INFO [2025-10-26 15:18:34] define_signif_tumor_subclusters(p_val=0.1\n",
      "INFO [2025-10-26 15:18:35] define_signif_tumor_subclusters(), tumor: Interm\n",
      "INFO [2025-10-26 15:18:35] Setting auto leiden resolution for Interm to 0.21198\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  AF117829.1, RIPK2, PDP1, PSMA2P2, OSGIN2, MYL12AP1, NBN, DECR1, TMEM67, RBM12B \n",
      "\t   AC120053.1, FAM26F, RWDD1, SLC35F1, AL450405.1, NUS1, RSPH4A, CBX3P9, CEP85L, ZUFSP \n",
      "\t   GOPC, MCM9, TSPYL1, DCBLD1, KPNA5, AL050331.1, DSE, TSPYL4, NT5DC1, FRK \n",
      "Negative:  PRKD2, CALM3, STRN4, PPP5D1, FKRP, PPP5C, SLC1A5, AP2S1, CCDC61, ARHGAP35 \n",
      "\t   MYPOP, IRF2BP1, SYMPK, DMWD, DMPK, DM1-AS, AC074212.1, SIX5, AC007191.1, FBXO46 \n",
      "\t   QPCTL, SNRPD2, GIPR, EML2-AS1, EML2, OPA3, VASP, RTN2, FOSB, ERCC1 \n",
      "PC_ 2 \n",
      "Positive:  HCST, NFKBID, APLP1, PRODH2, TYROBP, PROSER3, LIN37, LRFN3, AD000671.1, SDHAF1 \n",
      "\t   PSENEN, ALKBH6, THAP8, U2AF1L4, WDR62, POLR2I, IGFLR1, TBCB, AD000671.2, CAPNS1 \n",
      "\t   KMT2B, ZNF565, ZBTB32, ZNF146, LINC00665, RPL41P1, GINS1, AL035258.1, PPIAP2, NINL \n",
      "Negative:  BTN3A3, BTN2A3P, BTN2A1, BTN3A1, BTN1A1, BTN2A2, HCG11, BTN3A2, HIST1H3PS1, HMGN4 \n",
      "\t   HIST1H4H, HIST1H2BI, HIST1H3G, HIST1H2BH, HIST1H1D, HIST1H3E, ABT1, HIST1H2AE, HIST1H2BG, HIST1H4E \n",
      "\t   HIST1H3D, HIST1H2BE, ZNF322, AL353759.1, HIST1H2BD, HIST1H1E, HIST1H2AC, HIST1H2BC, HIST1H2BJ, HIST1H4C \n",
      "PC_ 3 \n",
      "Positive:  UBLCP1, RNF145, EBF1, AC008677.2, CLINT1, LSM11, THG1L, NIPAL4, ADAM19, CYFIP2 \n",
      "\t   MED7, HAVCR2, RARS, TIMD4, PANK3, RPL10P9, PPP1R2P3, SPDL1, AC008725.1, DOCK2 \n",
      "\t   LCP2, GEMIN5, RPL10P8, CNOT8, AC016582.2, ZNF573, SIPA1L3, ZNF607, AC093227.2, SPINT2 \n",
      "Negative:  SEMA4A, LMNA, MEX3A, RAB25, LAMTOR2, UBQLN4, SSR2, ARHGEF2, KIAA0907, RIT1 \n",
      "\t   GON4L, MSTO2P, DAP3, ASH1L-AS1, YY1AP1, MSTO1, NDRG1, ASH1L, ST13P6, WISP1-OT1 \n",
      "\t   RUSC1, ST3GAL1, SLA, ZFAT, FDPS, MUC1, KHDRBS3, THBS3, PHF20L1, GBAP1 \n",
      "PC_ 4 \n",
      "Positive:  PLEKHO2, ANKDD1A, AC103691.1, FOXN2, PPP1R21, NRXN1, AC079807.2, ASB3, SPG21, FBXO11 \n",
      "\t   CHAC2, ERLEC1, AC006509.1, PSME4, MTFMT, MSH6, ACYP2, AC079250.1, KCNK12, AC008280.3 \n",
      "\t   PDCD7, MSH2, SPTBN1, EPCAM, EML6, CALM2, RTN4, RPS27A, EPM2AIP1, TRANK1 \n",
      "Negative:  FAM189A2, TJP2, PTAR1, FXN, RPL24P8, FAM122A, SMC5, KLF9, ABHD17B, TMEM2 \n",
      "\t   TRPM3, C9orf85, RPL35AP21, PIP5K1B, BTF3P4, PGM5, ZFAND5, CBWD3, RPS27AP15, FAM27B \n",
      "\t   ANXA1, RORB, ZNF658, C9orf40, ANKRD20A3, CARNMT1, CBWD5, NMRK1, OSTF1, FRG1KP \n",
      "PC_ 5 \n",
      "Positive:  GOLGA5, RIN3, CPSF2, LGMN, ITPK1, NDUFB1, ATXN3, MOAP1, TMEM251, GON7 \n",
      "\t   AL049872.1, UBR7, TRIP11, BTBD7, PPP4R3A, UNC79, CCDC88C, AL122023.1, C14orf159, ASB2 \n",
      "\t   RPS6KA5, DDX24, AL122020.1, IFI27L1, TTC7B, IFI27, CALM1, IFI27L2, SERPINA1, RPL21P11 \n",
      "Negative:  PACRGL, ADGRA3, AC098583.1, SLIT2, GBA3, AC114728.1, CDC42P6, PPARGC1A, LCORL, DHX15 \n",
      "\t   NCAPG, ATP5LP3, DCAF16, MED28, CCDC149, LAP3, QDPR, SEPSECS, RPS21P4, TAPT1-AS1 \n",
      "\t   AC104662.3, TAPT1, CD38, FAM200B, PI4K2B, FBXL5, CPEB2, ZCCHC4, ANAPC4, SEL1L3 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 15:18:35] define_signif_tumor_subclusters(), tumor: Malignant_MM1_Non-Prolif\n",
      "INFO [2025-10-26 15:18:35] Setting auto leiden resolution for Malignant_MM1_Non-Prolif to 0.0838728\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  CCSAP, NUP133, SPHAR, ABCB10, TAF5L, URB2, RAB4A, GALNT2, RHOU, COG2 \n",
      "\t   C1orf198, TTC13, FTH1P2, ARV1, RNF187, MIR1182, AL732414.1, AL109810.1, HIST3H2A, C1orf131 \n",
      "\t   TRIM11, GNPAT, OBSCN, EXOC8, IBA57, SPRTN, GUK1, TARBP1, IRF2BP2, COA6 \n",
      "Negative:  FAM86C1, ALG1L9P, AP002495.1, RNF121, IL18BP, NUMA1, LRTOMT, LAMTOR1, ANAPC15, INPPL1 \n",
      "\t   CLPB, ARAP1, STARD10, ATG16L2, FCHSD2, AP002381.1, P2RY6, MPZL3, CEP164, BACE1-AS \n",
      "\t   IL10RA, RNF214, CD3E, PCSK7, CD3D, CD3G, UBE4A, RELT, TAGLN, ATP5L \n",
      "PC_ 2 \n",
      "Positive:  AC006511.3, NECAP1, FAM90A1, FAM86FP, FOXJ2, SLC2A3, AC092865.4, LAG3, AC131206.1, CD4 \n",
      "\t   CD163L1, HADHAP2, P3H3, GNB3, PEX5, PHC1, CDCA3, CLSTN3, USP5, TPI1 \n",
      "\t   M6PR, C1RL, SPSB2, RPL13P5, C1R, LRRC23, KLRG1, LPCAT3, EMG1, DSTNP2 \n",
      "Negative:  SLC25A20, ARIH2OS, ARIH2, PRKAR2A-AS1, PRKAR2A, IP6K2, NCKIPSD, CELSR3, P4HTM, SLC26A6 \n",
      "\t   WDR6, UQCRC1, DALRD3, PFKFB4, TREX1, SHISA5, AC134772.2, ATRIP, TMA7, CCDC51 \n",
      "\t   PLXNB1, NME6, APH1A, C1orf54, OTUD7B, ANP32E, VPS45, SF3B4, FCF1P2, PLEKHO1 \n",
      "PC_ 3 \n",
      "Positive:  ALDH9A1, MGST3, AL390119.1, TMCO1, PBX1, AL451074.2, NUF2, RGS5, AL451074.5, HSD17B7 \n",
      "\t   DDR2, UCK2, UAP1, UHMK1, ATF6, DUSP12, FCRLB, FCRLA, RPL31P11, FCGR2B \n",
      "\t   FCGR2C, FCGR3A, FCGR2A, AL592295.1, SDHC, TOMM40L, FCER1G, NDUFS2, ADAMTS4, B4GALT3 \n",
      "Negative:  RN7SL645P, CSNK2A2, GINS3, CFAP20, SETD6, USB1, CNOT1, ZNF319, SLC38A7, KATNB1 \n",
      "\t   GOT2, ADGRG1, APOOP5, ADGRG5, CCDC102A, DOK4, AC009094.1, POLR2C, AC009052.1, COQ9 \n",
      "\t   CIAPIN1, RPS27AP16, ARL2BP, AC093536.1, RSPRY1, FAM192A, LINC00920, CPNE2, TK2, NLRC5 \n",
      "PC_ 4 \n",
      "Positive:  GPR18, H2AFZP3, UBAC2, GPR183, STK24, DOCK9, FARP1, TM9SF2, IPO5, RAP2A \n",
      "\t   CLYBL, MBNL2, AL355338.1, UGGT2, PCCA, ABCC4, GPR180, RPL15P18, TGDS, PCCA-AS1 \n",
      "\t   GPC6, GGACT, LMO7, GPC5-AS1, RPS26P47, KCTD12, SLITRK5, RPL7P44, TMTC4, DHX9P1 \n",
      "Negative:  LUZP1, C6orf62, HNRNPR, GMNN, PPIAP29, RIPOR2, ZNF436, CMAHP, AL133268.1, TCEA3 \n",
      "\t   AL024509.2, E2F2, CARMIL1, TRIM38, ID3, HIST1H4B, HIST1H2AB, RPL11, HIST1H2BB, ELOA \n",
      "\t   HIST1H3C, HMGCL, FUCA1, ELOA-AS1, GALE, PITHD1, PNRC2, AL590609.2, LYPLA2, HIST1H1C \n",
      "PC_ 5 \n",
      "Positive:  HIST1H2BK, HIST1H4I, HIST1H2BJ, HIST1H2AG, ZNF322, HIST1H2AH, ABT1, HMGN4, HCG11, BTN1A1 \n",
      "\t   BTN2A1, BTN3A3, BTN2A3P, BTN3A1, BTN2A2, BTN3A2, HIST1H3PS1, HIST1H4H, HIST1H2BI, HIST1H3G \n",
      "\t   HIST1H2BH, HIST1H1D, HIST1H3E, HIST1H2AE, HIST1H2BG, USP35, KCTD21, GAB2, KCTD21-AS1, ALG8 \n",
      "Negative:  NLRP2, RDH13, PPP1R12C, TUBBP2, RGCC, NAA16, VWA8, NLRP7, MTRF1, DGKH \n",
      "\t   PPP6R1, AKAP11, KBTBD7, KIR3DL2, BRSK1, EPSTI1, CALM2P3, LILRB4, AL354696.2, LILRB1 \n",
      "\t   CCDC122, KBTBD6, TMEM150B, LILRA2, WBP4, LACC1, KMT5C, KIR3DX1, ELF1, IL11 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 15:18:37] define_signif_tumor_subclusters(), tumor: Malignant_MM1_Prolif\n",
      "INFO [2025-10-26 15:18:37] Setting auto leiden resolution for Malignant_MM1_Prolif to 0.258251\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  TNIP1, GPX3, DCTN4, ANXA6, RBM22, AC008453.1, SYNPO, NDST1, RPS14, CCDC69 \n",
      "\t   CD74, AL590094.1, GM2A, TCOF1, ZNF326, LRRC8D, LRRC8C, RPL7P1, SLC36A1, FLJ27354 \n",
      "\t   LRRC8B, GBP1P1, HMGXB3, GBP5, SPARC, GBP4, TIGD6, GBP2, SLC26A2, PTGES3P1 \n",
      "Negative:  SLX4IP, MKKS, SDAD1P2, ANKEF1, AL035456.1, LAMP5, JAG1, LAMP5-AS1, BTBD3, PLCB4 \n",
      "\t   PA2G4P2, SPTLC3, PLCB1, TASP1, TMX4, ESF1, AL109618.1, CRLS1, NDUFAF5, MCM8 \n",
      "\t   MACROD2, AL359918.1, TRMT6, CHRM3, C20orf196, AL590135.1, RYR2, GPCPD1, RPS10P2, RPL35P1 \n",
      "PC_ 2 \n",
      "Positive:  SMDT1, FAM109B, NDUFA6, OLA1P1, NAGA, TCF20, WBP2NL, Z83851.1, SEPT3, RRP7A \n",
      "\t   CENPM, RRP7BP, TNFRSF13C, POLDIP3, MIR378I, SREBF2, Z93241.1, AL021453.1, CYB5R3, CCDC134 \n",
      "\t   MEI1, ATP5L2, C22orf46, RPL5P34, SNU13, XRCC6, ARFGAP3, DESI1, PACSIN2, PMM1 \n",
      "Negative:  AL355512.1, DUSP5, SMC3, SMNDC1, PDCD4-AS1, MXI1, ADD3, PDCD4, BBIP1, XPNPEP1 \n",
      "\t   MIR4680, XIAPP1, SHOC2, YWHAZP5, RPL13AP6, GPAM, CFAP58-AS1, ACSL5, ZDHHC6, VTI1A \n",
      "\t   ITPRIP, AC022018.1, GSTO2, AL158212.3, GSTO1, CASP7, SFR1, DCLRE1A, SLK, STN1 \n",
      "PC_ 3 \n",
      "Positive:  AL512625.1, AL512625.3, LINC01410, LGALS17A, EID2, EID2B, SUPT5H, TIMM50, RPS16, PLEKHG2 \n",
      "\t   ZFP36, RNA5SP283, MED29, PAF1, SAMD4B, AL354718.1, AC011445.2, GMFG, LRFN1, CBWD6 \n",
      "\t   PAK4, AL353763.2, THOC6, FRG1HP, MMP25-AS1, GLIDR, IL32, ALDH1B1, ZNF213-AS1, ZNF213 \n",
      "Negative:  SOGA1, RPN2, RBL1, MROH8, MANBAL, TLDC2, SAMHD1, DSN1, SRC, NDRG3 \n",
      "\t   SLA2, BLCAP, C20orf24, TGIF2, CTNNBL1, MYL9, VSTM2L, TTI1, RPRD1B, SNHG17 \n",
      "\t   SNORA60, AEBP1, RALGAPB, ACTR5, PPP1R16B, FAM83D, POLD2, DHX35, NPM1P19, MAFB \n",
      "PC_ 4 \n",
      "Positive:  AEBP1, POLD2, YKT6, NUDCD3, AC004453.1, DDX56, TMED4, NPIPA5, MARF1, PKD1P6 \n",
      "\t   NPIPP1, NDE1, RRN3, OGDH, NTAN1, AC026401.3, AC138932.3, PDXDC1, ZMIZ2, NPIPA1 \n",
      "\t   MYH11, PPIA, NOMO1, AF001548.2, BFAR, FOPNL, H2AFV, PARN, MIR193BHG, AC130651.1 \n",
      "Negative:  CDK14, CLDN12, STEAP1, FZD1, EIF4A1P13, GTPBP10, SRI, MTERF1, ADAM22, AKAP9 \n",
      "\t   LRRD1, DBF4, CYP51A1, SLC25A40, RUNDC3B, KRIT1, AC000120.1, ABCB1, ANKIB1, ABCB4 \n",
      "\t   GATAD1, PEX1, CROT, RBM48, FAM133B, TP53TG1, CDK6, TMEM243, DMTF1, SAMD9 \n",
      "PC_ 5 \n",
      "Positive:  RPS15A, NOMO2, RPL7P47, AC109446.3, XYLT1, NOMO3, ABCC1, AC130651.1, FOPNL, AF001548.2 \n",
      "\t   MYH11, AC026401.3, NDE1, MARF1, NPIPA5, PKD1P6, NPIPP1, RRN3, NTAN1, AC138932.3 \n",
      "\t   PDXDC1, NPIPA1, NOMO1, BFAR, PARN, MIR193BHG, TVP23CP2, MKL2, ERCC4, AC109597.2 \n",
      "Negative:  CNIH4, WDR26, CNIH3, DNAH14, LBR, ENAH, SRP9, EPHX1, AL591895.1, TMEM63A \n",
      "\t   PYCR2, MIR6741, LEFTY2, SDE2, H3F3A, ACBD3, MIXL1, LIN9, AL359742.1, PARP1 \n",
      "\t   ITPKB, PSEN2, CCT6A, PSPH, SNORA22B, NIPSNAP2, SUMO2P3, SUMF2, PSPHP1, VOPP1 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 15:18:37] define_signif_tumor_subclusters(), tumor: Malignant_MM2_Non-Prolif\n",
      "INFO [2025-10-26 15:18:37] Setting auto leiden resolution for Malignant_MM2_Non-Prolif to 0.119745\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  MCEE, NAGK, MPHOSPH10, TEX261, PAIP2B, ANKRD53, CYSTM1, IGIP, PFDN1, PURA \n",
      "\t   ZNF638, ADD2, LINC01024, NRG2, SPR, RPS20P10, EXOC6B, TGFA, SFXN5, UBR1 \n",
      "\t   CXXC5, AC068724.1, TTBK2, RAB11FIP5, TMEM62, CDAN1, SPCS2P1, STARD9, CCNDBP1, SMYD5 \n",
      "Negative:  MYC, PVT1, FAM49B, AC020688.1, ASAP1, EFR3A, PHF20L1, PRNCR1, SLA, WISP1-OT1 \n",
      "\t   FAM84B, NDRG1, ST13P6, TRIB1, ST3GAL1, NSMCE2, ZFAT, KHDRBS3, WASHC5, LINC02055 \n",
      "\t   SQLE, KCNK9, ZNF572, TRAPPC9, MTSS1, AC107375.1, CHRAC1, NDUFB9, AGO2, AC090198.1 \n",
      "PC_ 2 \n",
      "Positive:  BIRC5, AFMID, TK1, SYNGR2, EIF5AP2, TMC8, TMC6, TNRC6C-AS1, TNRC6C, SEPT9 \n",
      "\t   MIR6516, TRIM16, ZNF286A, UBE2SP1, TVP23C, AC005517.2, TVP23C-CDRT4, AC013248.2, IL6STP1, HS3ST3B1 \n",
      "\t   AC015922.4, ZSWIM7, AC005224.1, TTC19, NCOR1, COX10, COX10-AS1, ELAC2, ARHGAP44, MAP2K4 \n",
      "Negative:  NXPE4, NXPE1, REXO2, RBM7, AP000805.1, C11orf71, CADM1, TUBAP2, SLC36A4, SMCO4 \n",
      "\t   DISC1FP1, CHORDC1, ZBTB16, LINC00900, FZD4, TMEM135, NCAM1, NAALAD2, AP001784.1, HIKESHI \n",
      "\t   EED, CTSC, USP28, TTC12, AP000797.2, PICALM, AP002884.1, ATF4P4, SYTL2, ZW10 \n",
      "PC_ 3 \n",
      "Positive:  EIF1P3, IER5, MR1, OVAAL, STX6, GLUL, XPR1, VDAC1P4, ACBD6, QSOX1 \n",
      "\t   RNASEL, CEP350, RGS16, TOR1AIP1, NPL, DHX9, TOR1AIP2, LAMC1, SOAT1, SMG7 \n",
      "\t   NCF2, EIF4A1P11, ARPC5, RGL1, SNORA67, TSEN15, ABL2, C1orf21, EDEM3, TOR3A \n",
      "Negative:  CCDC15, ROBO3, MSANTD2, SLC37A2, ESAM, TMEM218, FEZ1, SPA17, EI24, SIAE \n",
      "\t   STT3A, TBRG1, CHEK1, VWA5A, HYLS1, ZNF202, PUS3, CDON, AP002765.1, AP000926.3 \n",
      "\t   NAP1L1P1, AP003040.1, RPUSD4, C11orf63, UBASH3B, ARHGEF12, FAM118B, SORL1, TMEM136, NECTIN1 \n",
      "PC_ 4 \n",
      "Positive:  EXOSC9, CCNA2, BBS7, KIAA1109, CETN4P, BBS12, FGF2, AC093591.1, NUDT6, PLK4 \n",
      "\t   INTU, AC099340.1, SPATA5, MFSD8, ANKRD50, LARP1B, SPRY1, ABHD18, FAT4, PGRMC2 \n",
      "\t   AC108062.1, JADE1, SCLT1, C4orf33, PCDH10, AC104619.1, AC104619.3, SLC7A11, NOCT, ELF2 \n",
      "Negative:  SLC39A14, BMP1, REEP4, HR, POLR3D, NUDT18, FAM160B2, PPP3CC, SORBS3, DMTN \n",
      "\t   PDLIM2, C8orf58, XPO7, CCAR2, DOK2, ATP6V1B2, BIN3, INTS10, CSGALNACT1, PEBP4 \n",
      "\t   SH2D4A, AC107959.1, PSD3, RHOBTB2, TNFRSF10B, TNFRSF10D, TNFRSF10A, AC025062.1, AC100861.1, ADAM28 \n",
      "PC_ 5 \n",
      "Positive:  ARGLU1, ERCC5, BIVM, TEX30, ITGBL1, TPP2, TMTC4, RPS26P47, GGACT, PCCA-AS1 \n",
      "\t   RPL15P18, PCCA, AL355338.1, CLYBL, TM9SF2, GPR183, GPR18, H2AFZP3, UBAC2, EIF4A1P7 \n",
      "\t   POMP, DOCK9, SLC46A3, CYP51A1P2, UBE2L5P, HMGB1, PTPN2P2, LINC00426, KATNAL1, USPL1 \n",
      "Negative:  DYNC1LI2, NAE1, PDP2, CMTM4, FAM96B, CMTM3, CES2, CES3, CKLF-CMTM1, CKLF \n",
      "\t   CBFB, TK2, C16orf70, LINC00920, B3GNT9, TRADD, AC093536.1, FBXL8, RPS27AP16, NOL3 \n",
      "\t   AC009094.1, KIAA0895L, APOOP5, GOT2, E2F4, SLC38A7, CNOT1, LRRC29, SETD6, TMEM208 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 15:18:38] define_signif_tumor_subclusters(), tumor: Malignant_MM2_Prolif\n",
      "INFO [2025-10-26 15:18:38] Setting auto leiden resolution for Malignant_MM2_Prolif to 0.217096\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  MTCO1P40, AC034268.1, TOM1L1, COX11, CA10, AKAP1, AC007485.2, MSI2, SCPEP1, AC007114.1 \n",
      "\t   VEZF1, STXBP4, COIL, AC015912.3, RPL7P48, MMD, SRSF1, TRIM25, DGKE, PCTP \n",
      "\t   DYNLL2, C17orf67, AC015813.6, UTP18, MKS1, MPO, MBTD1, TSPOAP1-AS1, AC004687.1, SUPT4H1 \n",
      "Negative:  UFM1, PROSER1, LINC02334, TRPC4, NHLRC3, COG6, POSTN, LINC00598, SUPT20H, FOXO1 \n",
      "\t   EXOSC8, SLC25A15, ALG5, RFXAP, MIR621, SPART, ELF1, CCDC169-SOHLH2, DCLK1, WBP4 \n",
      "\t   NBEA, RFC3, STARD13, AL121603.2, KBTBD6, PDS5B, N4BP2L2, BAZ1A, AL354696.2, AL137247.1 \n",
      "PC_ 2 \n",
      "Positive:  RWDD3, TMEM56, ALG14, AC017013.1, AC016687.1, CNN3, ARAP2, RELL1, AC108022.1, SLC44A3 \n",
      "\t   PGM2, ABCD3, TBC1D1, KLF3, GCLM, TLR10, TLR1, DNTTIP2, TLR6, AL109613.1 \n",
      "\t   FAM114A1, TMEM156, BCAR3, KLHL5, AC079921.2, WDR19, RFC1, EXOSC6, PDPR, AARS \n",
      "Negative:  C12orf57, U47924.2, ATN1, PTPN6, ENO2, DSTNP2, LRRC23, PHB2, RPL13P5, SPSB2 \n",
      "\t   TPI1, PSMD4, PIP5K1A, AL391069.2, VPS72, SCARNA12, ZNF687, USP5, TMOD4, PI4KB \n",
      "\t   LYSMD1, SCNM1, RFX5, CDCA3, AL391069.3, TNFAIP8L2, EMG1, PSMB4, GABPB2, GNB3 \n",
      "PC_ 3 \n",
      "Positive:  TIMM8B, C11orf57, DLAT, DIXDC1, SDHD, RPL37AP8, PTS, C11orf1, FDXACB1, AP002884.1 \n",
      "\t   NCAM1, TTC12, ZW10, ATF4P4, FZD4, TMEM135, AP001784.1, HIKESHI, CTSC, NAALAD2 \n",
      "\t   CHORDC1, DISC1FP1, EED, TUBAP2, AP000805.1, PICALM, USP28, SLC36A4, SYTL2, SMCO4 \n",
      "Negative:  MAP2K4, ZNF18, ARHGAP44, ADPRM, ELAC2, TMEM220, COX10-AS1, SCO1, COX10, AC005224.1 \n",
      "\t   RPS27AP1, GAS7, HS3ST3B1, STX8, PIK3R5, AC025518.1, MYH10, AC013248.2, NDEL1, RPL26 \n",
      "\t   TVP23C-CDRT4, KRBA2, AC135178.3, AC005517.2, SNORA69, AC135178.6, TVP23C, RANGRF, SLC25A35, TRIM16 \n",
      "PC_ 4 \n",
      "Positive:  SELL, C1orf112, SLC19A2, CCDC181, BLZF1, METTL18, NME7, ATP1B1, TBX19, SCYL3 \n",
      "\t   SFT2D2, TIPRL, KIFAP3, GCSHP5, GORAB, MPZL1, DCAF6, MPC2, RCSD1, HMGB1P11 \n",
      "\t   CREG1, FMO4, POU2F1, PRRC2C, MYOCOS, PFN1P1, AL451050.1, RPL4P3, VAMP4, METTL13 \n",
      "Negative:  RNFT1, RPS6KB1, AC005702.1, AC004686.1, MIR4737, TUBD1, MIR21, VMP1, HEATR6, PTRH2 \n",
      "\t   USP32, RPL12P38, CLTC, AC091271.1, APPBP2, DHX40, PPM1D, YPEL2, BCAS3, GDPD1 \n",
      "\t   AC110602.1, SMG8, RPL23AP74, PRR11, SKA2, NACA2, TRIM37, RAD51C, BRIP1, TEX14 \n",
      "PC_ 5 \n",
      "Positive:  TLK2, AC008026.1, METTL2A, AC008026.2, AC008158.1, TANC2, AC018628.1, AC005828.4, MED13, INTS2 \n",
      "\t   CYB561, BRIP1, NACA2, DCAF7, RPL23AP74, AC110602.1, TACO1, BCAS3, MAP3K3, PPM1D \n",
      "\t   STRADA, APPBP2, LIMD2, RPL12P38, USP32, AC046185.3, HEATR6, CCDC47, MIR4737, AC005702.1 \n",
      "Negative:  PHLDA3, TMEM9, CSRP1, KIF21B, AC096677.1, RPS10P7, INAVA, NAV1, GPR25, IPO9 \n",
      "\t   RPL34P6, TIMM17A, CAMSAP2, RPL10P4, RNPEP, DDX59, ELF3, KIF14, AL691482.3, ZNF281 \n",
      "\t   ARL8A, AC096631.2, PTPN7, UBE2T, PTPRC, PPP1R12B, NEK7, AC099336.2, C1orf53, KDM5B \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 15:18:39] define_signif_tumor_subclusters(), tumor: Malignant_MM3_Non-Prolif\n",
      "INFO [2025-10-26 15:18:39] Setting auto leiden resolution for Malignant_MM3_Non-Prolif to 0.151456\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  SHC1, CKS1B, FLAD1, ZBTB7B, URB2, TTC13, ARV1, COG2, C1orf198, GALNT2 \n",
      "\t   TAF5L, DCST1-AS1, MIR1182, AL732414.1, ABCB10, AL109810.1, C1orf131, NUP133, GNPAT, CCSAP \n",
      "\t   ADAM15, EXOC8, SPHAR, EFNA4, SPRTN, RAB4A, EGLN1, EFNA3, RHOU, AL445524.1 \n",
      "Negative:  CLCC1, GPSM2, WDR47, BX679664.1, STXBP3, BX679664.3, TAF13, PRPF38B, HENMT1, AL356488.3 \n",
      "\t   FAM102B, SLC25A24P1, TMEM167B, NBPF4, AL356488.2, SLC25A24, SARS, VAV3, CELSR2, SORT1 \n",
      "\t   PSMA5, ATXN7L2, CYB561D1, AMIGO1, GNAI3, AMPD2, RPL7P8, GSTM4, GSTM2, GSTM3 \n",
      "PC_ 2 \n",
      "Positive:  ENPP6, AC074194.2, STOX2, TRAPPC11, RWDD4, ING2, CDKN2AIP, CD48, SLAMF1, AL121985.1 \n",
      "\t   SLAMF7, LY9, CD84, AL354714.2, F11R, SLAMF6, AL591806.3, USF1, MBD2, MYL12A \n",
      "\t   ITSN1, ARHGAP30, ATP5O, POLI, DCC, AC022701.1, NECTIN4, C18orf54, NCSTN, LINC00649 \n",
      "Negative:  GTF2A1, AL136040.2, TSHR, CEP128, DYNLL1P1, ADCK1, STON2, SNW1, DYNLL1P2, SLIRP \n",
      "\t   RPL21P10, SEL1L, ALKBH1, EEF1A1P2, SPTLC2, FLRT2, AHSA1, LINC02328, VIPAS39, GALC \n",
      "\t   TMED8, GPR65, GSTZ1, SPATA7, POMT2, ZC3H14, TMEM63C, EML5, CIPC, AC007686.3 \n",
      "PC_ 3 \n",
      "Positive:  HMGB1P5, LPP, TPRG1, BCL6, TP63, SGO1, RTP4, P3H2, RPL39L, IL1RAP \n",
      "\t   ST6GAL1, KAT2B, OSTN-AS1, RAB5A, SATB1-AS1, UTS2B, DPH3, SATB1, RFTN1, OXNAD1 \n",
      "\t   RN7SL4P, AC104297.1, ANKRD28, CCDC50, RPS20P14, TBC1D5, PLCL2, BTD, AC073365.1, HACL1 \n",
      "Negative:  CAMLG, DDX46, SEC24A, SAR1B, C5orf24, TXNDC15, PCBD2, JADE2, AC006077.2, PITX1 \n",
      "\t   H2AFY, CDKN2AIPNL, CXCL14, UBE2B, TGFBI, CDKL3, SMAD5, AC104109.4, KLHL3, PPP2CA \n",
      "\t   AC104109.3, HNRNPA0, AC106791.1, FAM13B, BRD8, KIF20A, CDC23, CDC25C, TRADD, FBXL8 \n",
      "PC_ 4 \n",
      "Positive:  EEF1A1P13, AC010491.1, CCT6P2, OTULIN, ANKH, FAM105A, RBBP4P1, TRIO, FBXL7, DAP \n",
      "\t   ZNF622, ANKRD33B, RETREG1, MYO10, MARCH6, RPS26P28, BASP1, PFKFB2, AC098935.1, AC012640.2 \n",
      "\t   YOD1, FCMR, AL445493.1, AL591846.1, CMBL, FTH1P10, MAPKAPK2, CD3E, IL10RA, MPZL3 \n",
      "Negative:  SYCP3, GNPTAB, CHPT1, RNU6-101P, RPS27P23, DRAM1, ARL1, WASHC3, NUP37, UTP20 \n",
      "\t   PARPBP, SCYL2, HELLPAR, IGF1, ACTR6, C12orf42, GOLGA2P5, NT5DC3, AC126474.2, TTC41P \n",
      "\t   UHRF1BP1L, C12orf73, RPS4XP1, TDG, AC078819.1, AC008126.1, HCFC2, APAF1, NFYB, IKBIP \n",
      "PC_ 5 \n",
      "Positive:  AL353795.3, VCP, FANCG, YWHAZP6, AL162231.2, PIGO, AL162231.1, IL11RA, GALT, STOML2 \n",
      "\t   SIGMAR1, FAM214B, ARID3C, UNC13B, DCTN3, ATP8B5P, RPP25L, RPS29P17, MIR4667, FAM219A \n",
      "\t   SIT1, C9orf24, RMRP, NUDT2, AL353662.2, CCDC107, ARHGEF39, UBAP1, TLN1, DCAF12 \n",
      "Negative:  SNHG18, SNORD123, FAM173B, CCT5, CMBL, FAM114A2, MFAP3, GALNT10, G3BP1, SAP30L \n",
      "\t   AC012640.2, CIR1P1, AC091982.4, LARP1, GM2A, CCDC69, MARCH6, ANXA6, FAXDC2, SLC36A1 \n",
      "\t   TNIP1, SPARC, CNOT8, AC091982.1, AC011374.2, ATOX1, GPX3, GEMIN5, AC008725.1, FGD6 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 15:18:40] define_signif_tumor_subclusters(), tumor: Malignant_MM3_Prolif\n",
      "INFO [2025-10-26 15:18:40] Setting auto leiden resolution for Malignant_MM3_Prolif to 0.209519\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  ZNF343, NOP56, MIR1292, SNORD86, IDH3B, TTC13, ARV1, C1orf198, MIR1182, COG2 \n",
      "\t   AL732414.1, PCED1A, GALNT2, URB2, VPS16, TAF5L, ABCB10, PTPRA, NUP133, CCSAP \n",
      "\t   SPHAR, OXT, RAB4A, RHOU, LIN9, PARP1, MIXL1, ITPKB, AL359742.1, FTH1P2 \n",
      "Negative:  GAREM1, RNF138, HNRNPA1P7, RNF125, AC009831.2, KLHL14, TRAPPC8, NOL4, DSG2, DTNA \n",
      "\t   CDH2, MAPRE2, AC018371.2, ZNF397, AC007996.1, ZSCAN30, AC011815.1, KCTD1, ZNF271P, TAF4B \n",
      "\t   ZNF24, ZNF396, DHFRP1, INO80C, PSMA8, GALNT1, SS18, C18orf21, RPRD1A, AC027229.1 \n",
      "PC_ 2 \n",
      "Positive:  SDHC, AL592295.1, TOMM40L, FCGR2A, FCER1G, FCGR3A, NDUFS2, ADAMTS4, FCGR2C, B4GALT3 \n",
      "\t   PPOX, USP21, UFC1, AL590714.1, FCGR2B, ACA64, DEDD, NIT1, RPL31P11, PFDN2 \n",
      "\t   TNFRSF13C, CENPM, MIR378I, SREBF2, KLHDC9, NECTIN4, FCRLA, AL021453.1, SEPT3, ARHGAP30 \n",
      "Negative:  PHF14, NDUFA4, THSD7A, AC007128.1, TMEM106B, VWDE, ICA1, AC006042.3, AC013470.3, GLCCI1 \n",
      "\t   AC006042.4, SCIN, UMAD1, RPA3, ARL4A, MIOS, C1GALT1, CCZ1B, AC005019.1, PMS2CL \n",
      "\t   ZNF12, ZNF316, C7orf26, ZDHHC4, KDELR2, DAGLB, RAC1, FAM220A, CYTH3, USP42 \n",
      "PC_ 3 \n",
      "Positive:  DCBLD1, GOPC, NUS1, KPNA5, ZUFSP, SLC35F1, RSPH4A, AL450405.1, RWDD1, FAM26F \n",
      "\t   CBX3P9, CEP85L, MFSD4B, TSPYL1, AL080317.1, REV3L, SLC16A10, NT5DC1, TSPYL4, TRAF3IP2-AS1 \n",
      "\t   DSE, AL050331.1, RPF2, FRK, TRAF3IP2, FYN, HDAC2, TUBE1, FAM229B, MCM9 \n",
      "Negative:  MRFAP1L1, AC093323.1, BLOC1S4, LINC02482, KIAA0232, MRFAP1, CYTL1, WFS1, MAN2B2, LDHAP1 \n",
      "\t   TBC1D14, RPS7P15, TADA2B, GRPEL1, STX18, LINC02447, ZBTB49, AFAP1-AS1, GMPSP1, LYAR \n",
      "\t   SH3TC1, AC104825.2, ACOX3, TRMT44, UNC93B8, TMEM128, FAM86EP, LRPAP1, RGS12, HTT \n",
      "PC_ 4 \n",
      "Positive:  HADH, CYP2U1, LEF1, PAPSS1, RPL34, OSTC, AIMP1, SEC24B, TBCK, MCUB \n",
      "\t   NPNT, GSTCD, CASP6, INTS12, PLA2G12A, EEF1A1P9, CFI, PPA2, GAR1, TET2 \n",
      "\t   ELOVL6, CXXC4, CENPE, BDH2, SLC9B2, SLC9B1, CISD2, AC018797.2, UBE2D3, MANBA \n",
      "Negative:  UBLCP1, NAP1L1P3, RPL7AP11, POLQ, GTF2E1, RABL3, HCLS1, NDUFB4, AC063952.2, LRRC58 \n",
      "\t   RNU4-62P, GOLGB1, STK10, IQCB1, EAF2, FBXW11, SLC15A2, ILDR1, NPM1, CD86 \n",
      "\t   RPL10P8, CSTA, CCDC58, LCP2, FAM162A, WDR5B, AC083798.2, DOCK2, KPNA1, SPDL1 \n",
      "PC_ 5 \n",
      "Positive:  COL4A3BP, HMGCR, ANKRD31, GCNT4, RPL27AP5, SCNM1, LYSMD1, TNFAIP8L2, FAM169A, GABPB2 \n",
      "\t   NSA2, MLLT11, GFM2, CDC42SE1, HEXB, C1orf56, ENC1, PRUNE1, MINDY1, UTP15 \n",
      "\t   CERS2, ANKRA2, SETDB1, FUNDC2P1, ARNT, BTF3, CTSK, FOXD1, CTSS, AC116345.1 \n",
      "Negative:  AL031847.1, RNF207, ICMT, ACOT7, NOL9, ZBTB48, KLHL21, PHF13, THAP3, CAMTA1 \n",
      "\t   VAMP3, PER3, TPT1, TPT1-AS1, AL138963.3, GTF2F2, RCN1P2, NUFIP1, AL359706.1, GPALPP1 \n",
      "\t   SMARCE1P5, SLC25A30, TSC22D1, LACC1, COG3, CCDC122, EPSTI1, AKAP11, LINC01055, DGKH \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 15:18:40] define_signif_tumor_subclusters(), tumor: Malignant_MM4_Non-Prolif\n",
      "INFO [2025-10-26 15:18:41] Setting auto leiden resolution for Malignant_MM4_Non-Prolif to 0.0741977\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  LINC01781, ADGRL2, PSAT1P3, AC093580.1, IFI44, HS2ST1, SELENOF, TTLL7, AL049597.1, AC093155.3 \n",
      "\t   SH3GLB1, PRKACB, ODF2L, COL24A1, IFI44L, ZNHIT6, LINC01140, AL450063.1, DDAH1, BCL10 \n",
      "\t   RPF1, C1orf52, GNG5, LMO4, SPATA1, SSX2IP, CTBS, AC103591.1, PKN2, GTF2B \n",
      "Negative:  ZNF354A, CLK4, ZNF354B, PIGFP1, PHYKPL, HNRNPAB, AC136632.2, NHP2, RMND5B, N4BP3 \n",
      "\t   AC136632.1, FAM153C, SUDS3P1, AC106795.1, AC139795.1, B4GALT7, TMED9, FAF2, RNF44, FAM193B \n",
      "\t   CLTB, HIGD2A, GPRIN1, DDX41, NOP16, TSPAN17, DOK3, ARL10, UIMC1, LMAN2 \n",
      "PC_ 2 \n",
      "Positive:  EPHB1, CEP63, MSL2, AC092991.1, ANAPC13, PCCB, STAG1, RPL39P5, AC117382.1, SLC35G2 \n",
      "\t   RYK, AC096992.2, SLCO2A1, NCK1-AS1, NCK1, SRPRB, IL20RB, DBR1, ARMC8, MRAS \n",
      "\t   CEP70, FAIM, PIK3CB, RNF7, EEF1A1P25, YWHAQP6, PXYLP1, ZBTB38, ATP1B3, SLC25A36 \n",
      "Negative:  RPS8, KIF2C, RNU5D-1, SNORD55, RNU5F-1, C1orf228, SNORD38A, RPS15AP11, PLK3, BTBD19 \n",
      "\t   RNU5E-6P, AL136380.1, EIF2B3, HECTD3, UROD, ZSWIM5, HPDL, MUTYH, TOE1, TESK2 \n",
      "\t   CCDC163, MMACHC, PRDX1, AKR1A1, NASP, GPBP1L1, TMEM69, IPP, AL604028.1, AL603882.1 \n",
      "PC_ 3 \n",
      "Positive:  AC131206.1, CD163L1, PEX5, CLSTN3, C1RL, C1R, LPCAT3, EMG1, SCARNA12, PHB2 \n",
      "\t   PTPN6, C12orf57, U47924.2, ATN1, ENO2, DSTNP2, LRRC23, RPL13P5, SPSB2, TPI1 \n",
      "\t   USP5, CDCA3, GNB3, P3H3, CD4, LAG3, PTMS, MLF2, COPS7A, ZNF384 \n",
      "Negative:  BECN1, PSME3, COA3, RUNDC1, AARSD1, RPL27, PTGES3L-AARSD1, VPS25, IFI35, RAMP2 \n",
      "\t   HMGN2P15, VAT1, EZH1, RND2, CNTNAP1, AC100793.2, BRCA1, PSMC3IP, RETREG3, CCR10 \n",
      "\t   TUBG1, PLEKHH3, TUBG2, AC135721.1, NBR2, AC060780.1, NBR1, TMEM106A, FCRL1, CD5L \n",
      "PC_ 4 \n",
      "Positive:  PLEKHG1, ULBP3, BTF3P10, LRP11, BTBD10P2, PCMT1, NUP43, LATS1, RPS18P9, KATNA1 \n",
      "\t   GINM1, PPIL4, AL031133.1, SUMO4, ZC3H12D, AL603766.1, TAB2, STXBP5, KATNBL1P6, RAB32 \n",
      "\t   SHPRH, AL356599.1, FBXO30, EPM2A, UTRN, TPT1P4, STX11, SF3B5, PLAGL1, ZC2HC1B \n",
      "Negative:  UBASH3B, VWA5A, ZNF202, C11orf63, TBRG1, AP003040.1, AP002765.1, SORL1, AP000926.3, SIAE \n",
      "\t   SPA17, ESAM, SC5D, MSANTD2, ROBO3, CCDC15, TECTA, SLC37A2, TBCEL, TMEM218 \n",
      "\t   FEZ1, GRIK4, EI24, ARHGEF12, STT3A, TMEM136, OAF, CHEK1, NECTIN1, HYLS1 \n",
      "PC_ 5 \n",
      "Positive:  MARS, ARHGAP9, DDIT3, MBD6, INHBE, DCTN2, R3HDM2, PIP4K2C, NDUFA4L2, DTX3 \n",
      "\t   SHMT2, OS9, NXPH4, LRP1, STAT6, AGAP2, NAB2, NEMP1, ZBTB39, TSPAN31 \n",
      "\t   PRIM1, NACA, PTGES3, MIR6759, SNORD59A, CDK4, ATP5B, MARCH9, BAZ2A, CYP27B1 \n",
      "Negative:  PDCD7, MTFMT, SPG21, AC103691.1, ANKDD1A, PLEKHO2, PIF1, OAZ2, AC090543.1, AC090543.2 \n",
      "\t   ZNF609, TRIP4, PCLAF, AC087632.1, CSNK1G1, UBL7-AS1, PPIB, SNX22, ARID3B, SNX1 \n",
      "\t   CLK3, FAM96A, EDC3, HERC1, FBXL22, CSK, USP3-AS1, USP3, ULK3, APH1B \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 15:18:42] define_signif_tumor_subclusters(), tumor: Malignant_MM4_Prolif\n",
      "INFO [2025-10-26 15:18:42] Setting auto leiden resolution for Malignant_MM4_Prolif to 0.133595\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  FAM78B, MIR3658, UCK2, AL596087.1, AL451074.5, RPL4P2, AL451074.2, POGK, TMCO1, TADA1 \n",
      "\t   ALDH9A1, MGST3, AL451050.1, AL390119.1, POU2F1, PBX1, CREG1, NUF2, RCSD1, MPZL1 \n",
      "\t   RGS5, MPC2, DCAF6, HSD17B7, GCSHP5, TIPRL, SFT2D2, DDR2, TBX19, ATP1B1 \n",
      "Negative:  ARFIP2, TIMM10B, TRIM3, APBB1, DNHD1, SMPD1, RRP8, AC090589.1, MIR6745, ACP2 \n",
      "\t   DDB2, PACSIN3, NR1H3, CAVIN3, ARFGAP2, MADD, AC091564.2, FAM160A2, C11orf49, SPI1 \n",
      "\t   CKAP5, ILK, TRIM22, ZNF408, SLC39A13, ARHGAP1, PSMC3, TRIM5, ATG13, TPP1 \n",
      "PC_ 2 \n",
      "Positive:  LRRC75A, SNORD49A, LRRC75A-AS1, TRPV2, UBB, AC093484.4, CENPV, PIGL, NCOR1, TTC19 \n",
      "\t   ZSWIM7, AC015922.4, IL6STP1, UBE2SP1, ZNF286A, TRIM16, ST20-AS1, BCL2A1, ST20, ST20-MTHFS \n",
      "\t   MTHFS, KIAA1024, TMED3, RASGRF1, CTSH, AC103975.1, MORF4L1, AC067863.1, CHRNA5, PSMA4 \n",
      "Negative:  PBRM1, GNL3, SNORD19, GLT8D1, SPCS1, NEK4, TMEM110, AC069271.1, NUP210, IQSEC1 \n",
      "\t   MKRN2, RAF1, TSEN2, CAND2, HDAC11, SNORA7A, AC091492.1, RPL32, TAMM41, FBLN2 \n",
      "\t   MIR1226, VGLL4, MAP4, CDC25A, SFMBT1, ATG7, NDUFB1P1, DHX30, SEC13, CHCHD4 \n",
      "PC_ 3 \n",
      "Positive:  MOAP1, ITPK1, TMEM251, GOLGA5, LGMN, RIN3, GON7, UBR7, CPSF2, BTBD7 \n",
      "\t   NDUFB1, UNC79, ATXN3, AL122023.1, AL049872.1, ASB2, TRIP11, DDX24, PPP4R3A, KCNK13 \n",
      "\t   IFI27L1, CCDC88C, PSMC1, TDP1, NRDE2, RPL21P11, IFI27, C14orf159, CALM1, RPS6KA5 \n",
      "Negative:  GRK6, PRR7, F12, DBN1, PDLIM7, DOK3, RGS14, DDX41, FAM193B, LMAN2 \n",
      "\t   TMED9, B4GALT7, AC139795.1, PRELID1, AC106795.1, SUDS3P1, MXD3, FAM153C, AC136632.1, NSD1 \n",
      "\t   ZNF346, N4BP3, RMND5B, UIMC1, NHP2, TSPAN17, AC136632.2, GPRIN1, HNRNPAB, RNF44 \n",
      "PC_ 4 \n",
      "Positive:  HIVEP1, RPL15P3, PHACTR1, AL008729.1, TBC1D7, GFOD1, RPS4XP7, SIRT5, AL441883.1, NOL7 \n",
      "\t   RANBP9, MCUR1, AL136162.1, JARID2, DTNBP1, ARPC3P5, MYLIP, GMPR, ATXN1, AL137003.1 \n",
      "\t   AL137003.2, FAM8A1, NUP153, AL138724.1, KIF13A, TPMT, KDM1B, DEK, RNF144B, AL589647.1 \n",
      "Negative:  RETSAT, TGOLN2, ELMOD3, KCMF1, CAPG, AC022210.1, MAT2A, GGCX, VAMP8, VAMP5 \n",
      "\t   AC005034.3, GCFC2, POLE4, HK2, AC019069.1, AC007387.1, SEMA4F, DOK1, HTRA2, AUP1 \n",
      "\t   CWF19L1, PCGF1, CHUK, ERLIN1, LBX2-AS1, TPM4P1, DNMBP, TTC31, CCDC142, MOGS \n",
      "PC_ 5 \n",
      "Positive:  CNP, NKIRAS2, DHX58, KAT2A, RAB5C, GHDC, STAT5B, STAT5A, STAT3, CAVIN1 \n",
      "\t   ATP6V0A1, ARHGEF11, Y-RNA, AC067852.4, SH2D2A, NAGLU, PRCC, PARK7, AC067852.2, COASY \n",
      "\t   HDGF, MLX, PER3, PSMC3IP, RRNAD1, RETREG3, ISG20L2, VAMP3, TUBG1, NES \n",
      "Negative:  PLTP, CTSA, NEURL2, PCIF1, ZNF335, SLC12A5, NCOA5, CD40, SLC35C2, ELMO2 \n",
      "\t   OCSTAMP, SLC13A3, AL133520.1, TP53RK, SLC2A10, EYA2, INO80B, INO80B-WBP1, WBP1, MOGS \n",
      "\t   LBX2-AS1, RTKN, TTC31, PCGF1, CCDC142, AUP1, WDR54, HTRA2, DOK1, C2orf81 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 15:18:43] define_signif_tumor_subclusters(), tumor: Malignant_MM5_Non-Prolif\n",
      "INFO [2025-10-26 15:18:43] Setting auto leiden resolution for Malignant_MM5_Non-Prolif to 0.126668\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  AC145285.4, AC145285.1, AC145285.2, EIF3C, SULT1A1, SULT1A2, SGF29, AC125807.1, TULP3, RHNO1 \n",
      "\t   FOXM1, AC005912.1, ITFG2, FKBP4, ITFG2-AS1, CRACR2A, CACNA1C-AS2, DCP1B, AC005831.1, PARP11 \n",
      "\t   AC007207.1, CCND2, TIGAR, AC006122.1, C12orf4, KCTD5, RAD51AP1, PRSS27, DYRK4, SRRM2 \n",
      "Negative:  AC114786.2, UGT2A2, ENAM, UGT2B15, UTP3, UGT2B17, RUFY3, GRSF1, MOB1B, COX18 \n",
      "\t   DCK, ANKRD17, HMGA1P2, RASSF6, TMPRSS11E, PF4V1, GALK2, MT2P1, AC093677.1, FAM227B \n",
      "\t   AC022306.2, MTHFD2L, DTWD1, SLC27A2, AREG, GABPB1, PARM1, GABPB1-AS1, LINC02562, USP8 \n",
      "PC_ 2 \n",
      "Positive:  E2F4, KIAA0895L, LRRC29, NOL3, FBXL8, TMEM208, TRADD, B3GNT9, FHOD1, C16orf70 \n",
      "\t   ATP6V0D1, CBFB, CES3, AC009061.2, CES2, FAM96B, PDP2, NAE1, DYNC1LI2, CMTM4 \n",
      "\t   CMTM3, CKLF-CMTM1, CKLF, TK2, MBTPS1, LINC00920, MLYCD, HSBP1, AC093536.1, AL606491.1 \n",
      "Negative:  GSR, UBXN8, PPP2CB, AC090281.1, GTF2E2, AC026979.3, WRN, DCTN6, AC090204.1, AC104027.1 \n",
      "\t   AC067838.1, FUT10, TTI2, MAK16, SNORD13, RNF122, DUSP26, AF279873.3, AF279873.2, AF279873.1 \n",
      "\t   AC087343.1, LSM12P1, AC124290.1, ZNF703, ERLIN2, PLPBP, BRF2, HIVEP1, RPL15P3, RAB11FIP1 \n",
      "PC_ 3 \n",
      "Positive:  TRIM45, TTF2, MAN1A2, CD2, NAP1L4P1, FAM46C, CD58, ATP1A1-AS1, SNORA40, ATP1A1 \n",
      "\t   PNRC2P1, U3, GDAP2, SLC22A15, WDR3, VANGL1, WARS2, AL359915.2, ZNF697, PHGDH \n",
      "\t   AL359752.1, NOTCH2, SEC22B, AC244669.1, U1, NBPF8, TTLL7, AC093580.1, FUBP1, AC103591.1 \n",
      "Negative:  TPD52L1, HDDC2, AL365259.1, HEY2, NCOA7, HINT3, TRMT11, CENPW, PRELID1P1, RNF146 \n",
      "\t   ECHDC1, RPL5P18, SOGA3, PTPRK, EEF1DP5, LAMA2, ARHGAP18, L3MBTL3, SAMD3, TMEM200A \n",
      "\t   EPB41L2, AKAP7, MED23, ENPP1, AL139805.1, CTGF, AL133346.1, LINC01013, MOXD1, STX7 \n",
      "PC_ 4 \n",
      "Positive:  JKAMP, L3HYPDH, PCNX4, DAAM1, DHRS7, PSMA3, AL135752.1, AL049873.1, ACTR10, TIMM9 \n",
      "\t   PSMA3-AS1, HNRNPCP1, ARID4A, KIAA0586, PPM1A, UBA52P3, C14orf37, NAA30, RBM8B, AP5M1 \n",
      "\t   SIX1, EXOC5, AL137100.1, RPL3P3, AL161757.4, TMEM260, RPL13AP3, KTN1, UGDH, UGDH-AS1 \n",
      "Negative:  FKRP, SLC1A5, STRN4, AP2S1, PRKD2, ARHGAP35, CALM3, PPP5D1, PPP5C, CCDC61 \n",
      "\t   MYPOP, IRF2BP1, IRS1, RHBDD1, DOCK10, COL4A4, AC073052.1, COL4A3, CUL3, SERPINE2 \n",
      "\t   AC097662.1, WDFY1, MFF, AP1S3, AGFG1, ACSL3, ARPC2, LINC01807, AC021016.2, AC016712.1 \n",
      "PC_ 5 \n",
      "Positive:  EMG1, SCARNA12, PHB2, PTPN6, C12orf57, U47924.2, ATN1, TAF7, PCDHB9, ENO2 \n",
      "\t   AC005618.1, VTRNA1-3, VTRNA1-1, PCDHGA6, PCDHGB6, ZMAT2, DSTNP2, PCDHGA10, HARS2, PCDHGC5 \n",
      "\t   LRRC23, HARS, DIAPH1, WDR55, RPL13P5, HDAC3, MIR3655, RELL2, SPSB2, IK \n",
      "Negative:  LYPLAL1, TGFB2, ZC3H11B, EPRS, RRP15, BPNT1, IARS2, GPATCH2, RPS15AP12, ESRRG \n",
      "\t   RAB3GAP2, USH2A, MORF4L1P1, KCTD3, AL451081.2, ABHD17AP3, MARK1, CENPF, C1orf115, TP53BP2 \n",
      "\t   ACTBP11, SMYD2, GTF2IP20, CAPN2, AL359742.1, DISP1, AC138393.1, MARC2, PROX1, FBXO28 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 15:18:43] define_signif_tumor_subclusters(), tumor: Malignant_MM5_Prolif\n",
      "INFO [2025-10-26 15:18:43] Setting auto leiden resolution for Malignant_MM5_Prolif to 0.0845483\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  TAPT1-AS1, TAPT1, RPS21P4, CD38, FAM200B, FBXL5, CPEB2, LINC00504, AC092546.1, AC006445.2 \n",
      "\t   BOD1L1, LINC01096, RNPS1P1, ZNF518B, RAB28, MIR3138, WDR1, SLC2A9, FAM86MP, ALG1L14P \n",
      "\t   FAM86KP, UNC93B8, TRMT44, GNPDA2, GABRG1, GUF1, RAC1P2, LINC02475, COX7B2, NDUFB4P12 \n",
      "Negative:  RNF34, ANAPC5, CAMKK2, P2RX4, SPPL3, ACADS, P2RX7, MIR4700, RPL12P33, UNC119B \n",
      "\t   OASL, MLEC, C12orf43, POP5, RNF10, AC063943.1, NRAV, DYNLL1, SRSF9, GATC \n",
      "\t   TRIAP1, AL021546.1, COX6A1, C6orf106, SIRT4, RNU4-1, RNU4-2, AL451165.2, RPLP0, PXN-AS1 \n",
      "PC_ 2 \n",
      "Positive:  PCSK7, RNF214, BACE1-AS, CEP164, TAGLN, IL10RA, SIDT2, MPZL3, PAFAH1B2, CD3E \n",
      "\t   AP000936.3, CD3D, SIK3, CD3G, UBE4A, ZPR1, ATP5L, BUD13, KMT2A, AP000797.1 \n",
      "\t   AP001267.3, AP000797.2, TMEM25, LINC00900, IFT46, CADM1, RPL5P30, ARCN1, NXPE4, DDX6 \n",
      "Negative:  LYST, B3GALNT2, TBCE, GGPS1, RBM34, ARID4B, TOMM20, IRF2BP2, TARBP1, COA6 \n",
      "\t   AL355472.1, AL355472.2, AL663058.1, KCNK1, MAP3K21, RPS7P3, PCNX2, NTPCR, MAP10, SIPA1L2 \n",
      "\t   DISC1, LINC00582, TSNAX, AL445524.1, EGLN1, SPRTN, EXOC8, GNPAT, C1orf131, AL109810.1 \n",
      "PC_ 3 \n",
      "Positive:  AC002558.3, MYL4, RPS7P11, AC068152.1, CDC27, GOSR2, EFCAB13, NSF, LRRC37A2, NPEPPS \n",
      "\t   ARL17B, AC025682.1, AC005829.1, KPNB1, KANSL1-AS1, TBX21, OSBPL7, KANSL1, SCRN2, AC018521.5 \n",
      "\t   LINC02210, SP2-AS1, RPS26P8, SP2, AC126544.1, PNPO, AC018521.1, LRRC37A4P, CDK5RAP3, COPZ2 \n",
      "Negative:  RNF168, UBXN7, TM4SF19-TCTEX1D2, TCTEX1D2, PCYT1A, TFRC, SDHAP1, TNK2, MUC20, LINC00969 \n",
      "\t   APOD, AC069213.3, PPP1R2, ABCF3, ALG3, AP2M1, EIF2B5, AC131235.1, DVL3, AC131235.3 \n",
      "\t   ABCC5, ECE2, PARL, PSMD2, CAMK2N2, EIF4G1, YEATS2, FAM131A, KLHL24, CLCN2 \n",
      "PC_ 4 \n",
      "Positive:  FGD3, ZNF484, EEF1DP2, BICD2, IPPK, CENPP, NOL8, SNORA84, IARS, SPTLC1 \n",
      "\t   AL391219.1, ROR2, NFIL3, CCDC66, AUH, ERC2, AL158071.3, WNT5A, SYK, RPS15P5 \n",
      "\t   SELENOK, DIRAS2, ACTR8, PA2G4P6, AC012467.2, SEMA4D, IL17RB, TMOD1, CHDH, VAPA \n",
      "Negative:  AC124312.1, SNURF, AC090602.1, RPL5P1, SNRPN, NDN, IMP3, SNX33, UBE2Q2, FBXO22 \n",
      "\t   ETFA, ISL2, SCAPER, RCN2, PSTPIP1, TSPAN3, ESCO2, MYRF, CCDC25, TMEM258 \n",
      "\t   MIR6843, CLU, MIR611, EPHX2, FEN1, PTK2B, ZNF815P, TRIM35, FADS2, CCZ1 \n",
      "PC_ 5 \n",
      "Positive:  PRKRA, DFNB59, AC009948.1, AC009948.2, FKBP7, PLEKHA3, OSBPL6, TTN-AS1, RBM45, TTN \n",
      "\t   TTC30A, SESTD1, TTC30B, FSIP2, ZNF804A, AGPS, ZNF385B, CACYBPP2, AC019080.1, NUP35 \n",
      "\t   CWC22, DUSP19, NCKAP1, AC079305.1, FRZB, SSFA2, SAP18P2, FTH1P20, UBE2E3, LINC01934 \n",
      "Negative:  FBXO46, AC007191.1, QPCTL, SNRPD2, SIX5, GIPR, AC074212.1, EML2-AS1, EML2, DM1-AS \n",
      "\t   OPA3, DMPK, VASP, DMWD, SYMPK, RTN2, IRF2BP1, FOSB, MYPOP, CCDC61 \n",
      "\t   PPP5C, PPP5D1, CALM3, ERCC1, PRKD2, STRN4, CD3EAP, ARHGAP35, AP2S1, FKRP \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 15:18:44] define_signif_tumor_subclusters(), tumor: Normal_Pb\n",
      "INFO [2025-10-26 15:18:44] Less cells in group Normal_Pb than k_nn setting. Keeping as a single subcluster.\n",
      "INFO [2025-10-26 15:18:44] define_signif_tumor_subclusters(), tumor: Normal_PC\n",
      "INFO [2025-10-26 15:18:44] Setting auto leiden resolution for Normal_PC to 0.0494373\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  BIRC5, PAQR4, TNFRSF12A, HCFC1R1, THOC6, MMP25-AS1, IL32, ZNF213-AS1, ZNF213, ZNF200 \n",
      "\t   ZNF263, TIGD7, ZNF75A, LINC01679, U2AF1, FAM189B, GBA, MSTO1, EPN1, U2AF2 \n",
      "\t   NLRP11, ZNF787, YY1AP1, PKNOX1, ZNF444, CCDC106, DAP3, ZSCAN5A, ZNF542P, MSTO2P \n",
      "Negative:  RPL19P11, AC008825.1, LINC01021, AC113347.1, MSNP1, LINC02228, TARS, AMACR, AC010342.1, C1QTNF3-AMACR \n",
      "\t   GUSBP1, C1QTNF3, LINC02241, CDH18, RPL36AP21, FTH1P10, BASP1, RPS26P28, MYO10, RETREG1 \n",
      "\t   ZNF622, EXOSC9, CCNA2, FBXL7, C8orf59, CA13, CA1, E2F5, LRRCC1, RALYL \n",
      "PC_ 2 \n",
      "Positive:  EEF1B2P6, AC008264.1, NDUFB9P2, CHCHD3, PODXL, MKLN1-AS, AC008038.1, AC009362.1, MIR3654, MKLN1 \n",
      "\t   AC083873.1, LINC-PINT, EXOC4, AC016831.1, COPG2, SLC35B4, MEST, AKR1B1, BPGM, CEP41 \n",
      "\t   CALD1, TMEM209, ZNF18, ADPRM, MAP2K4, TMEM220, ARHGAP44, COX10-AS1, ELAC2, SCO1 \n",
      "Negative:  NAA50, USF3, RN7SL767P, AC108693.2, SIDT1, SPICE1, CFAP44, ATP6V1A, BOC, NEPRO \n",
      "\t   GTPBP8, GRAMD1C, CD200R1, AC079944.2, AC048334.1, CCDC80, SLC35A5, ATG3, ZDHHC23, BTLA \n",
      "\t   CCDC191, CD200, GCSAM, QTRT2, C3orf52, ABHD10, TIGIT, PHLDB2, GGCT, PLCXD2 \n",
      "PC_ 3 \n",
      "Positive:  C7orf55-LUC7L2, UBN2, LUC7L2, FMC1, CLEC2L, AC009220.1, HIPK2, AC009220.3, TTC26, ZC3HAV1 \n",
      "\t   KIAA1549, TBXAS1, AC020983.1, PARP12, TRIM24, KDM7A, AC008155.1, JHDM1D-AS1, AC024082.1, SLC37A3 \n",
      "\t   AKR1D1, CREB3L2, MKRN1, AC009245.1, ADCK2, AC090498.1, DGKI, PTN, NDUFB2, MTPN \n",
      "Negative:  AL590714.1, ACA64, UFC1, DEDD, USP21, NIT1, PFDN2, PPOX, KLHDC9, B4GALT3 \n",
      "\t   NECTIN4, ARHGAP30, ADAMTS4, USF1, AL591806.3, NDUFS2, F11R, AL354714.2, FCER1G, LY9 \n",
      "\t   SLAMF7, CD48, TOMM40L, AL121985.1, SLAMF1, CD84, SDHC, SLAMF6, AL592295.1, NCSTN \n",
      "PC_ 4 \n",
      "Positive:  NUDCD1, PKHD1L1, ENY2, AC104248.1, EBAG9, TMEM74, EMC2, RNU4-37P, EIF3E, TAGLN2P1 \n",
      "\t   OXR1, CSMD3, AC027031.2, ZFPM2-AS1, TRPS1, LRP12, AC090686.1, EIF3H, UTP23, RAD21 \n",
      "\t   ZNF706, AP003469.4, AP003469.2, YWHAZ, SCD5, AC067942.1, PABPC1, ENOPH1, HNRNPDL, GAPDHP62 \n",
      "Negative:  PTENP1, UBE2R2, ANXA2P2, UBAP2, NOL6, DCAF12, AQP3, UBAP1, NFX1, CHMP5 \n",
      "\t   AL353662.2, BAG1, NUDT2, C9orf24, B4GALT1-AS1, FAM219A, B4GALT1, RPP25L, SMU1, DCTN3 \n",
      "\t   ARID3C, APTX, SIGMAR1, SMIM27, GALT, NDUFB6, IL11RA, AL162231.1, TOPORS, AL162231.2 \n",
      "PC_ 5 \n",
      "Positive:  C7orf71, SKAP2, AC004540.1, SNX10, TPM3P4, AC010677.1, HOXA1, CBX3, HOTAIRM1, HNRNPA2B1 \n",
      "\t   HOTAIRM1-4, AC004520.1, HOTAIRM1-5, NFE2L3, HOXA5, AC010719.1, HOXA7, AC005165.2, HOXA9, C7orf31 \n",
      "\t   HOXA10, EIF4HP1, CYCS, HIBADH, OSBPL3, TAX1BP1, MPP6, JAZF1, PCMTD1P3, CREB5 \n",
      "Negative:  TIGIT, ZBTB20, QTRT2, CCDC191, MIR568, ZDHHC23, AC079944.2, ZBTB20-AS4, LSAMP, GRAMD1C \n",
      "\t   ATP6V1A, BZW1P2, AC108693.2, NAA50, PTMAP8, USF3, RN7SL767P, AC068633.1, SIDT1, IGSF11 \n",
      "\t   SPICE1, CFAP44, B4GALT4, BOC, ARHGAP31, NEPRO, ARHGAP31-AS1, GTPBP8, TMEM39A, CD200R1 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 15:18:46] -mirroring for hspike\n",
      "INFO [2025-10-26 15:18:46] define_signif_tumor_subclusters(p_val=0.1\n",
      "INFO [2025-10-26 15:18:46] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_Normal_PC\n",
      "INFO [2025-10-26 15:18:46] cut tree into: 1 groups\n",
      "INFO [2025-10-26 15:18:46] -processing spike_tumor_cell_Normal_PC,spike_tumor_cell_Normal_PC_s1\n",
      "INFO [2025-10-26 15:18:46] define_signif_tumor_subclusters(), tumor: simnorm_cell_Normal_PC\n",
      "INFO [2025-10-26 15:18:46] cut tree into: 1 groups\n",
      "INFO [2025-10-26 15:18:46] -processing simnorm_cell_Normal_PC,simnorm_cell_Normal_PC_s1\n",
      "INFO [2025-10-26 15:19:09] ::plot_cnv:Start\n",
      "INFO [2025-10-26 15:19:09] ::plot_cnv:Current data dimensions (r,c)=14961,1875 Total=28191805.7872296 Min=0.548464152420786 Max=1.99385371368565.\n",
      "INFO [2025-10-26 15:19:10] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2025-10-26 15:19:10] plot_cnv(): auto thresholding at: (0.692221 , 1.317755)\n",
      "INFO [2025-10-26 15:19:11] plot_cnv_observation:Start\n",
      "INFO [2025-10-26 15:19:11] Observation data size: Cells= 1477 Genes= 14961\n",
      "INFO [2025-10-26 15:19:11] clustering observations via method: ward.D\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(1) is 61\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  61,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(2) is 10\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  10,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(3) is 2\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  2,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(4) is 44\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  44,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(5) is 32\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  32,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(6) is 29\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  29,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(7) is 24\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  24,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(8) is 20\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  20,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(9) is 19\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  19,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(10) is 18\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  18,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(11) is 17\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  17,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(12) is 12\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  12,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(13) is 21\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  21,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(14) is 21\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  21,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(15) is 16\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  16,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(16) is 46\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  46,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(17) is 25\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  25,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(18) is 21\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  21,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(19) is 18\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  18,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(20) is 17\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  17,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(21) is 14\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  14,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(22) is 1\n",
      "INFO [2025-10-26 15:19:11] Skipping group: 22, since less than 2 entries\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(23) is 29\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  29,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(24) is 25\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  25,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(25) is 17\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  17,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(26) is 37\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  37,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(27) is 32\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  32,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(28) is 27\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  27,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(29) is 12\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  12,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(30) is 37\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  37,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(31) is 24\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  24,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(32) is 13\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  13,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(33) is 33\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  33,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(34) is 32\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  32,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(35) is 30\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  30,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(36) is 26\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  26,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(37) is 22\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  22,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(38) is 20\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  20,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(39) is 19\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  19,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(40) is 17\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  17,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(41) is 17\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  17,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(42) is 17\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  17,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(43) is 15\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  15,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(44) is 29\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  29,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(45) is 26\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  26,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(46) is 25\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  25,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(47) is 24\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  24,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(48) is 21\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  21,14961\n",
      "INFO [2025-10-26 15:19:11] Number of cells in group(49) is 37\n",
      "INFO [2025-10-26 15:19:11] group size being clustered:  37,14961\n",
      "INFO [2025-10-26 15:19:12] Number of cells in group(50) is 25\n",
      "INFO [2025-10-26 15:19:12] group size being clustered:  25,14961\n",
      "INFO [2025-10-26 15:19:12] Number of cells in group(51) is 20\n",
      "INFO [2025-10-26 15:19:12] group size being clustered:  20,14961\n",
      "INFO [2025-10-26 15:19:12] Number of cells in group(52) is 20\n",
      "INFO [2025-10-26 15:19:12] group size being clustered:  20,14961\n",
      "INFO [2025-10-26 15:19:12] Number of cells in group(53) is 17\n",
      "INFO [2025-10-26 15:19:12] group size being clustered:  17,14961\n",
      "INFO [2025-10-26 15:19:12] Number of cells in group(54) is 14\n",
      "INFO [2025-10-26 15:19:12] group size being clustered:  14,14961\n",
      "INFO [2025-10-26 15:19:12] Number of cells in group(55) is 42\n",
      "INFO [2025-10-26 15:19:12] group size being clustered:  42,14961\n",
      "INFO [2025-10-26 15:19:12] Number of cells in group(56) is 33\n",
      "INFO [2025-10-26 15:19:12] group size being clustered:  33,14961\n",
      "INFO [2025-10-26 15:19:12] Number of cells in group(57) is 32\n",
      "INFO [2025-10-26 15:19:12] group size being clustered:  32,14961\n",
      "INFO [2025-10-26 15:19:12] Number of cells in group(58) is 27\n",
      "INFO [2025-10-26 15:19:12] group size being clustered:  27,14961\n",
      "INFO [2025-10-26 15:19:12] Number of cells in group(59) is 26\n",
      "INFO [2025-10-26 15:19:12] group size being clustered:  26,14961\n",
      "INFO [2025-10-26 15:19:12] Number of cells in group(60) is 20\n",
      "INFO [2025-10-26 15:19:12] group size being clustered:  20,14961\n",
      "INFO [2025-10-26 15:19:12] Number of cells in group(61) is 20\n",
      "INFO [2025-10-26 15:19:12] group size being clustered:  20,14961\n",
      "INFO [2025-10-26 15:19:12] Number of cells in group(62) is 13\n",
      "INFO [2025-10-26 15:19:12] group size being clustered:  13,14961\n",
      "INFO [2025-10-26 15:19:12] Number of cells in group(63) is 17\n",
      "INFO [2025-10-26 15:19:12] group size being clustered:  17,14961\n",
      "INFO [2025-10-26 15:19:12] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2025-10-26 15:19:12] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2025-10-26 15:19:12] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 15:19:12] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 15:19:15] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 15:19:15] Quantiles of plotted data range: 0.692221162989231,0.937577466951409,0.9995148640224,1.06709206710094,1.31775540985872\n",
      "INFO [2025-10-26 15:19:16] plot_cnv_references:Start\n",
      "INFO [2025-10-26 15:19:16] Reference data size: Cells= 398 Genes= 14961\n",
      "INFO [2025-10-26 15:19:17] plot_cnv_references:Number reference groups= 13\n",
      "INFO [2025-10-26 15:19:17] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2025-10-26 15:19:18] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 15:19:18] Quantiles of plotted data range: 0.692221162989231,0.957156923813384,0.999523544889441,1.04448758030643,1.31775540985872\n",
      "INFO [2025-10-26 15:19:42] ::plot_cnv:Start\n",
      "INFO [2025-10-26 15:19:42] ::plot_cnv:Current data dimensions (r,c)=14961,1875 Total=28191805.7872296 Min=0.548464152420786 Max=1.99385371368565.\n",
      "INFO [2025-10-26 15:19:43] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2025-10-26 15:19:43] plot_cnv(): auto thresholding at: (0.692221 , 1.317755)\n",
      "INFO [2025-10-26 15:19:44] plot_cnv_observation:Start\n",
      "INFO [2025-10-26 15:19:44] Observation data size: Cells= 1477 Genes= 14961\n",
      "INFO [2025-10-26 15:19:44] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2025-10-26 15:19:44] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2025-10-26 15:19:44] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 15:19:44] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 15:19:47] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 15:19:47] Quantiles of plotted data range: 0.692221162989231,0.937577466951409,0.9995148640224,1.06709206710094,1.31775540985872\n",
      "INFO [2025-10-26 15:19:48] plot_cnv_references:Start\n",
      "INFO [2025-10-26 15:19:48] Reference data size: Cells= 398 Genes= 14961\n",
      "INFO [2025-10-26 15:19:50] plot_cnv_references:Number reference groups= 1\n",
      "INFO [2025-10-26 15:19:50] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2025-10-26 15:19:50] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 15:19:50] Quantiles of plotted data range: 0.692221162989231,0.957156923813384,0.999523544889441,1.04448758030643,1.31775540985872\n",
      "INFO [2025-10-26 15:19:52] \n",
      "\n",
      "\tSTEP 17: HMM-based CNV prediction\n",
      "\n",
      "INFO [2025-10-26 15:19:52] predict_CNV_via_HMM_on_tumor_subclusters\n",
      "INFO [2025-10-26 15:20:29] -done predicting CNV based on initial tumor subclusters\n",
      "INFO [2025-10-26 15:20:29] get_predicted_CNV_regions(subcluster)\n",
      "INFO [2025-10-26 15:20:29] -processing cell_group_name: Interm.Interm_s2, size: 61\n",
      "INFO [2025-10-26 15:20:37] -processing cell_group_name: Interm.Interm_s1, size: 10\n",
      "INFO [2025-10-26 15:20:45] -processing cell_group_name: Interm.Interm_s3, size: 2\n",
      "INFO [2025-10-26 15:20:53] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s1, size: 44\n",
      "INFO [2025-10-26 15:21:02] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s6, size: 32\n",
      "INFO [2025-10-26 15:21:10] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s3, size: 29\n",
      "INFO [2025-10-26 15:21:18] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s7, size: 24\n",
      "INFO [2025-10-26 15:21:26] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s4, size: 20\n",
      "INFO [2025-10-26 15:21:35] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s2, size: 19\n",
      "INFO [2025-10-26 15:21:43] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s9, size: 18\n",
      "INFO [2025-10-26 15:21:51] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s5, size: 17\n",
      "INFO [2025-10-26 15:21:59] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s8, size: 12\n",
      "INFO [2025-10-26 15:22:08] -processing cell_group_name: Malignant_MM1_Prolif.Malignant_MM1_Prolif_s2, size: 21\n",
      "INFO [2025-10-26 15:22:15] -processing cell_group_name: Malignant_MM1_Prolif.Malignant_MM1_Prolif_s3, size: 21\n",
      "INFO [2025-10-26 15:22:23] -processing cell_group_name: Malignant_MM1_Prolif.Malignant_MM1_Prolif_s1, size: 16\n",
      "INFO [2025-10-26 15:22:31] -processing cell_group_name: Malignant_MM2_Non-Prolif.Malignant_MM2_Non-Prolif_s1, size: 46\n",
      "INFO [2025-10-26 15:22:40] -processing cell_group_name: Malignant_MM2_Non-Prolif.Malignant_MM2_Non-Prolif_s3, size: 25\n",
      "INFO [2025-10-26 15:22:48] -processing cell_group_name: Malignant_MM2_Non-Prolif.Malignant_MM2_Non-Prolif_s6, size: 21\n",
      "INFO [2025-10-26 15:22:56] -processing cell_group_name: Malignant_MM2_Non-Prolif.Malignant_MM2_Non-Prolif_s5, size: 18\n",
      "INFO [2025-10-26 15:23:05] -processing cell_group_name: Malignant_MM2_Non-Prolif.Malignant_MM2_Non-Prolif_s2, size: 17\n",
      "INFO [2025-10-26 15:23:13] -processing cell_group_name: Malignant_MM2_Non-Prolif.Malignant_MM2_Non-Prolif_s4, size: 14\n",
      "INFO [2025-10-26 15:23:21] -processing cell_group_name: Malignant_MM2_Non-Prolif.Malignant_MM2_Non-Prolif_s7, size: 1\n",
      "INFO [2025-10-26 15:23:29] -processing cell_group_name: Malignant_MM2_Prolif.Malignant_MM2_Prolif_s1, size: 29\n",
      "INFO [2025-10-26 15:23:37] -processing cell_group_name: Malignant_MM2_Prolif.Malignant_MM2_Prolif_s2, size: 25\n",
      "INFO [2025-10-26 15:23:45] -processing cell_group_name: Malignant_MM2_Prolif.Malignant_MM2_Prolif_s3, size: 17\n",
      "INFO [2025-10-26 15:23:53] -processing cell_group_name: Malignant_MM3_Non-Prolif.Malignant_MM3_Non-Prolif_s2, size: 37\n",
      "INFO [2025-10-26 15:24:02] -processing cell_group_name: Malignant_MM3_Non-Prolif.Malignant_MM3_Non-Prolif_s4, size: 32\n",
      "INFO [2025-10-26 15:24:10] -processing cell_group_name: Malignant_MM3_Non-Prolif.Malignant_MM3_Non-Prolif_s1, size: 27\n",
      "INFO [2025-10-26 15:24:19] -processing cell_group_name: Malignant_MM3_Non-Prolif.Malignant_MM3_Non-Prolif_s3, size: 12\n",
      "INFO [2025-10-26 15:24:27] -processing cell_group_name: Malignant_MM3_Prolif.Malignant_MM3_Prolif_s3, size: 37\n",
      "INFO [2025-10-26 15:24:35] -processing cell_group_name: Malignant_MM3_Prolif.Malignant_MM3_Prolif_s2, size: 24\n",
      "INFO [2025-10-26 15:24:43] -processing cell_group_name: Malignant_MM3_Prolif.Malignant_MM3_Prolif_s1, size: 13\n",
      "INFO [2025-10-26 15:24:51] -processing cell_group_name: Malignant_MM4_Non-Prolif.Malignant_MM4_Non-Prolif_s2, size: 33\n",
      "INFO [2025-10-26 15:24:59] -processing cell_group_name: Malignant_MM4_Non-Prolif.Malignant_MM4_Non-Prolif_s7, size: 32\n",
      "INFO [2025-10-26 15:25:07] -processing cell_group_name: Malignant_MM4_Non-Prolif.Malignant_MM4_Non-Prolif_s4, size: 30\n",
      "INFO [2025-10-26 15:25:16] -processing cell_group_name: Malignant_MM4_Non-Prolif.Malignant_MM4_Non-Prolif_s10, size: 26\n",
      "INFO [2025-10-26 15:25:24] -processing cell_group_name: Malignant_MM4_Non-Prolif.Malignant_MM4_Non-Prolif_s8, size: 22\n",
      "INFO [2025-10-26 15:25:32] -processing cell_group_name: Malignant_MM4_Non-Prolif.Malignant_MM4_Non-Prolif_s11, size: 20\n",
      "INFO [2025-10-26 15:25:40] -processing cell_group_name: Malignant_MM4_Non-Prolif.Malignant_MM4_Non-Prolif_s6, size: 19\n",
      "INFO [2025-10-26 15:25:48] -processing cell_group_name: Malignant_MM4_Non-Prolif.Malignant_MM4_Non-Prolif_s1, size: 17\n",
      "INFO [2025-10-26 15:25:56] -processing cell_group_name: Malignant_MM4_Non-Prolif.Malignant_MM4_Non-Prolif_s3, size: 17\n",
      "INFO [2025-10-26 15:26:05] -processing cell_group_name: Malignant_MM4_Non-Prolif.Malignant_MM4_Non-Prolif_s5, size: 17\n",
      "INFO [2025-10-26 15:26:13] -processing cell_group_name: Malignant_MM4_Non-Prolif.Malignant_MM4_Non-Prolif_s9, size: 15\n",
      "INFO [2025-10-26 15:26:21] -processing cell_group_name: Malignant_MM4_Prolif.Malignant_MM4_Prolif_s5, size: 29\n",
      "INFO [2025-10-26 15:26:29] -processing cell_group_name: Malignant_MM4_Prolif.Malignant_MM4_Prolif_s1, size: 26\n",
      "INFO [2025-10-26 15:26:37] -processing cell_group_name: Malignant_MM4_Prolif.Malignant_MM4_Prolif_s2, size: 25\n",
      "INFO [2025-10-26 15:26:45] -processing cell_group_name: Malignant_MM4_Prolif.Malignant_MM4_Prolif_s4, size: 24\n",
      "INFO [2025-10-26 15:26:53] -processing cell_group_name: Malignant_MM4_Prolif.Malignant_MM4_Prolif_s3, size: 21\n",
      "INFO [2025-10-26 15:27:02] -processing cell_group_name: Malignant_MM5_Non-Prolif.Malignant_MM5_Non-Prolif_s1, size: 37\n",
      "INFO [2025-10-26 15:27:10] -processing cell_group_name: Malignant_MM5_Non-Prolif.Malignant_MM5_Non-Prolif_s4, size: 25\n",
      "INFO [2025-10-26 15:27:18] -processing cell_group_name: Malignant_MM5_Non-Prolif.Malignant_MM5_Non-Prolif_s2, size: 20\n",
      "INFO [2025-10-26 15:27:26] -processing cell_group_name: Malignant_MM5_Non-Prolif.Malignant_MM5_Non-Prolif_s3, size: 20\n",
      "INFO [2025-10-26 15:27:35] -processing cell_group_name: Malignant_MM5_Non-Prolif.Malignant_MM5_Non-Prolif_s6, size: 17\n",
      "INFO [2025-10-26 15:27:43] -processing cell_group_name: Malignant_MM5_Non-Prolif.Malignant_MM5_Non-Prolif_s5, size: 14\n",
      "INFO [2025-10-26 15:27:51] -processing cell_group_name: Malignant_MM5_Prolif.Malignant_MM5_Prolif_s3, size: 42\n",
      "INFO [2025-10-26 15:27:59] -processing cell_group_name: Malignant_MM5_Prolif.Malignant_MM5_Prolif_s5, size: 33\n",
      "INFO [2025-10-26 15:28:07] -processing cell_group_name: Malignant_MM5_Prolif.Malignant_MM5_Prolif_s1, size: 32\n",
      "INFO [2025-10-26 15:28:15] -processing cell_group_name: Malignant_MM5_Prolif.Malignant_MM5_Prolif_s4, size: 27\n",
      "INFO [2025-10-26 15:28:23] -processing cell_group_name: Malignant_MM5_Prolif.Malignant_MM5_Prolif_s2, size: 26\n",
      "INFO [2025-10-26 15:28:32] -processing cell_group_name: Malignant_MM5_Prolif.Malignant_MM5_Prolif_s6, size: 20\n",
      "INFO [2025-10-26 15:28:40] -processing cell_group_name: Malignant_MM5_Prolif.Malignant_MM5_Prolif_s8, size: 20\n",
      "INFO [2025-10-26 15:28:48] -processing cell_group_name: Malignant_MM5_Prolif.Malignant_MM5_Prolif_s7, size: 13\n",
      "INFO [2025-10-26 15:28:56] -processing cell_group_name: Normal_Pb.Normal_Pb, size: 17\n",
      "INFO [2025-10-26 15:29:04] -processing cell_group_name: Normal_PC.Normal_PC_s7, size: 113\n",
      "INFO [2025-10-26 15:29:13] -processing cell_group_name: Normal_PC.Normal_PC_s2, size: 68\n",
      "INFO [2025-10-26 15:29:21] -processing cell_group_name: Normal_PC.Normal_PC_s5, size: 61\n",
      "INFO [2025-10-26 15:29:30] -processing cell_group_name: Normal_PC.Normal_PC_s8, size: 35\n",
      "INFO [2025-10-26 15:29:38] -processing cell_group_name: Normal_PC.Normal_PC_s6, size: 27\n",
      "INFO [2025-10-26 15:29:47] -processing cell_group_name: Normal_PC.Normal_PC_s3, size: 24\n",
      "INFO [2025-10-26 15:29:55] -processing cell_group_name: Normal_PC.Normal_PC_s9, size: 23\n",
      "INFO [2025-10-26 15:30:04] -processing cell_group_name: Normal_PC.Normal_PC_s4, size: 22\n",
      "INFO [2025-10-26 15:30:12] -processing cell_group_name: Normal_PC.Normal_PC_s1, size: 20\n",
      "INFO [2025-10-26 15:30:20] -processing cell_group_name: Normal_PC.Normal_PC_s12, size: 2\n",
      "INFO [2025-10-26 15:30:28] -processing cell_group_name: Normal_PC.Normal_PC_s10, size: 1\n",
      "INFO [2025-10-26 15:30:36] -processing cell_group_name: Normal_PC.Normal_PC_s11, size: 1\n",
      "INFO [2025-10-26 15:30:44] -processing cell_group_name: Normal_PC.Normal_PC_s13, size: 1\n",
      "INFO [2025-10-26 15:30:53] -writing cell clusters file: /home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_output/output_infercnv_r_arch_prolif_MARS/17_HMM_predHMMi6.leiden.hmm_mode-subclusters.cell_groupings\n",
      "INFO [2025-10-26 15:30:53] -writing cnv regions file: /home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_output/output_infercnv_r_arch_prolif_MARS/17_HMM_predHMMi6.leiden.hmm_mode-subclusters.pred_cnv_regions.dat\n",
      "INFO [2025-10-26 15:30:54] -writing per-gene cnv report: /home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_output/output_infercnv_r_arch_prolif_MARS/17_HMM_predHMMi6.leiden.hmm_mode-subclusters.pred_cnv_genes.dat\n",
      "INFO [2025-10-26 15:30:55] -writing gene ordering info: /home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_output/output_infercnv_r_arch_prolif_MARS/17_HMM_predHMMi6.leiden.hmm_mode-subclusters.genes_used.dat\n",
      "INFO [2025-10-26 15:31:07] ::plot_cnv:Start\n",
      "INFO [2025-10-26 15:31:07] ::plot_cnv:Current data dimensions (r,c)=14961,1875 Total=84978990 Min=2 Max=6.\n",
      "INFO [2025-10-26 15:31:08] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2025-10-26 15:31:08] plot_cnv_observation:Start\n",
      "INFO [2025-10-26 15:31:08] Observation data size: Cells= 1477 Genes= 14961\n",
      "INFO [2025-10-26 15:31:08] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2025-10-26 15:31:08] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2025-10-26 15:31:08] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 15:31:08] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 15:31:11] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 15:31:11] Quantiles of plotted data range: 2,3,3,3,6\n",
      "INFO [2025-10-26 15:31:13] plot_cnv_references:Start\n",
      "INFO [2025-10-26 15:31:13] Reference data size: Cells= 398 Genes= 14961\n",
      "INFO [2025-10-26 15:31:14] plot_cnv_references:Number reference groups= 1\n",
      "INFO [2025-10-26 15:31:14] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2025-10-26 15:31:14] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 15:31:14] Quantiles of plotted data range: 2,3,3,3,4\n",
      "INFO [2025-10-26 15:31:16] \n",
      "\n",
      "\tSTEP 18: Run Bayesian Network Model on HMM predicted CNVs\n",
      "\n",
      "INFO [2025-10-26 15:31:16] Creating the following Directory:  /home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_output/output_infercnv_r_arch_prolif_MARS/BayesNetOutput.HMMi6.leiden.hmm_mode-subclusters\n",
      "INFO [2025-10-26 15:31:16] Initializing new MCM InferCNV Object.\n",
      "INFO [2025-10-26 15:31:16] validating infercnv_obj\n",
      "INFO [2025-10-26 15:31:16] Total CNV's:  1345\n",
      "INFO [2025-10-26 15:31:16] Loading BUGS Model.\n",
      "INFO [2025-10-26 15:31:18] Running Sampling Using Parallel with  4 Cores\n",
      "INFO [2025-10-26 15:51:15] Obtaining probabilities post-sampling\n",
      "INFO [2025-10-26 15:52:52] Gibbs sampling time:  21.5720984061559  Minutes\n",
      "INFO [2025-10-26 15:54:14] ::plot_cnv:Start\n",
      "INFO [2025-10-26 15:54:14] ::plot_cnv:Current data dimensions (r,c)=14961,1875 Total=2965850.14652729 Min=0 Max=0.97694182336097.\n",
      "INFO [2025-10-26 15:54:15] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2025-10-26 15:54:15] plot_cnv_observation:Start\n",
      "INFO [2025-10-26 15:54:15] Observation data size: Cells= 1477 Genes= 14961\n",
      "INFO [2025-10-26 15:54:16] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2025-10-26 15:54:16] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2025-10-26 15:54:16] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 15:54:16] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 15:54:18] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 15:54:18] Quantiles of plotted data range: 0,0,0,0,0.97694182336097\n",
      "INFO [2025-10-26 15:54:20] plot_cnv_references:Start\n",
      "INFO [2025-10-26 15:54:20] Reference data size: Cells= 398 Genes= 14961\n",
      "INFO [2025-10-26 15:54:21] plot_cnv_references:Number reference groups= 1\n",
      "INFO [2025-10-26 15:54:21] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2025-10-26 15:54:22] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 15:54:22] Quantiles of plotted data range: 0,0,0,0,0.859889320539194\n",
      "INFO [2025-10-26 15:55:01] \n",
      "\n",
      "\tSTEP 19: Filter HMM predicted CNVs based on the Bayesian Network Model results and BayesMaxPNormal\n",
      "\n",
      "INFO [2025-10-26 15:55:02] Attempting to removing CNV(s) with a probability of being normal above  0.5\n",
      "INFO [2025-10-26 15:55:02] Removing  15  CNV(s) identified by the HMM.\n",
      "INFO [2025-10-26 15:55:02] Total CNV's after removing:  1330\n",
      "INFO [2025-10-26 15:55:02] Reassigning CNVs based on state probabilities.\n",
      "INFO [2025-10-26 15:55:02] Changing the following CNV's states assigned by the HMM to the following based on the CNV's state probabilities.\n",
      " chr19-region_55 : 2  (P= 0.249718218525346 ) ->  3 (P= 0.251672468007015 )\n",
      "chr10-region_58 : 4  (P= 0.24866630248206 ) ->  3 (P= 0.251656417279232 )\n",
      "chr5-region_63 : 4  (P= 0.24866630248206 ) ->  3 (P= 0.251656417279232 )\n",
      "chr11-region_72 : 4  (P= 0.250329898755975 ) ->  3 (P= 0.250762784902768 )\n",
      "chr17-region_75 : 4  (P= 0.250329898755975 ) ->  3 (P= 0.250762784902768 )\n",
      "chr10-region_136 : 4  (P= 0.394614196177415 ) ->  3 (P= 0.446477693622716 )\n",
      "chr22-region_334 : 2  (P= 0.374164183831052 ) ->  3 (P= 0.458828879383467 )\n",
      "chr14-region_443 : 2  (P= 0.33763324077974 ) ->  3 (P= 0.43992885818495 )\n",
      "chr10-region_447 : 2  (P= 0.386869815870867 ) ->  3 (P= 0.39095031853766 )\n",
      "chr14-region_931 : 2  (P= 0.429047487102184 ) ->  3 (P= 0.456101782967993 )\n",
      "chr14-region_933 : 2  (P= 0.400671732186097 ) ->  3 (P= 0.48429123055344 )\n",
      "chr19-region_976 : 2  (P= 0.378251664421661 ) ->  3 (P= 0.491442466552723 )\n",
      "chr22-region_1081 : 2  (P= 0.442715022917476 ) ->  3 (P= 0.464584661309319 )\n",
      "chr1-region_1176 : 5  (P= 0.0871137531818597 ) ->  4 (P= 0.413790924494521 )\n",
      "chr9-region_1193 : 2  (P= 0.335250923870136 ) ->  3 (P= 0.442972550797417 )\n",
      "chr20-region_1209 : 2  (P= 0.388361864012526 ) ->  3 (P= 0.388861307700543 )\n",
      "chr6-region_1219 : 2  (P= 0.332995334107316 ) ->  3 (P= 0.444065966937384 )\n",
      "chr1-region_1225 : 4  (P= 0.117044518575989 ) ->  3 (P= 0.393509668412474 )\n",
      "chr10-region_1229 : 2  (P= 0.443357808564101 ) ->  3 (P= 0.464778644578178 )\n",
      "chr19-region_1292 : 4  (P= 0.268383995363481 ) ->  3 (P= 0.333222727627819 )\n",
      "chr19-region_1352 : 4  (P= 0.312850419019473 ) ->  3 (P= 0.422940724786202 )\n",
      "chr5-region_1373 : 2  (P= 0.366629238265106 ) ->  3 (P= 0.42218285156421 )\n",
      "chr17-region_1395 : 4  (P= 0.367178690143868 ) ->  3 (P= 0.370059084263197 )\n",
      "chr10-region_2106 : 2  (P= 0.405441039325805 ) ->  3 (P= 0.468899305579623 )\n",
      "chr8-region_2194 : 4  (P= 0.257937449124163 ) ->  3 (P= 0.386372133919971 )\n",
      "chr6-region_2199 : 2  (P= 0.386939238926733 ) ->  3 (P= 0.484380267874418 )\n",
      "chr18-region_2213 : 4  (P= 0.255687706869103 ) ->  3 (P= 0.418032778718205 )\n",
      "chr19-region_2225 : 4  (P= 0.411528075173391 ) ->  3 (P= 0.455042461521162 )\n",
      "chr3-region_2233 : 4  (P= 0.3344057723096 ) ->  3 (P= 0.399116109692389 )\n",
      "chr6-region_2320 : 2  (P= 0.40851147060932 ) ->  3 (P= 0.442399657908666 )\n",
      "chr1-region_2329 : 4  (P= 0.371962190365203 ) ->  5 (P= 0.372760389025277 )\n",
      "chr12-region_2485 : 4  (P= 0.269772382741385 ) ->  3 (P= 0.461148686380144 )\n",
      "chr4-region_2668 : 2  (P= 0.433455306813746 ) ->  3 (P= 0.482632627123193 )\n",
      "chr8-region_2689 : 4  (P= 0.33341889689926 ) ->  3 (P= 0.458170772251569 )\n",
      "chr1-region_2781 : 4  (P= 0.315921484151745 ) ->  3 (P= 0.446569203953459 )\n",
      "chr14-region_2783 : 2  (P= 0.447124179042292 ) ->  3 (P= 0.448584808926408 )\n",
      "chr9-region_2801 : 4  (P= 0.394661589367508 ) ->  3 (P= 0.420317338837688 )\n",
      "chr15-region_2804 : 4  (P= 0.395704431433831 ) ->  3 (P= 0.473530897010131 )\n",
      "chr11-region_2808 : 2  (P= 0.422807182224687 ) ->  3 (P= 0.471730127395564 )\n",
      "chr17-region_2812 : 4  (P= 0.314203686225655 ) ->  3 (P= 0.47333657882577 )\n",
      "chr7-region_2827 : 4  (P= 0.30709294727306 ) ->  3 (P= 0.455184366485836 )\n",
      "chr19-region_3160 : 4  (P= 0.366795588703536 ) ->  5 (P= 0.423081821189991 )\n",
      "chr3-region_3459 : 4  (P= 0.245609219581847 ) ->  3 (P= 0.252312211801956 )\n",
      "chr2-region_3466 : 4  (P= 0.245609219581847 ) ->  3 (P= 0.252312211801956 )\n",
      "chr11-region_3469 : 4  (P= 0.248286982955074 ) ->  3 (P= 0.252057252208003 )\n",
      "INFO [2025-10-26 15:55:02] Creating Plots for CNV and cell Probabilities.\n",
      "INFO [2025-10-26 16:01:33] ::plot_cnv:Start\n",
      "INFO [2025-10-26 16:01:33] ::plot_cnv:Current data dimensions (r,c)=14961,1875 Total=2945469.04194569 Min=0 Max=0.97694182336097.\n",
      "INFO [2025-10-26 16:01:33] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2025-10-26 16:01:34] plot_cnv_observation:Start\n",
      "INFO [2025-10-26 16:01:34] Observation data size: Cells= 1477 Genes= 14961\n",
      "INFO [2025-10-26 16:01:34] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2025-10-26 16:01:34] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2025-10-26 16:01:34] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 16:01:34] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 16:01:36] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 16:01:36] Quantiles of plotted data range: 0,0,0,0,0.97694182336097\n",
      "INFO [2025-10-26 16:01:38] plot_cnv_references:Start\n",
      "INFO [2025-10-26 16:01:38] Reference data size: Cells= 398 Genes= 14961\n",
      "INFO [2025-10-26 16:01:39] plot_cnv_references:Number reference groups= 1\n",
      "INFO [2025-10-26 16:01:39] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2025-10-26 16:01:40] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 16:01:40] Quantiles of plotted data range: 0,0,0,0,0.859889320539194\n",
      "INFO [2025-10-26 16:01:53] ::plot_cnv:Start\n",
      "INFO [2025-10-26 16:01:53] ::plot_cnv:Current data dimensions (r,c)=14961,1875 Total=85023677 Min=2 Max=6.\n",
      "INFO [2025-10-26 16:01:54] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2025-10-26 16:01:54] plot_cnv_observation:Start\n",
      "INFO [2025-10-26 16:01:54] Observation data size: Cells= 1477 Genes= 14961\n",
      "INFO [2025-10-26 16:01:54] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2025-10-26 16:01:54] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2025-10-26 16:01:54] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 16:01:54] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 16:01:57] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 16:01:57] Quantiles of plotted data range: 2,3,3,3,6\n",
      "INFO [2025-10-26 16:01:58] plot_cnv_references:Start\n",
      "INFO [2025-10-26 16:01:58] Reference data size: Cells= 398 Genes= 14961\n",
      "INFO [2025-10-26 16:02:00] plot_cnv_references:Number reference groups= 1\n",
      "INFO [2025-10-26 16:02:00] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2025-10-26 16:02:00] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 16:02:00] Quantiles of plotted data range: 2,3,3,3,4\n",
      "INFO [2025-10-26 16:02:06] \n",
      "\n",
      "\tSTEP 20: Converting HMM-based CNV states to repr expr vals\n",
      "\n",
      "INFO [2025-10-26 16:02:19] ::plot_cnv:Start\n",
      "INFO [2025-10-26 16:02:19] ::plot_cnv:Current data dimensions (r,c)=14961,1875 Total=28487397.5 Min=0.5 Max=3.\n",
      "INFO [2025-10-26 16:02:20] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2025-10-26 16:02:20] plot_cnv_observation:Start\n",
      "INFO [2025-10-26 16:02:20] Observation data size: Cells= 1477 Genes= 14961\n",
      "INFO [2025-10-26 16:02:20] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2025-10-26 16:02:20] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2025-10-26 16:02:20] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 16:02:20] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 16:02:23] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 16:02:23] Quantiles of plotted data range: 0.5,1,1,1,3\n",
      "INFO [2025-10-26 16:02:25] plot_cnv_references:Start\n",
      "INFO [2025-10-26 16:02:25] Reference data size: Cells= 398 Genes= 14961\n",
      "INFO [2025-10-26 16:02:26] plot_cnv_references:Number reference groups= 1\n",
      "INFO [2025-10-26 16:02:26] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2025-10-26 16:02:27] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 16:02:27] Quantiles of plotted data range: 0.5,1,1,1,1.5\n",
      "INFO [2025-10-26 16:02:27] \n",
      "\n",
      "\tSTEP 22: Denoising\n",
      "\n",
      "INFO [2025-10-26 16:02:27] ::process_data:Remove noise, noise threshold defined via ref mean sd_amplifier:  1.5\n",
      "INFO [2025-10-26 16:02:27] denoising using mean(normal) +- sd_amplifier * sd(normal) per gene per cell across all data\n",
      "INFO [2025-10-26 16:02:27] :: **** clear_noise_via_ref_quantiles **** : removing noise between bounds:  0.902073711761648 - 1.10244986736508\n",
      "INFO [2025-10-26 16:02:55] \n",
      "\n",
      "## Making the final infercnv heatmap ##\n",
      "INFO [2025-10-26 16:02:56] ::plot_cnv:Start\n",
      "INFO [2025-10-26 16:02:56] ::plot_cnv:Current data dimensions (r,c)=14961,1875 Total=28251861.0629033 Min=0.548464152420786 Max=1.99385371368565.\n",
      "INFO [2025-10-26 16:02:56] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2025-10-26 16:02:57] plot_cnv(): auto thresholding at: (0.682245 , 1.317755)\n",
      "INFO [2025-10-26 16:02:57] plot_cnv_observation:Start\n",
      "INFO [2025-10-26 16:02:57] Observation data size: Cells= 1477 Genes= 14961\n",
      "INFO [2025-10-26 16:02:57] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2025-10-26 16:02:57] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2025-10-26 16:02:57] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 16:02:57] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 16:03:00] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 16:03:00] Quantiles of plotted data range: 0.68224459014128,1.00226178956337,1.00226178956337,1.00226178956337,1.31775540985872\n",
      "INFO [2025-10-26 16:03:02] plot_cnv_references:Start\n",
      "INFO [2025-10-26 16:03:02] Reference data size: Cells= 398 Genes= 14961\n",
      "INFO [2025-10-26 16:03:03] plot_cnv_references:Number reference groups= 1\n",
      "INFO [2025-10-26 16:03:03] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2025-10-26 16:03:03] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 16:03:03] Quantiles of plotted data range: 0.68224459014128,1.00226178956337,1.00226178956337,1.00226178956337,1.31775540985872\n"
     ]
    }
   ],
   "source": [
    "infercnv_obj = CreateInfercnvObject(raw_counts_matrix=\"/home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_input/exp_arch_prolif_MARS.tsv\",\n",
    "                                    annotations_file=\"/home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_input/ann_arch_prolif_MARS.txt\",\n",
    "                                    delim=\"\\t\",\n",
    "                                    gene_order_file=\"/home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_input/ann_gene.txt\",\n",
    "                                    ref_group_names=c(\"Normal_PC\"))\n",
    "\n",
    "infercnv_obj = infercnv::run(infercnv_obj,\n",
    "                             cutoff=0.1,  \n",
    "                             out_dir=\"/home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_output/output_infercnv_r_arch_prolif_MARS\",  \n",
    "                             cluster_by_groups=TRUE,   \n",
    "                             denoise=TRUE,\n",
    "                             HMM=TRUE,\n",
    "                             output_format=\"pdf\"\n",
    "                             )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "05ad2a4e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 16:03:04] Parsing matrix: /home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_input/exp_arch_prolif_SPID.tsv\n",
      "INFO [2025-10-26 16:03:17] Parsing gene order file: /home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_input/ann_gene.txt\n",
      "INFO [2025-10-26 16:03:17] Parsing cell annotations file: /home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_input/ann_arch_prolif_SPID.txt\n",
      "INFO [2025-10-26 16:03:17] ::order_reduce:Start.\n",
      "INFO [2025-10-26 16:03:17] .order_reduce(): expr and order match.\n",
      "INFO [2025-10-26 16:03:17] ::process_data:order_reduce:Reduction from positional data, new dimensions (r,c) = 52317,1460 Total=135049975 Min=0 Max=23107.\n",
      "INFO [2025-10-26 16:03:18] num genes removed taking into account provided gene ordering list: 2340 = 4.47273352829864% removed.\n",
      "INFO [2025-10-26 16:03:18] -filtering out cells < 100 or > Inf, removing 0 % of cells\n",
      "WARN [2025-10-26 16:03:19] Please use \"options(scipen = 100)\" before running infercnv if you are using the analysis_mode=\"subclusters\" option or you may encounter an error while the hclust is being generated.\n",
      "INFO [2025-10-26 16:03:21] validating infercnv_obj\n",
      "INFO [2025-10-26 16:03:21] ::process_data:Start\n",
      "INFO [2025-10-26 16:03:21] Checking for saved results.\n",
      "INFO [2025-10-26 16:03:21] Trying to reload from step 2\n",
      "INFO [2025-10-26 16:03:22] Using backup from step 2\n",
      "INFO [2025-10-26 16:03:22] \n",
      "\n",
      "\tSTEP 1: incoming data\n",
      "\n",
      "INFO [2025-10-26 16:03:22] \n",
      "\n",
      "\tSTEP 02: Removing lowly expressed genes\n",
      "\n",
      "INFO [2025-10-26 16:03:22] \n",
      "\n",
      "\tSTEP 03: normalization by sequencing depth\n",
      "\n",
      "INFO [2025-10-26 16:03:22] normalizing counts matrix by depth\n",
      "INFO [2025-10-26 16:03:22] Computed total sum normalization factor as median libsize: 76539.000000\n",
      "INFO [2025-10-26 16:03:22] Adding h-spike\n",
      "INFO [2025-10-26 16:03:22] -hspike modeling of Normal_PC\n",
      "INFO [2025-10-26 16:05:41] validating infercnv_obj\n",
      "INFO [2025-10-26 16:05:41] normalizing counts matrix by depth\n",
      "INFO [2025-10-26 16:05:41] Using specified normalization factor: 76539.000000\n",
      "INFO [2025-10-26 16:05:57] \n",
      "\n",
      "\tSTEP 04: log transformation of data\n",
      "\n",
      "INFO [2025-10-26 16:05:57] transforming log2xplus1()\n",
      "INFO [2025-10-26 16:05:57] -mirroring for hspike\n",
      "INFO [2025-10-26 16:05:57] transforming log2xplus1()\n",
      "INFO [2025-10-26 16:06:13] \n",
      "\n",
      "\tSTEP 08: removing average of reference data (before smoothing)\n",
      "\n",
      "INFO [2025-10-26 16:06:13] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
      "INFO [2025-10-26 16:06:13] subtracting mean(normal) per gene per cell across all data\n",
      "INFO [2025-10-26 16:06:17] -subtracting expr per gene, use_bounds=TRUE\n",
      "INFO [2025-10-26 16:06:19] -mirroring for hspike\n",
      "INFO [2025-10-26 16:06:19] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
      "INFO [2025-10-26 16:06:19] subtracting mean(normal) per gene per cell across all data\n",
      "INFO [2025-10-26 16:06:22] -subtracting expr per gene, use_bounds=TRUE\n",
      "INFO [2025-10-26 16:06:41] \n",
      "\n",
      "\tSTEP 09: apply max centered expression threshold: 3\n",
      "\n",
      "INFO [2025-10-26 16:06:41] ::process_data:setting max centered expr, threshold set to: +/-:  3\n",
      "INFO [2025-10-26 16:06:41] -mirroring for hspike\n",
      "INFO [2025-10-26 16:06:41] ::process_data:setting max centered expr, threshold set to: +/-:  3\n",
      "INFO [2025-10-26 16:07:00] \n",
      "\n",
      "\tSTEP 10: Smoothing data per cell by chromosome\n",
      "\n",
      "INFO [2025-10-26 16:07:00] smooth_by_chromosome: chr: chr1\n",
      "INFO [2025-10-26 16:07:03] smooth_by_chromosome: chr: chr14\n",
      "INFO [2025-10-26 16:07:04] smooth_by_chromosome: chr: chr19\n",
      "INFO [2025-10-26 16:07:06] smooth_by_chromosome: chr: chr10\n",
      "INFO [2025-10-26 16:07:07] smooth_by_chromosome: chr: chr12\n",
      "INFO [2025-10-26 16:07:09] smooth_by_chromosome: chr: chr22\n",
      "INFO [2025-10-26 16:07:10] smooth_by_chromosome: chr: chr3\n",
      "INFO [2025-10-26 16:07:12] smooth_by_chromosome: chr: chr5\n",
      "INFO [2025-10-26 16:07:14] smooth_by_chromosome: chr: chr4\n",
      "INFO [2025-10-26 16:07:15] smooth_by_chromosome: chr: chr9\n",
      "INFO [2025-10-26 16:07:16] smooth_by_chromosome: chr: chr15\n",
      "INFO [2025-10-26 16:07:18] smooth_by_chromosome: chr: chr2\n",
      "INFO [2025-10-26 16:07:20] smooth_by_chromosome: chr: chr11\n",
      "INFO [2025-10-26 16:07:21] smooth_by_chromosome: chr: chr17\n",
      "INFO [2025-10-26 16:07:23] smooth_by_chromosome: chr: chr20\n",
      "INFO [2025-10-26 16:07:24] smooth_by_chromosome: chr: chr8\n",
      "INFO [2025-10-26 16:07:26] smooth_by_chromosome: chr: chr16\n",
      "INFO [2025-10-26 16:07:27] smooth_by_chromosome: chr: chr6\n",
      "INFO [2025-10-26 16:07:29] smooth_by_chromosome: chr: chr7\n",
      "INFO [2025-10-26 16:07:30] smooth_by_chromosome: chr: chr21\n",
      "INFO [2025-10-26 16:07:31] smooth_by_chromosome: chr: chr13\n",
      "INFO [2025-10-26 16:07:32] smooth_by_chromosome: chr: chr18\n",
      "INFO [2025-10-26 16:07:33] -mirroring for hspike\n",
      "INFO [2025-10-26 16:07:33] smooth_by_chromosome: chr: chrA\n",
      "INFO [2025-10-26 16:07:34] smooth_by_chromosome: chr: chr_0\n",
      "INFO [2025-10-26 16:07:34] smooth_by_chromosome: chr: chr_B\n",
      "INFO [2025-10-26 16:07:35] smooth_by_chromosome: chr: chr_0pt5\n",
      "INFO [2025-10-26 16:07:35] smooth_by_chromosome: chr: chr_C\n",
      "INFO [2025-10-26 16:07:35] smooth_by_chromosome: chr: chr_1pt5\n",
      "INFO [2025-10-26 16:07:35] smooth_by_chromosome: chr: chr_D\n",
      "INFO [2025-10-26 16:07:35] smooth_by_chromosome: chr: chr_2pt0\n",
      "INFO [2025-10-26 16:07:35] smooth_by_chromosome: chr: chr_E\n",
      "INFO [2025-10-26 16:07:36] smooth_by_chromosome: chr: chr_3pt0\n",
      "INFO [2025-10-26 16:07:36] smooth_by_chromosome: chr: chr_F\n",
      "INFO [2025-10-26 16:07:56] \n",
      "\n",
      "\tSTEP 11: re-centering data across chromosome after smoothing\n",
      "\n",
      "INFO [2025-10-26 16:07:56] ::center_smooth across chromosomes per cell\n",
      "INFO [2025-10-26 16:07:58] -mirroring for hspike\n",
      "INFO [2025-10-26 16:07:58] ::center_smooth across chromosomes per cell\n",
      "INFO [2025-10-26 16:08:18] \n",
      "\n",
      "\tSTEP 12: removing average of reference data (after smoothing)\n",
      "\n",
      "INFO [2025-10-26 16:08:18] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
      "INFO [2025-10-26 16:08:18] subtracting mean(normal) per gene per cell across all data\n",
      "INFO [2025-10-26 16:08:22] -subtracting expr per gene, use_bounds=TRUE\n",
      "INFO [2025-10-26 16:08:23] -mirroring for hspike\n",
      "INFO [2025-10-26 16:08:23] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
      "INFO [2025-10-26 16:08:23] subtracting mean(normal) per gene per cell across all data\n",
      "INFO [2025-10-26 16:08:26] -subtracting expr per gene, use_bounds=TRUE\n",
      "INFO [2025-10-26 16:08:45] \n",
      "\n",
      "\tSTEP 14: invert log2(FC) to FC\n",
      "\n",
      "INFO [2025-10-26 16:08:45] invert_log2(), computing 2^x\n",
      "INFO [2025-10-26 16:08:46] -mirroring for hspike\n",
      "INFO [2025-10-26 16:08:46] invert_log2(), computing 2^x\n",
      "INFO [2025-10-26 16:09:08] \n",
      "\n",
      "\tSTEP 15: computing tumor subclusters via leiden\n",
      "\n",
      "INFO [2025-10-26 16:09:08] define_signif_tumor_subclusters(p_val=0.1\n",
      "INFO [2025-10-26 16:09:08] define_signif_tumor_subclusters(), tumor: Interm\n",
      "INFO [2025-10-26 16:09:08] Setting auto leiden resolution for Interm to 0.593499\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  PBXIP1, AL451085.1, PMVK, KCNN3, ADAR, PYGO2, UBE2Q1, AL451085.2, PSMD8P1, SHC1 \n",
      "\t   IPO4, IL6R, AL136295.4, TM9SF1, CKS1B, AL136295.6, AL162591.1, AL136295.7, MIR4258, AL096870.2 \n",
      "\t   AC145285.1, RPSAP17, AC145285.2, CHMP4A, EIF3C, FLAD1, SULT1A1, MDP1, SULT1A2, ORMDL3 \n",
      "Negative:  AP003352.1, RIDA, POP1, NIPAL2, RNU6-914P, STK3, AP003355.1, RNU6-748P, AP003467.2, AC104986.2 \n",
      "\t   AC104986.1, VPS13B, COX6C, FBXO43, ZNF880, POLR2K, MFAP3L, AC079768.1, LINC02382, AC108064.1 \n",
      "\t   PTGES3P3, AC105285.1, ZNF528-AS1, GALNT7, HPF1, CLCN3, NEK1, NOL8P1, HADHAP1, DDX60 \n",
      "PC_ 2 \n",
      "Positive:  RPL23AP18, AL360091.1, PPP2R5A, TMEM206, RPL21P28, NENF, DTL, AC092803.2, INTS7, ATF3 \n",
      "\t   LPGAT1, ATP5J2P5, TNFRSF1A, BATF3, CD9, LTBR, AC005840.1, NEK2, AC006206.1, CD27-AS1 \n",
      "\t   NSL1, SLC30A1, AC005840.2, KCNA5, ARPC3P2, CD27, TATDN3, SNX25P1, LINC00467, TAPBPL \n",
      "Negative:  UTP3, RUFY3, GRSF1, ENAM, UGT2B17, TMPRSS11E, MT2P1, MOB1B, YTHDC1, DCK \n",
      "\t   GCOM2, TMPRSS11A, UBA6-AS1, COX18, STAP1, UBA6, CENPC, RNU1-63P, ANKRD17, AC104806.2 \n",
      "\t   AC116049.1, AC115223.1, AC053527.1, AC093730.1, LARP1BP1, RASSF6, EXOC5P1, RPS15AP17, ADGRL3, UBE2CP3 \n",
      "PC_ 3 \n",
      "Positive:  NAP1L4, PHLDA2, SLC22A18, CDKN1C, MIR6755, EED, AP003097.1, PICALM, AP003084.1, SYTL2 \n",
      "\t   HIKESHI, CREBZF, ME3, PRSS23, TMEM126A, AP000654.2, FZD4, TMEM126B, AC024575.1, HNRNPCP8 \n",
      "\t   HNRNPCP6, PLPPR2, TMEM135, CCDC159, AP000852.1, AP003497.2, TMEM205, AP001784.1, DLG2, ARHGAP10 \n",
      "Negative:  PSME2P2, AL157813.1, RCBTB2, AL161421.1, FNDC3A, RNU6-60P, LPAR6, COX7CP1, CDADC1, PCNPP5 \n",
      "\t   CAB39L, SETDB2, DLEU2, SNRPGP14, RB1, SPRYD7, PHF11, CTAGE10P, RCBTB1, ARL11 \n",
      "\t   EBPL, KPNA3, LINC00441, ITM2B, MED4, NUDT15, SUCLA2, LINC00562, NAP1L4P3, ESD \n",
      "PC_ 4 \n",
      "Positive:  BBS7, CCNA2, EXOSC9, ANXA5, TNIP3, RN7SKP137, NDNF, SETP12, PRDM5, AC073475.1 \n",
      "\t   AC108866.1, MAD2L1, PDE5A, AC093752.1, GTF2IP12, C4orf3, USP53, AC008536.3, DCP2, ZRSR1 \n",
      "\t   XBP1P1, MCC, SEC24D, YTHDC2, TRIM36, METTL14, PGGT1B, AC110079.2, AC008494.3, CCDC112 \n",
      "Negative:  PLEKHM2, RPL12P14, UQCRHL, AL450998.1, FLJ37453, SPEN, ZBTB17, ARHGEF19, ANO7L1, MT1XP1 \n",
      "\t   FBXO42, COL19A1, SZRD1, AL365226.2, AL358794.1, FAM135A, NECAP2, SDHAF4, LINC01772, AL583856.1 \n",
      "\t   CROCCP3, SMAP1, NDUFAB1P1, RNU1-1, LYPLA1P3, NBPF1, OGFRL1, CROCCP2, RIMS1, KCNQ5 \n",
      "PC_ 5 \n",
      "Positive:  AL356535.1, AL133260.1, ARL4AP5, MTHFD1L, PDCL3P5, PLEKHG1, SMAP1, NDUFAB1P1, BTF3P10, OGFRL1 \n",
      "\t   LYPLA1P3, AL583856.1, RIMS1, KCNQ5, SDHAF4, KCNQ5-IT1, KHDC1L, KHDC1, FAM135A, AC019205.2 \n",
      "\t   LRP11, EIF3EP1, AL365226.2, AC019205.3, COL19A1, DDX43, MB21D1, BTBD10P2, MTO1, RNU6-975P \n",
      "Negative:  IFFO1, GAPDH, NOP2, NCAPD2, CHD4, VAMP1, LPAR5, TAPBPL, CD27, ACRBP \n",
      "\t   AC005840.2, CD27-AS1, AC005840.1, ING4, LTBR, ZNF384, COPS7A, TNFRSF1A, MLF2, ATP5J2P5 \n",
      "\t   PTMS, CD9, LAG3, AC006206.1, CD4, AP003469.2, YWHAZ, AP003469.4, RPS26P6, KCNA5 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 16:09:09] define_signif_tumor_subclusters(), tumor: Malignant_MM1_Non-Prolif\n",
      "INFO [2025-10-26 16:09:09] Setting auto leiden resolution for Malignant_MM1_Non-Prolif to 0.0683262\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  RPS29, AL512361.1, AL139099.1, LRR1, RN7SL1, AL157792.1, RHOQP1, FANCM, RPL36AL, MIS18BP1 \n",
      "\t   RPL10L, FKBP3, MGAT2, AL139099.2, PRPF39, DNAAF2, AL139099.3, POLE2, TOGARAM1, KLHDC1 \n",
      "\t   KLHL28, KLHDC2, SIPA1L1, PCNX1, AC005476.1, RGS6, AC004816.1, DPF3, DCAF4, NEMF \n",
      "Negative:  NCSTN, SUMO1P3, SLAMF6, CD84, COPA, RPSAP18, PEX19, AL139011.1, SLAMF1, DCAF8 \n",
      "\t   AL121985.1, CD48, SLAMF7, LY9, AL354714.2, F11R, AL591806.3, USF1, ARHGAP30, NECTIN4 \n",
      "\t   KLHDC9, PFDN2, NIT1, DEDD, ACA64, TRADD, B3GNT9, C16orf70, CBFB, CES4A \n",
      "PC_ 2 \n",
      "Positive:  HIST2H2AA4, HIST2H4B, HIST2H2BE, HIST2H2BC, HIST2H2AC, HIST2H2AB, BOLA1, SV2A, SF3B4, OTUD7B \n",
      "\t   POLR1C, TJAP1, YIPF3, ABCC10, AL355802.3, ZNF544, AL583834.1, ZSCAN22, XPO5, ZNF274 \n",
      "\t   RPL12P47, AL355802.2, ZNF318, ZNF329, ZSCAN18, DNPH1, LETM1P2, ZNF256, AC008969.1, ZNF606 \n",
      "Negative:  B3GLCT, ALOX5AP, USPL1, PTPN2P2, RBM22P2, HMGB1, UBE2L5P, LINC00426, PRDX2P1, KATNAL1 \n",
      "\t   UBL3, SLC7A1, GAPDHP69, CYP51A1P2, SLC46A3, POMP, EIF4A1P7, PPIC, SNX24, CEP120 \n",
      "\t   AC008669.1, KRT8P33, CSNK1G3, SNX2, SRFBP1, AC106806.1, AC008568.1, HSD17B4, TNFAIP8, AC118465.1 \n",
      "PC_ 3 \n",
      "Positive:  PROX1-AS1, RPS6KC1, PROX1, ANGEL2, SMYD2, VASH2, AL445305.1, CENPF, FLVCR1, FLVCR1-AS1 \n",
      "\t   KCTD3, TATDN3, NSL1, USH2A, BATF3, ATF3, ESRRG, AC092803.2, NENF, TMEM206 \n",
      "\t   AL360091.1, GPATCH2, RPL23AP18, PPP2R5A, RPL21P28, DTL, INTS7, SPATA17, LPGAT1, NEK2 \n",
      "Negative:  DCTN6, AC026979.3, AC026979.1, LEPROTL1, AC044849.2, SARAF, LINC02099, MAK16, AC069181.1, SNORD13 \n",
      "\t   DUSP4, RNF122, DUSP26, KIF13B, AF279873.3, AF279873.1, AC108449.2, AC087343.1, LSM12P1, HMBOX1 \n",
      "\t   AC124290.1, INTS9, TPT1P8, EXTL3, SMARCE1P4, FZD3, ERLIN2, FBXO16, AC138356.2, ZNF395 \n",
      "PC_ 4 \n",
      "Positive:  PPP2CA, SKP1, TCF7, AC008608.2, AC104109.4, VDAC1, CDKL3, C5orf15, AC010307.4, UBE2B \n",
      "\t   FSTL4, AC109454.2, AC113410.1, CDKN2AIPNL, ZCCHC10, AC109454.1, JADE2, AC010240.1, SAR1B, AFF4 \n",
      "\t   SEC24A, LEAP2, UQCRQ, CAMLG, GDF9, DDX46, SHROOM1, AC010301.1, SEPT8, C5orf24 \n",
      "Negative:  CBX3P9, TSPYL1, ZBTB2, AKAP12, RMND1, AL356535.1, AL050331.1, ARMT1, AL133260.1, ARL4AP5 \n",
      "\t   ESR1, MTHFD1L, SYNE1, PDCL3P5, PLEKHG1, AL080276.1, BTF3P10, FBXO5, LRP11, BTBD10P2 \n",
      "\t   DSE, PCMT1, NUP43, AL080276.2, AL358852.1, MTRF1L, LATS1, TSPYL4, RPS18P9, RGS17 \n",
      "PC_ 5 \n",
      "Positive:  AC010336.1, EVI5L, FCER2, TRAPPC5, RPS27AP19, STXBP2, HLA-E, GNL1, NRM, PPP1R18 \n",
      "\t   MDC1, PRR3, TUBB, DHX16, RANP1, ABCF1, FLOT1, C6orf136, MIR877, ATAT1 \n",
      "\t   PPP1R10, SUCLA2P1, AC008763.2, AL662797.2, RPP21, TRIM39-RPP21, IER3, TRIM39, DDR1, HCG18 \n",
      "Negative:  C11orf53, FDX1, RDX, ZC3H12C, DDX10, EXPH5, KDELC2, C11orf65, ATM, NPAT \n",
      "\t   ACAT1, CUL5, AP001024.1, SLC35F2, ELMOD1, ALKBH8, SMARCE1P1, CWF19L2, AP000766.1, NDUFB11P1 \n",
      "\t   SMCO4, SLC36A4, AP004242.1, OSBPL9P2, AP002782.1, CEP295, snoU2-19, snoU2-30, TAF1D, TUBAP2 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 16:09:10] define_signif_tumor_subclusters(), tumor: Malignant_MM1_Prolif\n",
      "INFO [2025-10-26 16:09:10] Setting auto leiden resolution for Malignant_MM1_Prolif to 0.24731\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  KCTD6, C3orf67, FHIT, AC135507.1, PTPRG, PDHB, PXK, AC096919.1, HTD2, RPP14 \n",
      "\t   ABHD6, FLNB, PPIAP16, SLMAP, DENND6A, ARF4-AS1, ARF4, PDE12, AC092418.1, RNF7P1 \n",
      "\t   APPL1, PTMAP5, NDFIP2, DDX6P2, RBM26-AS1, MIR4500HG, SLITRK5, RBM26, CCT5P2, SP3P \n",
      "Negative:  ZFYVE21, AL049840.4, XRCC3, AL049840.3, AL049840.2, AL049840.5, APOPT1, KLC1, BAG5, TRMT61A \n",
      "\t   CKB, MATN2, RPL30, AP003352.1, RPS23P1, RPL10AP1, RIDA, LAPTM4B, MARK3, POP1 \n",
      "\t   NIPAL2, SNORA28, RNU6-914P, MTDH, EIF5, STK3, AP003355.1, RNU6-748P, AC004884.1, TNFAIP2 \n",
      "PC_ 2 \n",
      "Positive:  AL122023.1, ASB2, BTBD7, OTUB2, UBR7, AL132838.1, IFI27L1, DDX24, IFI27, GON7 \n",
      "\t   TMEM251, IFI27L2, SERPINA1, MOAP1, ADIPOR1P2, ITPK1-AS1, AL121612.1, DICER1, ITPK1, DICER1-AS1 \n",
      "\t   CLMN, SYNE3, GOLGA5, AL133467.1, SCARNA13, SNHG10, GLRX5, LGMN, AL133467.2, RIN3 \n",
      "Negative:  APEH, DAG1, NICN1, AMT, TCTA, RHOA-IT1, C3orf62, RHOA, KLHDC8B, USP4 \n",
      "\t   GPX1, CCDC71, USP19, QARS, QRICH1, AC137630.3, IMPDH2, NDUFAF3, MIR425, DALRD3 \n",
      "\t   WDR6, P4HTM, AC137630.4, ARIH2, ARIH2OS, SLC25A20, PRKAR2A-AS1, PRKAR2A, IP6K2, NCKIPSD \n",
      "PC_ 3 \n",
      "Positive:  PRADC1, CCT7, FBXO41, ALMS1, ALMS1-IT1, TPRKB, AC092653.1, DUSP11, STAMBP, DGUOK \n",
      "\t   DGUOK-AS1, TET3, FNBP1P1, BOLA3, BOLA3-AS1, MOB1A, LRP12, MTHFD2, SLC4A5, DCTN1 \n",
      "\t   AC090686.1, C2orf81, RIMS2, WDR54, DCAF13, RTKN, INO80B, SLC25A32, INO80B-WBP1, AC012213.3 \n",
      "Negative:  AL161742.1, CACHD1, RAVER2, JAK1, LINC01359, AL357078.3, RNU6-1176P, MIR3671, GDI2P2, RPL31P12 \n",
      "\t   NEGR1, ZRANB2-AS2, MIR186, LEPROT, ZRANB2, AL354872.1, CTH, HHLA3, ANKRD13C, LEPR \n",
      "\t   SRSF11, LRRC40, PDE4B, PIN1P1, WDR78, DEPDC1, MIER1, RPS7P4, SLC35D1, WLS \n",
      "PC_ 4 \n",
      "Positive:  NHP2, AC136632.2, MEF2A, LRRC28, AC110588.1, ARRDC4, IGF1R, SYNM, TTC23, NR2F2-AS1 \n",
      "\t   GMCL1P1, AC087633.2, MCTP2, AC104390.1, CHD2, AC013394.1, HAUS6P1, SLC25A5P3, LANCL2, SEC61G \n",
      "\t   AC074397.1, CALM1P2, VOPP1, COBL, SUMO2P3, PSPHP1, FIGNL1, NIPSNAP2, IKZF1, PSPH \n",
      "Negative:  FBXL15, CUEDC2, C10orf95, RPARP-AS1, ACTR1A, MFSD13A, SUFU, RPL23AP58, AL391121.1, MPP7 \n",
      "\t   ARMC4, MKX, RAB18, ACBD5, MASTL, YME1L1, ANKRD26, SNORA57, RNU6-946P, ABI1 \n",
      "\t   MIR1226, CCDC51, PLXNB1, NME6, CDC25A, AC134772.1, FCF1P2, PDSS1, TMA7, AC134772.2 \n",
      "PC_ 5 \n",
      "Positive:  INTS12, AC105391.1, GSTCD, ARHGEF38, EEF1A1P9, NPNT, PPA2, TET2, AC004069.1, TBCK \n",
      "\t   CXXC4, BDH2, LINC02428, CENPE, SLC9B2, SLC9B1, AIMP1, CISD2, AC018797.2, UBE2D3 \n",
      "\t   PAPSS1, AC018797.3, MANBA, NFKB1, CYP2U1, AF213884.3, AC098487.1, SLC39A8, BANK1, FLJ20021 \n",
      "Negative:  UQCR10, ASCC2, AC003681.1, CNN2P1, AC004997.1, CASTOR1, TBC1D10A, PARP12, TBXAS1, SF3A1 \n",
      "\t   HIPK2, ERHP1, CCDC157, CLEC2L, AC005531.1, RNF215, AC083880.1, C7orf55-LUC7L2, LUC7L2, AC004832.4 \n",
      "\t   FMC1, UBN2, AC009220.1, TTC26, ZC3HAV1, KIAA1549, AC020983.1, TRIM24, MTFP1, AC008155.1 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 16:09:11] define_signif_tumor_subclusters(), tumor: Malignant_MM2_Non-Prolif\n",
      "INFO [2025-10-26 16:09:11] Setting auto leiden resolution for Malignant_MM2_Non-Prolif to 0.0916085\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  TTC38, CDPF1, FP325332.1, PPARA, MIRLET7BHG, PRR34-AS1, AL121672.3, LINC00899, BX537318.1, ATXN10 \n",
      "\t   Z95331.1, RIBC2, SMC1B, CDC45, CLDN5, AC000068.3, GP1BB, UFD1, GNB1L, AC000089.1 \n",
      "\t   RTL10, TXNRD2, AC000068.1, COMT, TANGO2, C22orf39, DGCR8, TRMT2A, HIRA, RANBP1 \n",
      "Negative:  HMCES, COPG1, H1FX, CNBP, H1FX-AS1, AC108673.3, RPL32P3, ISY1, RAB43, MBD4 \n",
      "\t   GP9, IFT122, ACAD9, PLXND1, AC112484.2, TMCC1, FTH1P4, TMCC1-AS1, RAB7A, AC083906.1 \n",
      "\t   SNORA24, ALG1L2, RPN1, LINC02014, EEFSEC, FAM86HP, RUVBL1, LINC02021, SEC61A1, SNRPCP8 \n",
      "PC_ 2 \n",
      "Positive:  RPSAP17, AL162591.1, IL6R, PSMD8P1, UBE2Q1, ADAR, KCNN3, PMVK, AL451085.1, PBXIP1 \n",
      "\t   AL353807.2, ASH1L-AS1, RNU6-1297P, ASH1L, RUSC1, PYGO2, RUSC1-AS1, FDPS, HCN3, AL451085.2 \n",
      "\t   CLK2, SCAMP3, SHC1, FAM189B, GBA, CKS1B, MTX1P1, MIR4258, GBAP1, MTX1 \n",
      "Negative:  AP002360.2, EMSY, ACER3, AP002360.4, CAPN5, AP002360.1, TOMM20P1, PAK1, THAP12, CLNS1A \n",
      "\t   WNT11, AQP11, AP003168.2, RSF1, FTH1P16, AAMDC, UVRAG, INTS4, ALG8, KCTD21-AS1 \n",
      "\t   NDUFC2, THRSP, NDUFC2-KCTD14, KCTD21, USP35, DGAT2, GAB2, SERPINH1, AP003086.1, FABP5P1 \n",
      "PC_ 3 \n",
      "Positive:  SCARNA24, GMEB1, LINC01715, RAB42, TAF12, SNORA61, TRNAU1AP, SNHG3, RCC1, CR391992.1 \n",
      "\t   PHACTR4, MED18, SESN2, JMJD7, MAPKBP1, AC073657.2, ATG16L2, FCHSD2, ARAP1, RNU6-672P \n",
      "\t   RPS12P20, MGA, AP000593.3, STARD10, CLPB, AP002761.2, INPPL1, AP002761.4, TYRO3, FOLR2 \n",
      "Negative:  AC092447.2, AC095038.3, AC092447.7, AC073136.2, AC073136.1, AC099654.2, SLC25A1P2, AC006970.2, AC092634.6, AC006970.1 \n",
      "\t   NMD3P1, CHCHD2, AC091685.2, PHKG1, ZNF736, SUMF2, YWHAEP1, SNORA22B, CCT6A, ZNF680 \n",
      "\t   PSPH, NIPSNAP2, PSPHP1, ZNF107, SUMO2P3, AC091799.1, VOPP1, LANCL2, ZNF138, CALM1P2 \n",
      "PC_ 4 \n",
      "Positive:  AP000902.1, VAPA, RNA5SP450, RAB31, NAPG, PPP4R1, AP005432.1, CCDC58P1, RALBP1, AP001180.5 \n",
      "\t   TWSG1, PIEZO2, AP001033.2, GNAL, ANKRD12, CHMP1B, NDUFV2-AS1, MPPE1, IMPA2, AP005263.1 \n",
      "\t   AP002414.4, CCDC58P3, NDUFV2, TUBB6, AFG3L2, RPS4XP19, SPIRE1, PSMG2, CEP76, AP005482.3 \n",
      "Negative:  DYNC1I1, SLC25A13, AC002451.1, PDK4, PON2, AC084368.1, PPP1R9A, SEM1, RPS3AP25, SDHAF3 \n",
      "\t   PEG10, SGCE, ASNS, CASD1, COL1A2, RPS3AP29, BET1, AC006378.2, OR7E7P, GNG11 \n",
      "\t   OR7E38P, AC002075.2, VPS50, CRYGS, DGKG, HEPACAM2, ETV5, TRA2B, SAMD9L, IGF2BP2 \n",
      "PC_ 5 \n",
      "Positive:  CENPT, THAP11, NUTF2, AC040162.2, CKLF, GOT2, SLC38A7, AC009118.3, RPS27P27, APOOP5 \n",
      "\t   RPS27AP16, AC009094.1, CNOT1, SETD6, GLI4, AC009118.2, MINCR, AC138696.2, GINS3, ZNF696 \n",
      "\t   CCDC113, CSNK2A2, CFAP20, USB1, C11orf21, TSPAN32, FBXO31, AC010531.6, RPL26P30, AC010531.2 \n",
      "Negative:  CEP85L, AL589993.1, AL450405.1, SLC35F1, NUS1, GOPC, DCBLD1, KPNA5, ZUFSP, RSPH4A \n",
      "\t   RWDD1, TRAPPC3L, TSPYL1, AL050331.1, FAM26F, CBX3P9, TSPYL4, NT5DC1, FRK, HDAC2 \n",
      "\t   KLHDC2, KLHDC1, NEMF, POLE2, RN7SL3, AL139099.3, DNAAF2, AL139099.2, RN7SL2, MGAT2 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 16:09:12] define_signif_tumor_subclusters(), tumor: Malignant_MM2_Prolif\n",
      "INFO [2025-10-26 16:09:12] Setting auto leiden resolution for Malignant_MM2_Prolif to 0.139357\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  C7orf26, ZNF316, ZNF12, PMS2CL, CCZ1B, C1GALT1, AC004982.2, MIOS, RPA3, UMAD1 \n",
      "\t   CCNB2P1, AC006042.4, GLCCI1, AC010719.1, OSBPL3, CYCS, DFNA5, C7orf31, AC005165.2, HDAC9 \n",
      "\t   NFE2L3, MPP6, TWISTNB, RPL21P75, MACC1, AC004520.1, AC006042.3, PCMTD1P3, STK31, AC080080.1 \n",
      "Negative:  TTC7B, AL122020.1, CALM1, RPS6KA5, RPL21P11, C14orf159, NRDE2, CCDC88C, PSMC1, AL133153.2 \n",
      "\t   TDP1, PPP4R3A, POLR3GP1, TRIP11, AL049872.1, AC145285.1, ATXN3, AC145285.2, NDUFB1, EIF3C \n",
      "\t   CPSF2, SULT1A1, RIN3, SULT1A2, LGMN, L2HGDH, ATP5S, CDKL1, SOS2, MAP4K5 \n",
      "PC_ 2 \n",
      "Positive:  VCP, ATP8B5P, AL353795.3, FANCG, ZFAND6P1, UNC13B, FAM214B, STOML2, PIGO, SYF2P2 \n",
      "\t   RPS29P17, AL355377.1, CD72, YWHAZP6, AL162231.2, MIR4667, UBAP1, DCAF12, AL353662.2, KIF24 \n",
      "\t   OSTCP8, AL162231.1, DCTN3, FAM219A, C9orf24, NUDT2, RPP25L, SIGMAR1, IL11RA, SIT1 \n",
      "Negative:  KIF13A, AL138724.1, NUP153, FAM8A1, RPL7P26, USB1, CFAP20, UNC45A, HDDC3, RCCD1 \n",
      "\t   MAN2A2, PRC1, FES, VPS33B, FURIN, SV2B, CSNK2A2, SLCO3A1, AC124248.1, NPM1P5 \n",
      "\t   FAM174B, HMGN1P38, BLM, CCDC113, AC091544.3, AC106028.4, CRTC3-AS1, AC106028.5, LINC01578, CRTC3 \n",
      "PC_ 3 \n",
      "Positive:  CEP164, BACE1-AS, RNF214, PCSK7, TAGLN, SIDT2, PAFAH1B2, AP000936.3, SIK3, APOA1 \n",
      "\t   ZPR1, BUD13, LINC00900, CADM1, NXPE4, NXPE1, AP002373.2, OTUD7B, SF3B4, SV2A \n",
      "\t   REXO2, BOLA1, HIST2H2AB, HIST2H2AC, RBM7, HIST2H2BE, HIST2H4B, C11orf71, HIST2H2AA4, ZBTB16 \n",
      "Negative:  C17orf75, UBL5P2, ZNF207, RHOT1, AC005899.6, AC116407.2, AC090616.7, AC005899.7, LRRC37B, PSMD11 \n",
      "\t   CDK5R1, MYO1D, AC079336.3, TMEM98, SPACA3, AC024610.2, AC123769.1, NIPA2, CYFIP1, CCL1 \n",
      "\t   TUBGCP5, AC242376.1, CCT6B, AC242376.2, ZNF830, RN7SL536P, LIG3, AC100756.2, RFFL, MKRN3 \n",
      "PC_ 4 \n",
      "Positive:  RPS15AP11, SNORD38A, BEST4, PLK3, SNORD55, RNU5E-6P, AL136380.1, EIF2B3, RPS8, HECTD3 \n",
      "\t   UROD, KIF2C, ZSWIM5, RNU5D-1, OSTCP5, RNU5F-1, LINC01144, TMEM53, HPDL, RNF220 \n",
      "\t   MUTYH, ERI3, TOE1, DMAP1, AL356653.1, TESK2, AL139220.2, CCDC163, CCDC24, SLC6A9 \n",
      "Negative:  ZNF572, SQLE, MTSS1, WASHC5, NDUFB9, NSMCE2, RN7SL329P, AC090198.1, TRIB1, TATDN1 \n",
      "\t   SOD1P3, RNF139, FAM84B, PCAT1, RNF139-AS1, PRNCR1, TRMT12, AC020688.1, MYC, TMEM65 \n",
      "\t   PVT1, FAM91A1, TMEM75, FBXO32, FAM49B, WDYHV1, ASAP1, DUTP2, ASAP1-IT2, ATAD2 \n",
      "PC_ 5 \n",
      "Positive:  TIGD7, ZNF75A, AC025283.2, AC004232.1, ZSCAN32, ZNF174, ZNF597, NEDD8, GMPR2, NEDD8-MDP1 \n",
      "\t   TINF2, NAA60, MDP1, RABGGTA, KMT5B, CHMP4A, AC004224.1, C11orf24, DHRS1, AL096870.2 \n",
      "\t   LRP5, NOP9, AL136295.7, CLUAP1, AL136295.6, CREB3, CCDC107, TM9SF1, CIDEB, ARHGEF39 \n",
      "Negative:  ZNF830, LIG3, RFFL, CCT6B, RAD51D, CCL1, NLE1, AC022916.1, AC123769.1, SLFN5 \n",
      "\t   AC024610.2, AC022706.1, SPACA3, SLFN11, TMEM98, AC060766.3, AC079336.3, SLFN12, MYO1D, SLFN13 \n",
      "\t   CDK5R1, AC015911.4, PSMD11, SLFN12L, AC005899.7, AC015911.8, AC005899.6, AC015911.3, ZNF207, C17orf75 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 16:09:13] define_signif_tumor_subclusters(), tumor: Malignant_MM3_Non-Prolif\n",
      "INFO [2025-10-26 16:09:13] Setting auto leiden resolution for Malignant_MM3_Non-Prolif to 0.167569\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  AC005394.2, AC005381.1, AC079466.1, AC005307.1, UQCRFS1, POP4, PLEKHF1, AC006504.8, C19orf12, THBS3 \n",
      "\t   CCNE1, URI1, RPL9P32, AC008507.1, MTX1, AC006504.5, ZNF507, GBAP1, AC007773.1, MTX1P1 \n",
      "\t   AC006504.1, GBA, FAM189B, DPY19L3, SCAMP3, LINC00662, CLK2, PDCD5, HCN3, FDPS \n",
      "Negative:  MTCYBP4, RPL23AP46, SNORD100, EYA4, TBPL1, SLC2A12, RPS12, SGK1, AL121970.1, SLC18B1 \n",
      "\t   ALDH8A1, VNN2, HBS1L, MYB, VNN1, AHI1, STX7, LINC00271, MOXD1, MTFR2 \n",
      "\t   LINC01013, BCLAF1, AL357084.1, RN7SKP106, AL035446.1, AL589674.2, ATP5F1P6, AL133346.1, MAP7, CITED2 \n",
      "PC_ 2 \n",
      "Positive:  SAMD4A, GCH1, WDHD1, SYNE2, AL161670.2, CHMP4BP1, ESR2, SGPP1, ATG14, AL161756.1 \n",
      "\t   AL356801.1, TEX21P, AL136038.5, TBPL2, MTHFD1, AL122035.2, KTN1-AS1, ZBTB25, WDR89, AKAP5 \n",
      "\t   RPL7AP4, ZBTB1, AL136038.4, KTN1, AL049869.4, AL136038.3, RPL13AP3, AL049869.1, PPP2R5E, PELI2 \n",
      "Negative:  PHKG2, CCDC189, TMEM265, RNF40, SNORA30, ZNF629, AC106886.5, BCL7C, MIR762HG, FBXL19-AS1 \n",
      "\t   ORAI3, SETD1A, HSD3B7, STX4, ZNF668, ZNF646, AC073389.1, AC073389.3, AC022400.3, USP54 \n",
      "\t   AGAP5, PPP3CB-AS1, BMS1P4, PPP3CB, GLUD1P3, C7orf26, DUSP8P5, ZNF316, MSS51, ZNF12 \n",
      "PC_ 3 \n",
      "Positive:  HIST3H2A, HIST3H2BB, AL139288.1, RNF187, MIR6742, FTH1P2, TRIM11, OBSCN, RHOU, IBA57 \n",
      "\t   AL162595.1, PRAMENP, GUK1, RAB4A, ABHD17AP5, FAM96AP2, TOP3BP1, SPHAR, GMCL1P1, VPREB1 \n",
      "\t   C1orf35, AC245060.5, AC136632.2, CCSAP, ARF1, NHP2, PTGES3L-AARSD1, SNAP47, AL117350.1, JMJD4 \n",
      "Negative:  AL163051.1, GSKIP, ATG2B, TCL1A, TCL6, AL133467.2, IFI27L2, SERPINA1, GLRX5, ADIPOR1P2 \n",
      "\t   AL121612.1, IFI27, DICER1, DICER1-AS1, CLMN, IFI27L1, SNHG10, SYNE3, SCARNA13, AL133467.1 \n",
      "\t   DDX24, OTUB2, ASB2, AL122023.1, BTBD7, AL132838.1, UBR7, GON7, TMEM251, MOAP1 \n",
      "PC_ 4 \n",
      "Positive:  ZSCAN22, AC010642.2, AC020915.1, AC020915.2, ZNF8, AC020915.4, ZNF544, ZNF274, ZNF329, ZSCAN18 \n",
      "\t   LETM1P2, ZNF154, ZNF551, ZNF418, ZNF211, ZNF134, ZNF256, ZNF417, ZNF606, ZNF671 \n",
      "\t   AC008969.1, ZNF776, AC010326.2, ZNF552, ZNF586, ZNF814, ZNF587B, ZNF587, AC010326.4, AC010326.3 \n",
      "Negative:  PPP1R15B, PIK3C2B, MDM4, ACA64, DEDD, NIT1, PFDN2, KLHDC9, NECTIN4, ARHGAP30 \n",
      "\t   USF1, AL591806.3, F11R, AL354714.2, LY9, HLTF, HPS3, GYG1, WWTR1, COMMD2 \n",
      "\t   SLAMF7, UBQLN4P1, RNF13, CD48, NPM1P29, AL121985.1, PLSCR1, NCSTN, SLAMF1, SLAMF6 \n",
      "PC_ 5 \n",
      "Positive:  RPS15AP1, AL117332.1, XRN2, RNA5SP477, RPL41P1, KIZ, AL035258.1, CYB5AP4, RPL24P2, NXT1 \n",
      "\t   RALGAPA2, LINC01431, CRNKL1, GZF1, NAPB, NAA20, RNA5SP479, CST3, RPL12P12, SYNDIG1 \n",
      "\t   CST7, APMAP, AL035563.1, ACSS1, VSX1, ENTPD6, DTD1, AL035252.3, PYGB, ABHD12 \n",
      "Negative:  MORC3, SRSF9P1, RPL3P1, AP000692.1, DOPEY2, CHAF1B, RPS9P1, CLDN14, CBR1, PSMD4P1 \n",
      "\t   HLCS, SETD4, AP000704.1, RPL23AP3, PIGP, HDAC5, RUNX1, TTC3, C17orf53, RCAN1 \n",
      "\t   ASB16-AS1, AP001429.1, SMIM11A, TMUB2, CCNJ, AP001432.1, ZNF518A, RPS5P2, ATXN7L3, NPM1P25 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 16:09:14] define_signif_tumor_subclusters(), tumor: Malignant_MM3_Prolif\n",
      "INFO [2025-10-26 16:09:14] Less cells in group Malignant_MM3_Prolif than k_nn setting. Keeping as a single subcluster.\n",
      "INFO [2025-10-26 16:09:14] define_signif_tumor_subclusters(), tumor: Malignant_MM4_Non-Prolif\n",
      "INFO [2025-10-26 16:09:14] Setting auto leiden resolution for Malignant_MM4_Non-Prolif to 0.115566\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  SERTAD2, RNU6-100P, AC007365.1, MIR4434, AC007365.2, LINC01800, AFTPH, LINC02245, AC008074.3, RN7SL211P \n",
      "\t   LGALSL, AC007386.1, RPL23AP37, SLC1A4, LINC00309, AC007386.2, CEP68, AC012368.1, RAB1A, PELI1 \n",
      "\t   AC007318.1, VPS54, ACTR2, UGP2, SPRED2, AC096664.2, ACA64, DEDD, AC012370.1, AC016734.1 \n",
      "Negative:  ARRB2, AC091153.1, CXCL16, MED11, PSMB6, AC091153.3, PLD2, PELP1, MINK1, TXNP4 \n",
      "\t   CHRNE, GGT6, C17orf107, MYBBP1A, SPNS2, SLC25A11, MFSD1P1, RNF167, UBE2G1, PFN1 \n",
      "\t   AC087742.2, ENO3, AC087742.1, ANKFY1, SPAG7, CAMTA2, AC004771.4, AC004771.3, KIF1C, AC004771.1 \n",
      "PC_ 2 \n",
      "Positive:  PSME2P2, AL161421.1, AL157813.1, FNDC3A, RCBTB2, RNU6-60P, LPAR6, COX7CP1, PCNPP5, CDADC1 \n",
      "\t   RB1, CAB39L, LINC00441, ITM2B, SETDB2, MED4, SNRPGP14, NUDT15, SUCLA2, PHF11 \n",
      "\t   LINC00562, NAP1L4P3, RCBTB1, ESD, LRCH1, ARL11, COX17P1, RUBCNL, LCP1, EBPL \n",
      "Negative:  AL162591.1, IL6R, RPSAP17, PSMD8P1, KCNN3, ADAR, UBE2Q1, PMVK, AL451085.1, PBXIP1 \n",
      "\t   PYGO2, AL451085.2, SHC1, CKS1B, MIR4258, FLAD1, ZBTB7B, DCST1-AS1, ADAM15, EFNA4 \n",
      "\t   EFNA3, EFNA1, SLC50A1, DPM3, KRTCAP2, TRIM46, MUC1, THBS3, MTX1, GBAP1 \n",
      "PC_ 3 \n",
      "Positive:  CST7, APMAP, SYNDIG1, CST3, ACSS1, RNA5SP479, VSX1, NAPB, GZF1, ENTPD6 \n",
      "\t   LINC01431, AL035252.3, NXT1, CYB5AP4, AL035258.1, RPL41P1, XRN2, AL117332.1, RPS15AP1, RNA5SP477 \n",
      "\t   KIZ, RPL24P2, RALGAPA2, CRNKL1, NAA20, RPL12P12, AL035563.1, DTD1, SMIM26, SEC23B \n",
      "Negative:  PERP, LINC02528, ARFGEF3, AL591468.1, HEBP2, TNFAIP3, AL590617.2, GVQW2, AL357060.2, AL356234.1 \n",
      "\t   CCDC28A, IFNGR1, PEX7, REPS1, MAP3K5, ABRACL, AL023284.2, AL023284.3, HECA, MAP7 \n",
      "\t   BCLAF1, AL031772.1, MTFR2, LINC00271, TXLNB, AHI1, AL158850.1, MYB, CITED2, HBS1L \n",
      "PC_ 4 \n",
      "Positive:  BLOC1S5-TXNDC5, PIP5K1P1, BLOC1S5, EEF1E1-BLOC1S5, EEF1E1, SLC35B3, HULC, AL591475.1, AL136226.1, TFAP2A-AS1 \n",
      "\t   GCNT2, C6orf52, PAK1IP1, TMEM14C, AL024498.1, TMEM14B, SMIM13, NEDD9, TMEM170B, ADTRP \n",
      "\t   AL157373.2, HIVEP1, EDN1, RPL15P3, PHACTR1, AL008729.1, TBC1D7, AL008729.2, GFOD1, RPS4XP7 \n",
      "Negative:  MUC12, MUC3A, ACHE, TRIM56, SRRT, PILRB, STAG3L5P-PVRIG2P-PILRB, TRIP6, PILRA, STAG3L5P \n",
      "\t   SLC12A9, MIR4653, FAM102A, ZCWPW1, POP7, MEPCE, GIGYF1, TSC22D4, PPP1R35, MOSPD3 \n",
      "\t   TFR2, GNB2, AGFG2, LRCH4, PLOD3, DPM2, ZNHIT1, AC004453.1, NUDCD3, POLD2 \n",
      "PC_ 5 \n",
      "Positive:  AC015883.1, AC091181.1, MSI2, RN7SL449P, AKAP1, VEZF1, AC015813.1, AC007114.1, SRSF1, SCPEP1 \n",
      "\t   DYNLL2, AC015813.6, AC004584.2, MKS1, COIL, MPO, AC015912.3, TSPOAP1, TRIM25, TSPOAP1-AS1 \n",
      "\t   AC015912.2, AC004687.1, RNU6-1158P, DGKE, MIR142, SUPT4H1, C17orf67, RNF43, SEPT4-AS1, ANKFN1 \n",
      "Negative:  AC074397.1, COBL, FIGNL1, HAUS6P1, IKZF1, SLC25A5P3, UPP1, C7orf69, SEC61G, LINC00525 \n",
      "\t   HUS1, CALM1P2, LANCL2, TRAF3, AMN, CDC42BPB, RPL13P6, VOPP1, EXOC3L4, TNFAIP2 \n",
      "\t   SUMO2P3, EIF5, PSPHP1, SNORA28, NIPSNAP2, MARK3, PSPH, RPL10AP1, IMMTP1, TSPEAR-AS2 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 16:09:15] define_signif_tumor_subclusters(), tumor: Malignant_MM4_Prolif\n",
      "INFO [2025-10-26 16:09:15] Setting auto leiden resolution for Malignant_MM4_Prolif to 0.139357\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  PIGT, MIR6812, WFDC2, DBNDD2, RPS2P7, SYS1, STK4, DNTTIP1, TOMM34, UBE2C \n",
      "\t   TNNC2, ACOT8, PABPC1L, ZSWIM3, ZSWIM1, YWHAB, NEURL2, CTSA, KCNK15-AS1, PLTP \n",
      "\t   PCIF1, LINC01260, ZNF335, ADA, MMP9, PKIG, PPIAP3, CTNNBL1, BLCAP, RN7SKP185 \n",
      "Negative:  SNX10, AC010677.1, CBX3, AC004540.1, BET1, AC010719.1, AC005165.2, HNRNPA2B1, AC006378.2, COL1A2 \n",
      "\t   NFE2L3, AC004520.1, GNG11, C7orf71, CASD1, C7orf31, SGCE, AC002075.2, VPS50, SKAP2 \n",
      "\t   CYCS, HEPACAM2, PEG10, SAMD9L, SAMD9, RPS3AP25, OSBPL3, PPP1R9A, CDK6, HMGB3P20 \n",
      "PC_ 2 \n",
      "Positive:  SPARC, AC011374.2, SLC36A1, ATOX1, GM2A, CCDC69, AC091982.1, AC091982.4, ANXA6, G3BP1 \n",
      "\t   TNIP1, FAM114A2, GPX3, ZNF300, SMIM3, DCTN4, RBM22, SYNPO, BTF3L4, TXNDC12 \n",
      "\t   ZFYVE9, RAB3B, NDST1, AL139156.1, AL050343.1, NRDC, ANAPC10P1, SNORA26, CC2D1B, PLA2G12AP1 \n",
      "Negative:  PCP4, BACE2, MYL6P2, POMP, FAM3B, SLC46A3, SH3BGR, EIF4A1P7, CYP51A1P2, GAPDHP69 \n",
      "\t   LCA5L, SLC7A1, MX2, WRB, UBL3, MX1, PRDM15, HMGN1, KATNAL1, AP001619.1 \n",
      "\t   C2CD2, PRDX2P1, ZBTB21, H2AFZP3, LINC00426, BRWD1-AS2, UBAC2, ABCG1, UBAC2-AS1, GPR183 \n",
      "PC_ 3 \n",
      "Positive:  PCDHB10, PCDHB9, PCDHB11, PCDHB13, PCDHB14, PCDHB16, TAF7, PCDHB2, AC244517.9, AC005618.1 \n",
      "\t   VTRNA1-3, PCDHGA1, VTRNA1-1, PCDHGA5, ZMAT2, PCDHGA6, HARS2, HARS, PCDHGA8, DND1 \n",
      "\t   PCDHGB6, WDR55, PCDHGA10, MIR3655, PCDHGB7, IK, PCDHGC5, TMCO6, DIAPH1, NDUFA2 \n",
      "Negative:  AC017002.1, ANAPC1, EEF1E1P1, MERTK, AC104651.1, TMEM87B, ZC3H8, ZC3H6, RGPD8, TTL \n",
      "\t   POLR1B, AC079922.1, CHCHD5, AC079922.2, SLC20A1, CKAP2L, XIAPP3, AC016724.1, IL1RN, PSD4 \n",
      "\t   PAX8-AS1, PAX8, AC016745.2, CBWD2, WASH2P, RPL23AP7, RABL2A, AL078621.1, SLC35F5, AC104653.1 \n",
      "PC_ 4 \n",
      "Positive:  LRRC3, TRPM2-AS, TRPM2, TSPEAR-AS2, AP001062.1, C21orf2, PFKL, IMMTP1, ICOSLG, UBE2G2 \n",
      "\t   SUMO3, C21orf33, PTTG1IP, ITGB2, LINC01547, FAM207A, PWP2, AP001505.2, H2AFZP1, SSR4P1 \n",
      "\t   ADARB1, TRAPPC10, POFUT2, AGPAT3, BX322557.1, MYL6P1, RRP1, LINC00205, CSTB, COL18A1 \n",
      "Negative:  SLC15A2, EAF2, IQCB1, ILDR1, CD86, CCDC58, HLA-L, HCG17, AL662795.2, TRIM26 \n",
      "\t   TRIM31, HCG18, FAM162A, PPP1R11, TRIM39, ZNRD1, TRIM39-RPP21, ETF1P1, WDR5B, HLA-J \n",
      "\t   RPP21, AL671277.2, ZNRD1ASP, SUCLA2P1, AC083798.2, HCG9, RANP1, HLA-A, HLA-E, KPNA1 \n",
      "PC_ 5 \n",
      "Positive:  CES3, CES4A, CBFB, C16orf70, CES2, B3GNT9, FAM96B, PDP2, TRADD, NAE1 \n",
      "\t   DYNC1LI2, FBXL8, CMTM4, AC074143.1, CMTM3, CMTM1, CKLF-CMTM1, CKLF, AC010542.4, AC010542.2 \n",
      "\t   TK2, LINC00920, CDH8, RPS27AP16, AC009094.1, APOOP5, RPS27P27, GOT2, SLC38A7, AC009118.3 \n",
      "Negative:  SNORA58, NDC1, AC119428.2, LINC01771, LRP8, MAGOH, C1orf123, CPT2, AL445183.3, SCP2 \n",
      "\t   AC099677.1, ECHDC2, ZYG11A, RPS13P2, ZYG11B, NDUFS5P3, COA7, FAM159A, GPX7, ZCCHC11 \n",
      "\t   PRPF38A, DPH3, IMPDH1P8, OXNAD1, MIR563, RFTN1, ORC1, RN7SL4P, PLCL2, ANKRD28 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 16:09:16] define_signif_tumor_subclusters(), tumor: Malignant_MM5_Non-Prolif\n",
      "INFO [2025-10-26 16:09:16] Setting auto leiden resolution for Malignant_MM5_Non-Prolif to 0.103125\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  IL16, AC103858.1, STARD5, TMC3-AS1, TLNRD1, AC109809.1, AC068870.2, AC023034.1, AC068870.1, MESD \n",
      "\t   AC104041.1, AC027808.2, ABHD17C, AC026956.1, ARNT2, MEX3B, AC016705.2, AC105036.1, EFL1, SNUPN \n",
      "\t   FAH, AC105020.2, SAXO2, MARCH6, IMP3, AC012640.2, RPL30P7, ANKRD33B, CMBL, AC105020.6 \n",
      "Negative:  AL161787.1, DAPK1-IT1, CTSL, RPS10P3, AL353726.1, DAPK1, AL353752.1, SPIN1, CKS2, CDC20P1 \n",
      "\t   GAS1RR, SECISBP2, GAS1, AL353678.1, SEMA4D, ZCCHC6, ISCA1, C9orf153, PA2G4P6, GOLM1 \n",
      "\t   NAA35, DIRAS2, AL353743.3, PHBP7, SYK, AL353743.1, AGTPBP1, NTRK2, AL355607.1, RMI1 \n",
      "PC_ 2 \n",
      "Positive:  AC009413.1, LDHAP3, PKDCC, SLC8A1, AC007253.1, EML4, COX7A2L, AC025750.2, MTA3, HAAO \n",
      "\t   AC098824.1, ZFP36L2, LINC01126, GPATCH2L, AC016526.3, THADA, TGFB3, IFT43, AC010883.1, NEK9 \n",
      "\t   DYNC2LI1, ZC2HC1C, ACYP1, MLH3, LRPPRC, EIF2B2, RPS6KL1, AC019129.1, DLST, AC006530.1 \n",
      "Negative:  AL359183.1, SLC29A3, PCBD1, CDH23, SGPL1, VSIR, PRF1, PALD1, PSAP, EIF4EBP2 \n",
      "\t   CEP57L1P1, SPOCK2, LRRC20, PPA1, ASCC1, CALM2P2, RPL15P14, SAR1A, ANAPC16, TYSND1 \n",
      "\t   CDIP1, SUB1P3, HMOX2, UBALD1, AIFM2, MGRN1, NMRAL1, NUDT16L1, DDIT4, ANKS3 \n",
      "PC_ 3 \n",
      "Positive:  LRIG1, SUCLG2, KBTBD8, PRKAR2A, PRKAR2A-AS1, SLC25A20, ARIH2OS, ARIH2, AC137630.4, P4HTM \n",
      "\t   WDR6, DALRD3, MIR425, NDUFAF3, IMPDH2, AC137630.3, QRICH1, QARS, USP19, CCDC71 \n",
      "\t   KLHDC8B, TRMT12, RNF139-AS1, TMEM65, RNF139, FAM91A1, TATDN1, AC090198.1, MTSS1, ZNF572 \n",
      "Negative:  CHD4, NOP2, LPAR5, IFFO1, ACRBP, AC005840.2, GAPDH, CD27-AS1, CD27, AC005840.1 \n",
      "\t   TAPBPL, ING4, NCAPD2, VAMP1, ZNF384, COPS7A, MLF2, PTMS, LAG3, CD4 \n",
      "\t   P3H3, GNB3, CDCA3, USP5, C6orf99, RSPH3, TCP1, SNORA29, IGF2R, AL353625.1 \n",
      "PC_ 4 \n",
      "Positive:  AL357079.1, AL451062.2, KDM4A, IPO13, ST3GAL3, DPH2, KDM4A-AS1, PTPRF, ATP6V0B, B4GALT2 \n",
      "\t   HYI, SLC6A9, SZT2, CCDC24, AL139220.2, MED8, AL356653.1, ELOVL1, DMAP1, ERI3 \n",
      "\t   CDC20, RNF220, AL139289.2, TMEM53, RNU5F-1, TMEM125, RNU5D-1, KIF2C, RPS8, MIR6733 \n",
      "Negative:  GTF2H2, GUSBP9, NAIP, BDP1, SMN1, MCCC2, SERF1A, MAP1B, AC139272.1, PTCD2 \n",
      "\t   GTF2H2B, RPL7P22, TNPO1, AC138866.2, AC008953.1, SMN2, FCHO2, SERF1B, AC020893.2, TMEM171 \n",
      "\t   AC131392.1, AC116345.1, GTF2H2C, AC099522.1, OCLN, AC099522.2, BTF3, AC145146.1, MARVELD2, RAD17 \n",
      "PC_ 5 \n",
      "Positive:  PPIH, YBX1, P3H1, C1orf50, SVBP, ERMAP, AL512353.1, ZNF691, ATP6V1E1P1, SLC2A1 \n",
      "\t   EBNA1BP2, MIR6733, TMEM125, AL139289.2, CDC20, ELOVL1, LINC01144, MED8, OSTCP5, SZT2 \n",
      "\t   ZSWIM5, HYI, UROD, PTPRF, HECTD3, KDM4A, EIF2B3, KDM4A-AS1, AL136380.1, ST3GAL3 \n",
      "Negative:  TTC38, CDPF1, FP325332.1, PPARA, MIRLET7BHG, PRR34-AS1, AL121672.3, LINC00899, BX537318.1, ATXN10 \n",
      "\t   Z95331.1, RIBC2, SMC1B, AC069181.1, RBM8B, SIX1, MAVS, AP5S1, CDC25B, SIX4 \n",
      "\t   AL391834.2, AL158206.1, AL158077.1, RPS6, CENPB, DUSP4, AL391834.1, MLLT3, MNAT1, MIR491 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 16:09:17] define_signif_tumor_subclusters(), tumor: Malignant_MM5_Prolif\n",
      "INFO [2025-10-26 16:09:17] Setting auto leiden resolution for Malignant_MM5_Prolif to 0.145685\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  CDKN2A, AL359922.3, MTAP, KLHL9, CDKN2B-AS1, HACD4, MIR491, AL449423.1, FOCAD, MLLT3 \n",
      "\t   CDKN2B, UBA52P6, AL158077.1, DMRTA1, AC091167.1, GDPGP1, CIB1, AL158206.1, SEMA4B, IDH2 \n",
      "\t   LINC01239, NGRN, ZNF710-AS1, AC087284.1, AC091167.2, CLIC4P1, AL391834.2, ZNF710, AC091167.5, ELAVL2 \n",
      "Negative:  MFSD5, ESPL1, PFDN5, RARG, AC073611.1, ITGB7, C12orf10, AAAS, ZNF740, SP1 \n",
      "\t   PRR13, CSAD, PCBP2, EIF4A1P4, MAP3K12, AC023509.4, HIGD1AP1, TARBP2, IGFBP6, USB1 \n",
      "\t   ATF7, SPRYD3, CFAP20, AC073594.1, AC068888.1, ATP5G2, CSNK2A2, AC068888.2, CALCOCO1, CCDC113 \n",
      "PC_ 2 \n",
      "Positive:  PRKRIP1, AC091390.2, SH2B2, AC005072.1, AC091390.4, AC073517.1, CUX1, IFT22, AC006329.1, ORAI2 \n",
      "\t   FIS1, CLDN15, ZNHIT1, ALKBH4, PLOD3, MIR4653, POLR2J, TRIM56, MIR4467, MUC12 \n",
      "\t   MUC3A, ACHE, SRRT, TRIP6, POLR2J3, SLC12A9, POLR2J2, POP7, GIGYF1, FAM185A \n",
      "Negative:  AP4B1, AL137856.1, DCLRE1B, PTPN22, HIPK1-AS1, RSBN1, HIPK1, AL133517.1, TRIM33, PHTF1 \n",
      "\t   AL035410.1, AL391058.1, PKMP1, BCAS2, MAGI3, DENND2C, AL357055.2, AMPD1, AL357055.1, NRAS \n",
      "\t   LRIG2, CSDE1, SLC16A1-AS1, AKR7A2P1, SIKE1, SLC16A1, TSPAN2, NUTF2P4, CNOT7P2, AL603832.3 \n",
      "PC_ 3 \n",
      "Positive:  DPAGT1, C2CD2L, HINFP, NLRX1, CCDC153, CBL, MCAM, RNF26, USP2, NECTIN1 \n",
      "\t   OAF, CEP164, TMEM136, BACE1-AS, ARHGEF12, GRIK4, RNF214, TBCEL, PCSK7, TECTA \n",
      "\t   AC010627.1, CCT6P2, SC5D, FAM105A, EEF1A1P13, AC010491.1, TAGLN, ANKH, OTULIN, RBBP4P1 \n",
      "Negative:  DDX1, AC007738.1, NBAS, AC016730.1, AC073062.1, TRIB2, MIR3681HG, LPIN1, GREB1, E2F6 \n",
      "\t   ROCK2, PQLC3, PDIA6, AC092687.3, ATP6V1C2, RN7SL832P, NOL10, ODC1, HPCAL1, C2orf48 \n",
      "\t   AC007240.1, RRM2, KLF11, AC104794.2, GRHL1, AC010969.2, SF3B4, SV2A, BOLA1, OTUD7B \n",
      "PC_ 4 \n",
      "Positive:  AC080078.3, AC080078.2, LINC02273, FBXW7, AC097375.1, MIR3140, AC107074.1, LINC02486, MIR4453, AC023424.2 \n",
      "\t   AC023424.1, PSMD5-AS1, PHF19, PSMD5, AL161911.1, TRAF1, FBXW2, WHRN, AHCYP2, AKNA \n",
      "\t   C5, RGS3, AL138895.2, C9orf43, SNORA70C, CDK5RAP2, POLE3, MEGF9, TLR4, TRIM32 \n",
      "Negative:  AC026412.1, NDUFS6, AC025183.1, AC116359.1, ICE1, MED10, UBE2QL1, NSUN2, SRD5A1, PAPD7 \n",
      "\t   MTRR, FASTKD3, CPSF3, MIR4458HG, IAH1, SEMA5A, ADAM17, SNHG18, YWHAQ, SNORD123 \n",
      "\t   TAF1B, FAM173B, LIX1L-AS1, GNRHR2, LIX1L, ANKRD34A, POLR3GL, RBM8A, TXNIP, PEX11B \n",
      "PC_ 5 \n",
      "Positive:  PNRC2P1, SNORA40, VDAC2P3, FAM46C, MAN1A2, AL358072.1, TRIM45, TTF2, AL445231.1, CD2 \n",
      "\t   NAP1L4P1, CD58, RNU6-817P, ATP1A1-AS1, ATP1A1, U3, SLC22A15, VANGL1, CNOT7P2, TSPAN2 \n",
      "\t   RNVU1-19, SIKE1, LINC00623, CSDE1, NRAS, RNVU1-4, AMPD1, SRGAP2C, DENND2C, AL592494.2 \n",
      "Negative:  AL159163.1, FAM103A2P, RPS6KA2, AL022069.1, MPC1, SFT2D1, GNG5P1, QKI, CAHM, AL590286.1 \n",
      "\t   PRKN, AGPAT4, MAP3K4, SLC22A1, AL353625.1, IGF2R, SNORA29, TCP1, VKORC1, AC135050.2 \n",
      "\t   DYNC1LI2, CMTM4, NAE1, RSPH3, CMTM3, PDP2, FAM96B, CMTM1, C6orf99, CES2 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 16:09:18] define_signif_tumor_subclusters(), tumor: Normal_Pb\n",
      "INFO [2025-10-26 16:09:18] Less cells in group Normal_Pb than k_nn setting. Keeping as a single subcluster.\n",
      "INFO [2025-10-26 16:09:18] define_signif_tumor_subclusters(), tumor: Normal_PC\n",
      "INFO [2025-10-26 16:09:18] Setting auto leiden resolution for Normal_PC to 0.124265\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  GAPDHP35, NPIPB6, EIF3CL, CLN3, APOBR, AC020765.2, SGF29, SULT1A2, SULT1A1, EIF3C \n",
      "\t   AC145285.2, AC145285.1, VARS2, TUBB, GTF2H4, FLOT1, DDR1, MDC1, IER3, AL662797.2 \n",
      "\t   SFTA2, NRM, SCD, PHBP9, HCG21, BLOC1S2, CWF19L1, PPP1R18, CHUK, MUC21 \n",
      "Negative:  AC005019.1, AC011287.1, ETV1, ARL4A, SCIN, AGMO, AC013470.3, AC013470.2, AC005550.2, VWDE \n",
      "\t   ANKMY2, TMEM106B, BZW2, AC073109.1, TSPAN13, AGR2, THSD7A, RAD17P1, AC003075.1, PHF14 \n",
      "\t   AHR, SNX13, AC007029.1, AC080080.1, HDAC9, NDUFA4, TWISTNB, RPL21P75, AC009945.1, MACC1 \n",
      "PC_ 2 \n",
      "Positive:  AC118553.1, SPCS2P1, AC068724.2, CCNDBP1, AC009852.2, LCMT2, RTCA-AS1, ADAL, ZSCAN29, TUBGCP4 \n",
      "\t   TP53BP1, CDC14A, MAP1A, PPIP5K1, BCAS2P2, CATSPER2, RPS28P7, EXTL2, AC011330.2, ARL6IP1P3 \n",
      "\t   GLO1, DNAH8, AP006295.1, BTBD9, SAYSD1, SLC30A7, AL589655.1, KIF6, AP001284.1, ZFAND3 \n",
      "Negative:  AL049543.1, NOP56P1, AL662890.1, TRIM27, ZNF311, AL662791.2, AL645927.1, CD4, LAG3, P3H3 \n",
      "\t   GABBR1, PTMS, GNB3, CDCA3, MLF2, RNU5D-1, KIF2C, ACRBP, COPS7A, ING4 \n",
      "\t   ZNF384, LPAR5, RPS8, USP5, SUMO2P1, CHD4, SNORD55, NCAPD2, GAPDH, NOP2 \n",
      "PC_ 3 \n",
      "Positive:  RALGPS2, AL449106.1, AL137796.1, C1orf220, AL365357.1, RASAL2, SEC16B, FBXO28, AC138393.1, MIR320B2 \n",
      "\t   ASTN1, CNIH4, NVL, AC092809.3, DEGS1, GTF2IP20, WDR26, ACTBP11, CNIH3, AL359265.1 \n",
      "\t   PHBP11, AKR1B1P1, TP53BP2, CAPN2, DNAH14, U7, DISP1, LBR, AL392172.1, ENAH \n",
      "Negative:  AC104046.1, MIR7706, RNU7-79P, AC087286.4, AKAP13, RNU6-185P, ADAMTS7P4, LINC00052, NTRK3, PDE8A \n",
      "\t   ALPK3, AC013489.1, ZNF592, DET1, SEC11A, NMB, AEN, WDR73, EGLN1P1, ISG20 \n",
      "\t   SCAND2P, ZSCAN2, MFGE8, UBE2Q2P1, AC107954.1, ADAMTSL3, HDGFL3, ABHD2, TM6SF1, HMGB1P8 \n",
      "PC_ 4 \n",
      "Positive:  FAM174B, NPM1P5, HMGN1P38, SLCO3A1, AC091544.3, SV2B, VPS33B, PRC1, RCCD1, UNC45A \n",
      "\t   HDDC3, MAN2A2, FES, FURIN, DYNC1I2, SLC25A12, CYBRD1, AC007969.1, AC124248.1, HAT1 \n",
      "\t   AC007969.2, METAP1D, BLM, DAP3P2, DCAF17, CRTC3-AS1, AC008065.1, CRTC3, METTL8, IQGAP1 \n",
      "Negative:  LIN9, PARP1, AL359704.1, ENAH, ITPKB, LBR, PSEN2, DNAH14, AKR1B1P1, COQ8A \n",
      "\t   CNIH3, CDC42BPA, WDR26, CNIH4, AL627308.2, MIR320B2, NUCKS1P1, NVL, AL606462.1, DEGS1 \n",
      "\t   BTF3P9, AC092809.3, ABCB8, ATG9B, NOS3, ZNF678, FBXO28, JMJD4, TMEM176A, AC138393.1 \n",
      "PC_ 5 \n",
      "Positive:  FAIM, CEP70, AC020890.1, IQGAP1, PIK3CB, CRTC3, MRAS, ZNF774, EEF1A1P25, RN7SL736P \n",
      "\t   CRTC3-AS1, ARMC8, AC091167.5, COPB2, AC091167.2, BLM, DBR1, NGRN, AC097103.2, AC124248.1 \n",
      "\t   GDPGP1, FURIN, IL20RB, CIB1, AC097103.1, AC091167.1, SEMA4B, FES, IDH2, ZNF710-AS1 \n",
      "Negative:  WNT11, THAP12, AP003168.2, AP002360.1, UVRAG, AP002360.4, DGAT2, AP002360.2, EMSY, SERPINH1 \n",
      "\t   ACER3, CAPN5, GDPD5, TOMM20P1, KLHL35, PAK1, CLNS1A, RPS3, AQP11, NEU3 \n",
      "\t   RSF1, SPCS2, FTH1P16, AP001992.1, AAMDC, RN7SL239P, INTS4, AP000560.1, XRRA1, NDUFC2-KCTD14 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-10-26 16:09:19] -mirroring for hspike\n",
      "INFO [2025-10-26 16:09:19] define_signif_tumor_subclusters(p_val=0.1\n",
      "INFO [2025-10-26 16:09:19] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_Normal_PC\n",
      "INFO [2025-10-26 16:09:19] cut tree into: 1 groups\n",
      "INFO [2025-10-26 16:09:19] -processing spike_tumor_cell_Normal_PC,spike_tumor_cell_Normal_PC_s1\n",
      "INFO [2025-10-26 16:09:19] define_signif_tumor_subclusters(), tumor: simnorm_cell_Normal_PC\n",
      "INFO [2025-10-26 16:09:19] cut tree into: 1 groups\n",
      "INFO [2025-10-26 16:09:19] -processing simnorm_cell_Normal_PC,simnorm_cell_Normal_PC_s1\n",
      "INFO [2025-10-26 16:09:41] ::plot_cnv:Start\n",
      "INFO [2025-10-26 16:09:41] ::plot_cnv:Current data dimensions (r,c)=18168,1460 Total=26640208.5862418 Min=0.506573411933416 Max=2.13705129387919.\n",
      "INFO [2025-10-26 16:09:41] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2025-10-26 16:09:42] plot_cnv(): auto thresholding at: (0.691167 , 1.317499)\n",
      "INFO [2025-10-26 16:09:42] plot_cnv_observation:Start\n",
      "INFO [2025-10-26 16:09:42] Observation data size: Cells= 1324 Genes= 18168\n",
      "INFO [2025-10-26 16:09:42] clustering observations via method: ward.D\n",
      "INFO [2025-10-26 16:09:42] Number of cells in group(1) is 22\n",
      "INFO [2025-10-26 16:09:42] group size being clustered:  22,18168\n",
      "INFO [2025-10-26 16:09:42] Number of cells in group(2) is 61\n",
      "INFO [2025-10-26 16:09:42] group size being clustered:  61,18168\n",
      "INFO [2025-10-26 16:09:42] Number of cells in group(3) is 37\n",
      "INFO [2025-10-26 16:09:42] group size being clustered:  37,18168\n",
      "INFO [2025-10-26 16:09:42] Number of cells in group(4) is 37\n",
      "INFO [2025-10-26 16:09:42] group size being clustered:  37,18168\n",
      "INFO [2025-10-26 16:09:42] Number of cells in group(5) is 34\n",
      "INFO [2025-10-26 16:09:42] group size being clustered:  34,18168\n",
      "INFO [2025-10-26 16:09:42] Number of cells in group(6) is 26\n",
      "INFO [2025-10-26 16:09:42] group size being clustered:  26,18168\n",
      "INFO [2025-10-26 16:09:42] Number of cells in group(7) is 23\n",
      "INFO [2025-10-26 16:09:42] group size being clustered:  23,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(8) is 22\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  22,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(9) is 17\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  17,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(10) is 16\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  16,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(11) is 36\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  36,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(12) is 25\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  25,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(13) is 39\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  39,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(14) is 25\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  25,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(15) is 22\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  22,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(16) is 21\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  21,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(17) is 21\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  21,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(18) is 18\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  18,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(19) is 17\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  17,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(20) is 17\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  17,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(21) is 14\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  14,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(22) is 36\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  36,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(23) is 36\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  36,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(24) is 27\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  27,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(25) is 20\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  20,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(26) is 36\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  36,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(27) is 27\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  27,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(28) is 19\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  19,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(29) is 14\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  14,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(30) is 5\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  5,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(31) is 38\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  38,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(32) is 28\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  28,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(33) is 26\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  26,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(34) is 24\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  24,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(35) is 16\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  16,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(36) is 16\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  16,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(37) is 40\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  40,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(38) is 35\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  35,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(39) is 32\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  32,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(40) is 12\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  12,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(41) is 48\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  48,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(42) is 43\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  43,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(43) is 31\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  31,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(44) is 27\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  27,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(45) is 18\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  18,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(46) is 2\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  2,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(47) is 35\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  35,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(48) is 35\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  35,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(49) is 29\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  29,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(50) is 14\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  14,18168\n",
      "INFO [2025-10-26 16:09:43] Number of cells in group(51) is 5\n",
      "INFO [2025-10-26 16:09:43] group size being clustered:  5,18168\n",
      "INFO [2025-10-26 16:09:43] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2025-10-26 16:09:43] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2025-10-26 16:09:43] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 16:09:43] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 16:09:46] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 16:09:46] Quantiles of plotted data range: 0.691167043801423,0.933755536876456,0.998709494487031,1.06883885885783,1.31749854511917\n",
      "INFO [2025-10-26 16:09:48] plot_cnv_references:Start\n",
      "INFO [2025-10-26 16:09:48] Reference data size: Cells= 136 Genes= 18168\n",
      "INFO [2025-10-26 16:09:48] plot_cnv_references:Number reference groups= 5\n",
      "INFO [2025-10-26 16:09:48] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2025-10-26 16:09:48] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 16:09:48] Quantiles of plotted data range: 0.691167043801423,0.952178263344931,0.998725152982485,1.04861125595042,1.31749854511917\n",
      "INFO [2025-10-26 16:10:11] ::plot_cnv:Start\n",
      "INFO [2025-10-26 16:10:11] ::plot_cnv:Current data dimensions (r,c)=18168,1460 Total=26640208.5862418 Min=0.506573411933416 Max=2.13705129387919.\n",
      "INFO [2025-10-26 16:10:11] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2025-10-26 16:10:12] plot_cnv(): auto thresholding at: (0.691167 , 1.317499)\n",
      "INFO [2025-10-26 16:10:12] plot_cnv_observation:Start\n",
      "INFO [2025-10-26 16:10:12] Observation data size: Cells= 1324 Genes= 18168\n",
      "INFO [2025-10-26 16:10:13] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2025-10-26 16:10:13] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2025-10-26 16:10:13] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 16:10:13] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 16:10:15] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 16:10:15] Quantiles of plotted data range: 0.691167043801423,0.933755536876456,0.998709494487031,1.06883885885783,1.31749854511917\n",
      "INFO [2025-10-26 16:10:18] plot_cnv_references:Start\n",
      "INFO [2025-10-26 16:10:18] Reference data size: Cells= 136 Genes= 18168\n",
      "INFO [2025-10-26 16:10:18] plot_cnv_references:Number reference groups= 1\n",
      "INFO [2025-10-26 16:10:18] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2025-10-26 16:10:18] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 16:10:18] Quantiles of plotted data range: 0.691167043801423,0.952178263344931,0.998725152982485,1.04861125595042,1.31749854511917\n",
      "INFO [2025-10-26 16:10:19] \n",
      "\n",
      "\tSTEP 17: HMM-based CNV prediction\n",
      "\n",
      "INFO [2025-10-26 16:10:19] predict_CNV_via_HMM_on_tumor_subclusters\n",
      "INFO [2025-10-26 16:10:52] -done predicting CNV based on initial tumor subclusters\n",
      "INFO [2025-10-26 16:10:52] get_predicted_CNV_regions(subcluster)\n",
      "INFO [2025-10-26 16:10:52] -processing cell_group_name: Interm.Interm_s1, size: 22\n",
      "INFO [2025-10-26 16:11:02] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s2, size: 61\n",
      "INFO [2025-10-26 16:11:12] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s5, size: 37\n",
      "INFO [2025-10-26 16:11:22] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s6, size: 37\n",
      "INFO [2025-10-26 16:11:32] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s1, size: 34\n",
      "INFO [2025-10-26 16:11:43] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s8, size: 26\n",
      "INFO [2025-10-26 16:11:53] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s3, size: 23\n",
      "INFO [2025-10-26 16:12:02] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s7, size: 22\n",
      "INFO [2025-10-26 16:12:12] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s4, size: 17\n",
      "INFO [2025-10-26 16:12:22] -processing cell_group_name: Malignant_MM1_Non-Prolif.Malignant_MM1_Non-Prolif_s9, size: 16\n",
      "INFO [2025-10-26 16:12:32] -processing cell_group_name: Malignant_MM1_Prolif.Malignant_MM1_Prolif_s2, size: 36\n",
      "INFO [2025-10-26 16:12:43] -processing cell_group_name: Malignant_MM1_Prolif.Malignant_MM1_Prolif_s1, size: 25\n",
      "INFO [2025-10-26 16:12:52] -processing cell_group_name: Malignant_MM2_Non-Prolif.Malignant_MM2_Non-Prolif_s2, size: 39\n",
      "INFO [2025-10-26 16:13:02] -processing cell_group_name: Malignant_MM2_Non-Prolif.Malignant_MM2_Non-Prolif_s6, size: 25\n",
      "INFO [2025-10-26 16:13:12] -processing cell_group_name: Malignant_MM2_Non-Prolif.Malignant_MM2_Non-Prolif_s7, size: 22\n",
      "INFO [2025-10-26 16:13:22] -processing cell_group_name: Malignant_MM2_Non-Prolif.Malignant_MM2_Non-Prolif_s3, size: 21\n",
      "INFO [2025-10-26 16:13:32] -processing cell_group_name: Malignant_MM2_Non-Prolif.Malignant_MM2_Non-Prolif_s9, size: 21\n",
      "INFO [2025-10-26 16:13:42] -processing cell_group_name: Malignant_MM2_Non-Prolif.Malignant_MM2_Non-Prolif_s5, size: 18\n",
      "INFO [2025-10-26 16:13:53] -processing cell_group_name: Malignant_MM2_Non-Prolif.Malignant_MM2_Non-Prolif_s1, size: 17\n",
      "INFO [2025-10-26 16:14:02] -processing cell_group_name: Malignant_MM2_Non-Prolif.Malignant_MM2_Non-Prolif_s8, size: 17\n",
      "INFO [2025-10-26 16:14:12] -processing cell_group_name: Malignant_MM2_Non-Prolif.Malignant_MM2_Non-Prolif_s4, size: 14\n",
      "INFO [2025-10-26 16:14:22] -processing cell_group_name: Malignant_MM2_Prolif.Malignant_MM2_Prolif_s3, size: 36\n",
      "INFO [2025-10-26 16:14:32] -processing cell_group_name: Malignant_MM2_Prolif.Malignant_MM2_Prolif_s4, size: 36\n",
      "INFO [2025-10-26 16:14:42] -processing cell_group_name: Malignant_MM2_Prolif.Malignant_MM2_Prolif_s1, size: 27\n",
      "INFO [2025-10-26 16:14:53] -processing cell_group_name: Malignant_MM2_Prolif.Malignant_MM2_Prolif_s2, size: 20\n",
      "INFO [2025-10-26 16:15:03] -processing cell_group_name: Malignant_MM3_Non-Prolif.Malignant_MM3_Non-Prolif_s3, size: 36\n",
      "INFO [2025-10-26 16:15:13] -processing cell_group_name: Malignant_MM3_Non-Prolif.Malignant_MM3_Non-Prolif_s1, size: 27\n",
      "INFO [2025-10-26 16:15:23] -processing cell_group_name: Malignant_MM3_Non-Prolif.Malignant_MM3_Non-Prolif_s2, size: 19\n",
      "INFO [2025-10-26 16:15:33] -processing cell_group_name: Malignant_MM3_Non-Prolif.Malignant_MM3_Non-Prolif_s4, size: 14\n",
      "INFO [2025-10-26 16:15:42] -processing cell_group_name: Malignant_MM3_Prolif.Malignant_MM3_Prolif, size: 5\n",
      "INFO [2025-10-26 16:15:52] -processing cell_group_name: Malignant_MM4_Non-Prolif.Malignant_MM4_Non-Prolif_s4, size: 38\n",
      "INFO [2025-10-26 16:16:03] -processing cell_group_name: Malignant_MM4_Non-Prolif.Malignant_MM4_Non-Prolif_s5, size: 28\n",
      "INFO [2025-10-26 16:16:12] -processing cell_group_name: Malignant_MM4_Non-Prolif.Malignant_MM4_Non-Prolif_s1, size: 26\n",
      "INFO [2025-10-26 16:16:22] -processing cell_group_name: Malignant_MM4_Non-Prolif.Malignant_MM4_Non-Prolif_s3, size: 24\n",
      "INFO [2025-10-26 16:16:32] -processing cell_group_name: Malignant_MM4_Non-Prolif.Malignant_MM4_Non-Prolif_s2, size: 16\n",
      "INFO [2025-10-26 16:16:42] -processing cell_group_name: Malignant_MM4_Non-Prolif.Malignant_MM4_Non-Prolif_s6, size: 16\n",
      "INFO [2025-10-26 16:16:52] -processing cell_group_name: Malignant_MM4_Prolif.Malignant_MM4_Prolif_s3, size: 40\n",
      "INFO [2025-10-26 16:17:03] -processing cell_group_name: Malignant_MM4_Prolif.Malignant_MM4_Prolif_s1, size: 35\n",
      "INFO [2025-10-26 16:17:13] -processing cell_group_name: Malignant_MM4_Prolif.Malignant_MM4_Prolif_s2, size: 32\n",
      "INFO [2025-10-26 16:17:23] -processing cell_group_name: Malignant_MM4_Prolif.Malignant_MM4_Prolif_s4, size: 12\n",
      "INFO [2025-10-26 16:17:32] -processing cell_group_name: Malignant_MM5_Non-Prolif.Malignant_MM5_Non-Prolif_s3, size: 48\n",
      "INFO [2025-10-26 16:17:43] -processing cell_group_name: Malignant_MM5_Non-Prolif.Malignant_MM5_Non-Prolif_s1, size: 43\n",
      "INFO [2025-10-26 16:17:53] -processing cell_group_name: Malignant_MM5_Non-Prolif.Malignant_MM5_Non-Prolif_s2, size: 31\n",
      "INFO [2025-10-26 16:18:03] -processing cell_group_name: Malignant_MM5_Non-Prolif.Malignant_MM5_Non-Prolif_s5, size: 27\n",
      "INFO [2025-10-26 16:18:14] -processing cell_group_name: Malignant_MM5_Non-Prolif.Malignant_MM5_Non-Prolif_s6, size: 18\n",
      "INFO [2025-10-26 16:18:23] -processing cell_group_name: Malignant_MM5_Non-Prolif.Malignant_MM5_Non-Prolif_s4, size: 2\n",
      "INFO [2025-10-26 16:18:32] -processing cell_group_name: Malignant_MM5_Prolif.Malignant_MM5_Prolif_s1, size: 35\n",
      "INFO [2025-10-26 16:18:43] -processing cell_group_name: Malignant_MM5_Prolif.Malignant_MM5_Prolif_s2, size: 35\n",
      "INFO [2025-10-26 16:18:53] -processing cell_group_name: Malignant_MM5_Prolif.Malignant_MM5_Prolif_s3, size: 29\n",
      "INFO [2025-10-26 16:19:03] -processing cell_group_name: Malignant_MM5_Prolif.Malignant_MM5_Prolif_s4, size: 14\n",
      "INFO [2025-10-26 16:19:13] -processing cell_group_name: Normal_Pb.Normal_Pb, size: 5\n",
      "INFO [2025-10-26 16:19:22] -processing cell_group_name: Normal_PC.Normal_PC_s3, size: 58\n",
      "INFO [2025-10-26 16:19:33] -processing cell_group_name: Normal_PC.Normal_PC_s1, size: 43\n",
      "INFO [2025-10-26 16:19:44] -processing cell_group_name: Normal_PC.Normal_PC_s5, size: 22\n",
      "INFO [2025-10-26 16:19:54] -processing cell_group_name: Normal_PC.Normal_PC_s4, size: 8\n",
      "INFO [2025-10-26 16:20:04] -processing cell_group_name: Normal_PC.Normal_PC_s2, size: 5\n",
      "INFO [2025-10-26 16:20:14] -writing cell clusters file: /home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_output/output_infercnv_r_arch_prolif_SPID/17_HMM_predHMMi6.leiden.hmm_mode-subclusters.cell_groupings\n",
      "INFO [2025-10-26 16:20:14] -writing cnv regions file: /home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_output/output_infercnv_r_arch_prolif_SPID/17_HMM_predHMMi6.leiden.hmm_mode-subclusters.pred_cnv_regions.dat\n",
      "INFO [2025-10-26 16:20:16] -writing per-gene cnv report: /home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_output/output_infercnv_r_arch_prolif_SPID/17_HMM_predHMMi6.leiden.hmm_mode-subclusters.pred_cnv_genes.dat\n",
      "INFO [2025-10-26 16:20:16] -writing gene ordering info: /home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_output/output_infercnv_r_arch_prolif_SPID/17_HMM_predHMMi6.leiden.hmm_mode-subclusters.genes_used.dat\n",
      "INFO [2025-10-26 16:20:28] ::plot_cnv:Start\n",
      "INFO [2025-10-26 16:20:28] ::plot_cnv:Current data dimensions (r,c)=18168,1460 Total=80940824 Min=2 Max=6.\n",
      "INFO [2025-10-26 16:20:29] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2025-10-26 16:20:29] plot_cnv_observation:Start\n",
      "INFO [2025-10-26 16:20:29] Observation data size: Cells= 1324 Genes= 18168\n",
      "INFO [2025-10-26 16:20:29] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2025-10-26 16:20:29] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2025-10-26 16:20:30] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 16:20:30] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 16:20:33] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 16:20:33] Quantiles of plotted data range: 2,3,3,3,6\n",
      "INFO [2025-10-26 16:20:35] plot_cnv_references:Start\n",
      "INFO [2025-10-26 16:20:35] Reference data size: Cells= 136 Genes= 18168\n",
      "INFO [2025-10-26 16:20:35] plot_cnv_references:Number reference groups= 1\n",
      "INFO [2025-10-26 16:20:35] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2025-10-26 16:20:35] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 16:20:35] Quantiles of plotted data range: 2,3,3,3,4\n",
      "INFO [2025-10-26 16:20:37] \n",
      "\n",
      "\tSTEP 18: Run Bayesian Network Model on HMM predicted CNVs\n",
      "\n",
      "INFO [2025-10-26 16:20:37] Creating the following Directory:  /home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_output/output_infercnv_r_arch_prolif_SPID/BayesNetOutput.HMMi6.leiden.hmm_mode-subclusters\n",
      "INFO [2025-10-26 16:20:37] Initializing new MCM InferCNV Object.\n",
      "INFO [2025-10-26 16:20:37] validating infercnv_obj\n",
      "INFO [2025-10-26 16:20:37] Total CNV's:  1673\n",
      "INFO [2025-10-26 16:20:37] Loading BUGS Model.\n",
      "INFO [2025-10-26 16:20:39] Running Sampling Using Parallel with  4 Cores\n",
      "INFO [2025-10-26 16:49:26] Obtaining probabilities post-sampling\n",
      "INFO [2025-10-26 16:51:50] Gibbs sampling time:  31.1744776010513  Minutes\n",
      "INFO [2025-10-26 16:53:25] ::plot_cnv:Start\n",
      "INFO [2025-10-26 16:53:25] ::plot_cnv:Current data dimensions (r,c)=18168,1460 Total=4269120.10010831 Min=0 Max=0.981616515338111.\n",
      "INFO [2025-10-26 16:53:26] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2025-10-26 16:53:26] plot_cnv_observation:Start\n",
      "INFO [2025-10-26 16:53:26] Observation data size: Cells= 1324 Genes= 18168\n",
      "INFO [2025-10-26 16:53:26] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2025-10-26 16:53:26] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2025-10-26 16:53:26] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 16:53:26] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 16:53:29] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 16:53:29] Quantiles of plotted data range: 0,0,0,0,0.981616515338111\n",
      "INFO [2025-10-26 16:53:31] plot_cnv_references:Start\n",
      "INFO [2025-10-26 16:53:31] Reference data size: Cells= 136 Genes= 18168\n",
      "INFO [2025-10-26 16:53:31] plot_cnv_references:Number reference groups= 1\n",
      "INFO [2025-10-26 16:53:31] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2025-10-26 16:53:32] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 16:53:32] Quantiles of plotted data range: 0,0,0,0,0.854748054554991\n",
      "INFO [2025-10-26 16:54:14] \n",
      "\n",
      "\tSTEP 19: Filter HMM predicted CNVs based on the Bayesian Network Model results and BayesMaxPNormal\n",
      "\n",
      "INFO [2025-10-26 16:54:15] Attempting to removing CNV(s) with a probability of being normal above  0.5\n",
      "INFO [2025-10-26 16:54:15] Removing  5  CNV(s) identified by the HMM.\n",
      "INFO [2025-10-26 16:54:15] Total CNV's after removing:  1668\n",
      "INFO [2025-10-26 16:54:15] Reassigning CNVs based on state probabilities.\n",
      "INFO [2025-10-26 16:54:15] Changing the following CNV's states assigned by the HMM to the following based on the CNV's state probabilities.\n",
      " chr11-region_285 : 4  (P= 0.404025530679923 ) ->  3 (P= 0.424410943571656 )\n",
      "chr19-region_315 : 2  (P= 0.404945974582536 ) ->  3 (P= 0.469532135679121 )\n",
      "chr11-region_422 : 2  (P= 0.414478699899208 ) ->  3 (P= 0.448182816129959 )\n",
      "chr4-region_483 : 4  (P= 0.405303024097322 ) ->  3 (P= 0.416708684379041 )\n",
      "chr16-region_505 : 2  (P= 0.393825130122651 ) ->  3 (P= 0.462893451600025 )\n",
      "chr16-region_507 : 2  (P= 0.426026314304597 ) ->  3 (P= 0.429593510538129 )\n",
      "chr3-region_622 : 4  (P= 0.388662275909313 ) ->  3 (P= 0.428900966898535 )\n",
      "chr4-region_951 : 4  (P= 0.385563632665396 ) ->  3 (P= 0.419911360146806 )\n",
      "chr9-region_1315 : 2  (P= 0.395005620134498 ) ->  3 (P= 0.404939085737452 )\n",
      "chr6-region_2058 : 2  (P= 0.272658091176975 ) ->  3 (P= 0.365051743744697 )\n",
      "chr3-region_2089 : 4  (P= 0.447663793373664 ) ->  3 (P= 0.461499121886816 )\n",
      "chr7-region_2262 : 5  (P= 0.406928092970282 ) ->  4 (P= 0.467939190212043 )\n",
      "chr17-region_2401 : 4  (P= 0.366978162873919 ) ->  3 (P= 0.451302178252425 )\n",
      "chr17-region_2533 : 2  (P= 0.455903041524089 ) ->  3 (P= 0.457629295956128 )\n",
      "chr1-region_2829 : 5  (P= 0.278706249064277 ) ->  4 (P= 0.425097294260741 )\n",
      "chr1-region_2831 : 5  (P= 0.427332901109428 ) ->  4 (P= 0.48074969470334 )\n",
      "chr8-region_2860 : 4  (P= 0.35187590917645 ) ->  3 (P= 0.443549166583148 )\n",
      "chr6-region_2864 : 4  (P= 0.277112636532896 ) ->  3 (P= 0.499065612838338 )\n",
      "chr4-region_2945 : 4  (P= 0.405585949701413 ) ->  3 (P= 0.459874767251627 )\n",
      "chr16-region_2980 : 2  (P= 0.426933134609764 ) ->  3 (P= 0.465629891670113 )\n",
      "chr1-region_2991 : 4  (P= 0.395171867840795 ) ->  3 (P= 0.454613753071146 )\n",
      "chr3-region_3217 : 4  (P= 0.247132787590186 ) ->  3 (P= 0.254094208650617 )\n",
      "chr3-region_3223 : 4  (P= 0.2481359793517 ) ->  3 (P= 0.249253902638915 )\n",
      "chr15-region_3249 : 4  (P= 0.249526992856934 ) ->  3 (P= 0.250414561972283 )\n",
      "chr11-region_3260 : 2  (P= 0.247081304363607 ) ->  3 (P= 0.252495018676466 )\n",
      "chr17-region_3276 : 2  (P= 0.251676647049296 ) ->  3 (P= 0.25207276796445 )\n",
      "chr8-region_3288 : 6  (P= 0.251360629777571 ) ->  5 (P= 0.251903135088546 )\n",
      "chr6-region_3297 : 5  (P= 0.249793401137984 ) ->  4 (P= 0.250352634241294 )\n",
      "chr9-region_3347 : 4  (P= 0.391598514961896 ) ->  3 (P= 0.414315649015835 )\n",
      "chr17-region_3708 : 4  (P= 0.295189695270772 ) ->  3 (P= 0.338719864739144 )\n",
      "chr1-region_3818 : 4  (P= 0.179228962863817 ) ->  3 (P= 0.364530165691949 )\n",
      "chr6-region_3846 : 4  (P= 0.180535091187919 ) ->  3 (P= 0.274167576973692 )\n",
      "chr13-region_3853 : 2  (P= 0.273859368327761 ) ->  3 (P= 0.362935708975308 )\n",
      "INFO [2025-10-26 16:54:15] Creating Plots for CNV and cell Probabilities.\n",
      "INFO [2025-10-26 17:02:39] ::plot_cnv:Start\n",
      "INFO [2025-10-26 17:02:39] ::plot_cnv:Current data dimensions (r,c)=18168,1460 Total=4263326.90293415 Min=0 Max=0.981616515338111.\n",
      "INFO [2025-10-26 17:02:40] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2025-10-26 17:02:40] plot_cnv_observation:Start\n",
      "INFO [2025-10-26 17:02:40] Observation data size: Cells= 1324 Genes= 18168\n",
      "INFO [2025-10-26 17:02:40] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2025-10-26 17:02:40] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2025-10-26 17:02:40] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 17:02:40] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 17:02:43] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 17:02:43] Quantiles of plotted data range: 0,0,0,0,0.981616515338111\n",
      "INFO [2025-10-26 17:02:45] plot_cnv_references:Start\n",
      "INFO [2025-10-26 17:02:45] Reference data size: Cells= 136 Genes= 18168\n",
      "INFO [2025-10-26 17:02:45] plot_cnv_references:Number reference groups= 1\n",
      "INFO [2025-10-26 17:02:45] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2025-10-26 17:02:46] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 17:02:46] Quantiles of plotted data range: 0,0,0,0,0.854748054554991\n",
      "INFO [2025-10-26 17:02:58] ::plot_cnv:Start\n",
      "INFO [2025-10-26 17:02:58] ::plot_cnv:Current data dimensions (r,c)=18168,1460 Total=80925076 Min=2 Max=5.\n",
      "INFO [2025-10-26 17:02:59] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2025-10-26 17:02:59] plot_cnv_observation:Start\n",
      "INFO [2025-10-26 17:02:59] Observation data size: Cells= 1324 Genes= 18168\n",
      "INFO [2025-10-26 17:02:59] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2025-10-26 17:02:59] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2025-10-26 17:02:59] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 17:02:59] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 17:03:02] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 17:03:02] Quantiles of plotted data range: 2,3,3,3,5\n",
      "INFO [2025-10-26 17:03:04] plot_cnv_references:Start\n",
      "INFO [2025-10-26 17:03:04] Reference data size: Cells= 136 Genes= 18168\n",
      "INFO [2025-10-26 17:03:04] plot_cnv_references:Number reference groups= 1\n",
      "INFO [2025-10-26 17:03:04] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2025-10-26 17:03:04] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 17:03:04] Quantiles of plotted data range: 2,3,3,3,4\n",
      "INFO [2025-10-26 17:03:12] \n",
      "\n",
      "\tSTEP 20: Converting HMM-based CNV states to repr expr vals\n",
      "\n",
      "INFO [2025-10-26 17:03:24] ::plot_cnv:Start\n",
      "INFO [2025-10-26 17:03:24] ::plot_cnv:Current data dimensions (r,c)=18168,1460 Total=27199898 Min=0.5 Max=2.\n",
      "INFO [2025-10-26 17:03:25] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2025-10-26 17:03:25] plot_cnv_observation:Start\n",
      "INFO [2025-10-26 17:03:25] Observation data size: Cells= 1324 Genes= 18168\n",
      "INFO [2025-10-26 17:03:25] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2025-10-26 17:03:25] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2025-10-26 17:03:25] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 17:03:25] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 17:03:28] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 17:03:28] Quantiles of plotted data range: 0.5,1,1,1,2\n",
      "INFO [2025-10-26 17:03:30] plot_cnv_references:Start\n",
      "INFO [2025-10-26 17:03:30] Reference data size: Cells= 136 Genes= 18168\n",
      "INFO [2025-10-26 17:03:30] plot_cnv_references:Number reference groups= 1\n",
      "INFO [2025-10-26 17:03:30] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2025-10-26 17:03:30] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 17:03:30] Quantiles of plotted data range: 0.5,1,1,1,1.5\n",
      "INFO [2025-10-26 17:03:31] \n",
      "\n",
      "\tSTEP 22: Denoising\n",
      "\n",
      "INFO [2025-10-26 17:03:31] ::process_data:Remove noise, noise threshold defined via ref mean sd_amplifier:  1.5\n",
      "INFO [2025-10-26 17:03:31] denoising using mean(normal) +- sd_amplifier * sd(normal) per gene per cell across all data\n",
      "INFO [2025-10-26 17:03:31] :: **** clear_noise_via_ref_quantiles **** : removing noise between bounds:  0.891203221677113 - 1.11431646868135\n",
      "INFO [2025-10-26 17:03:58] \n",
      "\n",
      "## Making the final infercnv heatmap ##\n",
      "INFO [2025-10-26 17:03:59] ::plot_cnv:Start\n",
      "INFO [2025-10-26 17:03:59] ::plot_cnv:Current data dimensions (r,c)=18168,1460 Total=26709245.760419 Min=0.506573411933416 Max=2.13705129387919.\n",
      "INFO [2025-10-26 17:03:59] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2025-10-26 17:04:00] plot_cnv(): auto thresholding at: (0.682501 , 1.317499)\n",
      "INFO [2025-10-26 17:04:00] plot_cnv_observation:Start\n",
      "INFO [2025-10-26 17:04:00] Observation data size: Cells= 1324 Genes= 18168\n",
      "INFO [2025-10-26 17:04:00] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2025-10-26 17:04:00] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2025-10-26 17:04:00] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 17:04:00] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2025-10-26 17:04:03] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 17:04:03] Quantiles of plotted data range: 0.682501454880828,1.00275984517923,1.00275984517923,1.00275984517923,1.31749854511917\n",
      "INFO [2025-10-26 17:04:05] plot_cnv_references:Start\n",
      "INFO [2025-10-26 17:04:05] Reference data size: Cells= 136 Genes= 18168\n",
      "INFO [2025-10-26 17:04:06] plot_cnv_references:Number reference groups= 1\n",
      "INFO [2025-10-26 17:04:06] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2025-10-26 17:04:06] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2025-10-26 17:04:06] Quantiles of plotted data range: 0.682501454880828,1.00275984517923,1.00275984517923,1.00275984517923,1.31749854511917\n"
     ]
    }
   ],
   "source": [
    "infercnv_obj = CreateInfercnvObject(raw_counts_matrix=\"/home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_input/exp_arch_prolif_SPID.tsv\",\n",
    "                                    annotations_file=\"/home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_input/ann_arch_prolif_SPID.txt\",\n",
    "                                    delim=\"\\t\",\n",
    "                                    gene_order_file=\"/home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_input/ann_gene.txt\",\n",
    "                                    ref_group_names=c(\"Normal_PC\"))\n",
    "\n",
    "infercnv_obj = infercnv::run(infercnv_obj,\n",
    "                             cutoff=0.1,  \n",
    "                             out_dir=\"/home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_output/output_infercnv_r_arch_prolif_SPID\",  \n",
    "                             cluster_by_groups=TRUE,   \n",
    "                             denoise=TRUE,\n",
    "                             HMM=TRUE,\n",
    "                             output_format=\"pdf\"\n",
    "                             )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
