{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from omegaconf import OmegaConf\n",
    "#\n",
    "repo_dir = '/home/projects/amit/annaku/repos/Blueprint'\n",
    "sys.path.append(repo_dir)\n",
    "#\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "#\n",
    "plt.rcParams[\"image.cmap\"] = \"Set1\"\n",
    "sns.set_theme(style='ticks', rc={'axes.grid': False})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.cnv_utils import *\n",
    "from src.palettes import *\n",
    "\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', 300)\n",
    "%config InlineBackend.figure_format = 'png'\n",
    "plt.rcParams['pdf.fonttype'] = 'truetype'\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "\n",
    "sns.set_style('ticks')\n",
    "sns.set_style('white')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "version = '20250306'\n",
    "\n",
    "project_root = '/home/projects/amit/annaku/repos/Blueprint'\n",
    "conf_path = os.path.join(project_root, 'configs', 'config.yaml')\n",
    "conf = OmegaConf.load(conf_path)\n",
    "\n",
    "save_figures = True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4",
   "metadata": {},
   "source": [
    "# cnv outputs to adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path =conf['outputs']['output_dir']\n",
    "filename = f'adata_PC_with_ann_merged_v_{version}.h5ad' \n",
    "adata_cnv = sc.read_h5ad(os.path.join(data_path, filename))\n",
    "adata_cnv = adata_cnv[adata_cnv.obs['cells_rem_dupl_between_methods'] == False]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_cnv.obs['Sample.Code.Cell'] = adata_cnv.obs['Sample.Code'].astype(str) + '.' + adata_cnv.obs['Populations'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# arch merge, attention to version\n",
    "\n",
    "path_save_8arch = '/home/projects/amit/annaku/repos/Blueprint/data/processed/nmf_outputs/renamed_with_8_separated/'\n",
    "\n",
    "arch = pd.read_csv(os.path.join(path_save_8arch,f'arch_sample_v7_samplelevel_only_malignant_renamed_with_added_samples_v_{version}.csv'), index_col = 0)\n",
    "arch[['Cluster', 'Cluster_exp']] = arch[['Cluster', 'Cluster_exp']].astype(str)\n",
    "arch_gene = pd.read_csv(os.path.join(path_save_8arch,'arch_gene_v7_samplelevel_only_malignant_renamed.csv'), index_col = 0)\n",
    "arch_gene[['Cluster']] = arch_gene[['Cluster']].astype(str)\n",
    "\n",
    "adata_cnv.obs['PID'] = 'z.' + adata_cnv.obs['Method'].astype(str) + '_' + adata_cnv.obs['Populations'].astype(str) + '_' + adata_cnv.obs[\"Sample.Code\"].astype(str).str.lower()\n",
    "adata_cnv.obs['PID'] = adata_cnv.obs['PID'].str.lower()\n",
    "\n",
    "arch['PID'] = arch['PID'].str.lower() \n",
    "arch['patient'] = arch['patient'].str.lower()\n",
    "\n",
    "adata_cnv.obs['index'] = adata_cnv.obs.index\n",
    "adata_cnv.obs = pd.merge(adata_cnv.obs, arch[['Cluster_exp', 'Cluster', 'PID', 'prolif_high_0.1', '8.0']], how='left', on='PID')\n",
    "adata_cnv.obs.index = adata_cnv.obs['index']\n",
    "\n",
    "adata_cnv.obs = adata_cnv.obs.rename(columns = {'Cluster_exp':'arch', 'Cluster':'arch_with_8', '8.0':'prolif_coef'})\n",
    "\n",
    "arch_rename_dict = {'2.0':'MM1',\n",
    "                    '3.0':'MM2',\n",
    "                    '4.0':'MM3',\n",
    "                    '6.0':'MM4',\n",
    "                    '7.0':'MM5'}\n",
    "\n",
    "adata_cnv.obs['arch'] = adata_cnv.obs['arch'].replace(arch_rename_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_minibulk_combined = process_minibulk_cnv(adata_cnv,\n",
    "                                            to_filter = False,\n",
    "                                            reference_key = 'Populations',\n",
    "                                        reference_cat=['Normal_PC'],\n",
    "                                            preproc_standart = False,\n",
    "                                            cat_to_group=['Sample.Code', 'Populations'])\n",
    "\n",
    "adata_minibulk_combined.obs['Populations'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_minibulk_combined.obs = adata_minibulk_combined.obs.drop(columns = ['cnv_leiden'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "methods = ['SPID', 'MARS']\n",
    "base_data_path =conf['outputs']['output_dir'] + '/infercnv_r_output/output_infercnv_r_arch_prolif_'\n",
    "\n",
    "dataframes = []\n",
    "\n",
    "for method in methods:\n",
    "    data_path = base_data_path + method\n",
    "    filename = 'infercnv.observation_groupings.txt'\n",
    "    \n",
    "    obs_df = pd.read_csv(os.path.join(data_path, filename), sep=' \"', engine='python', header=0,\n",
    "                         names=['cell_id', 'group', 'dendrogram_color', 'annotation', 'color'])\n",
    "    \n",
    "    obs_df['cell_id'] = obs_df['cell_id'].str.strip('\"')\n",
    "    obs_df = obs_df.set_index('cell_id')\n",
    "    obs_df = obs_df.apply(lambda x: x.str.strip('\"'))\n",
    "    \n",
    "    dataframes.append(obs_df)\n",
    "\n",
    "merged_obs_df = pd.concat(dataframes)\n",
    "\n",
    "ann = adata_minibulk_combined.obs.copy()\n",
    "ann_cell_level = pd.concat([ann, merged_obs_df], axis=1)\n",
    "print(ann_cell_level.shape)\n",
    "\n",
    "ann_cell_level['prolif_high_0.1'] = ann_cell_level['prolif_high_0.1'].astype(str)\n",
    "\n",
    "ann_cell_level.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o expr_data_MARS,row_names_MARS,col_names_MARS,gene_order_MARS\n",
    "result <- readRDS('/home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_output/output_infercnv_r_arch_prolif_MARS/run.final.infercnv_obj')\n",
    "\n",
    "expr_data_MARS <- result@expr.data\n",
    "\n",
    "gene_order_MARS <- result@gene_order\n",
    "\n",
    "row_names_MARS <- rownames(expr_data_MARS)\n",
    "col_names_MARS <- colnames(expr_data_MARS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o expr_data_SPID,row_names_SPID,col_names_SPID,gene_order_SPID\n",
    "result <- readRDS('/home/projects/amit/annaku/repos/Blueprint/data/processed/infercnv_r_output/output_infercnv_r_arch_prolif_SPID/run.final.infercnv_obj')\n",
    "\n",
    "expr_data_SPID <- result@expr.data\n",
    "\n",
    "gene_order_SPID <- result@gene_order\n",
    "\n",
    "row_names_SPID <- rownames(expr_data_SPID)\n",
    "col_names_SPID <- colnames(expr_data_SPID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": [
    "expr_data_MARS_df = pd.DataFrame(data=expr_data_MARS, index=row_names_MARS, columns=col_names_MARS)\n",
    "print(expr_data_MARS_df.shape)\n",
    "expr_data_SPID_df = pd.DataFrame(data=expr_data_SPID, index=row_names_SPID, columns=col_names_SPID)\n",
    "print(expr_data_SPID_df.shape)\n",
    "expr_data_MARS_df.head(2)\n",
    "\n",
    "common_genes = gene_order_MARS.index.intersection(gene_order_SPID.index)\n",
    "gene_order = gene_order_MARS.loc[common_genes]\n",
    "\n",
    "expr_data_MARS_common = expr_data_MARS_df.loc[common_genes]\n",
    "expr_data_SPID_common = expr_data_SPID_df.loc[common_genes]\n",
    "\n",
    "cnv_pred = pd.concat([expr_data_MARS_common, expr_data_SPID_common], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '/home/projects/amit/annaku/repos/Blueprint/data/processed/'\n",
    "cb_file = pd.read_csv(os.path.join(path, 'cytoBand.txt'), sep='\\t', )\n",
    "\n",
    "def get_cytoband(chrom, pos, cytoband_df):\n",
    "    chr_bands = cytoband_df[cytoband_df['chr'] == chrom]\n",
    "    \n",
    "    band = chr_bands[(chr_bands['start'] <= pos) & (chr_bands['end'] >= pos)]\n",
    "    \n",
    "    if len(band) == 0:\n",
    "        return None\n",
    "    return band.iloc[0]['band']\n",
    "\n",
    "def annotate_cytobands(df, cytoband_df):\n",
    "    df = df.copy()\n",
    "    df['cytoband'] = df.apply(\n",
    "        lambda row: get_cytoband(row['chr'], row['start'], cytoband_df), \n",
    "        axis=1\n",
    "    )\n",
    "    return df\n",
    "\n",
    "gene_df = pd.DataFrame(gene_order).reset_index().rename(columns={'index': 'gene_name'})\n",
    "ann_genes = annotate_cytobands(gene_df, cb_file)\n",
    "ann_genes.index = ann_genes['gene_name']\n",
    "\n",
    "def get_arm(cytoband):\n",
    "    if pd.isna(cytoband):\n",
    "        return None\n",
    "    return cytoband[0]\n",
    "\n",
    "ann_genes['arm'] = ann_genes['cytoband'].apply(get_arm)\n",
    "ann_genes['chr_arm'] = ann_genes['chr'].astype(str) + '_' + ann_genes['arm'].astype(str)\n",
    "\n",
    "ann_genes['chr_band'] = ann_genes['chr'].astype(str) + '_' + ann_genes['cytoband'].astype(str)\n",
    "\n",
    "print(ann_genes)\n",
    "ann_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(pd.crosstab(ann_genes['chr'], ann_genes['arm']))\n",
    "\n",
    "# # too few genes per arm\n",
    "ann_genes['chr_arm'] = ann_genes['chr_arm'].replace({'chr14_p':'chr14',\n",
    "                                                                                'chr14_q':'chr14',\n",
    "                                                                                'chr22_p':'chr22',\n",
    "                                                                                'chr22_q':'chr22',\n",
    "                                                                                'chr15_p':'chr15',\n",
    "                                                                                'chr15_q':'chr15',\n",
    "                                                                                'chr13_p':'chr13',\n",
    "                                                                                'chr13_q':'chr13',\n",
    "                                                                                'chr21_p':'chr21',\n",
    "                                                                                'chr21_q':'chr21',\n",
    "                                                                                 })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16",
   "metadata": {},
   "outputs": [],
   "source": [
    "# minibulk level\n",
    "\n",
    "adata_cell_level = ad.AnnData(\n",
    "    X=cnv_pred.T.loc[ann_cell_level.index].values, \n",
    "    obs=ann_cell_level,\n",
    "    var=ann_genes\n",
    ")\n",
    "\n",
    "# aggregate adata pat level\n",
    "\n",
    "pat_level_expression = adata_cell_level.to_df().groupby(adata_cell_level.obs['Sample.Code.Cell']).mean()\n",
    "pat_level_ann = adata_cell_level.obs.groupby('Sample.Code.Cell').first()\n",
    "\n",
    "adata_pat_level = ad.AnnData(\n",
    "    X=pat_level_expression.values,\n",
    "    obs=pat_level_ann,\n",
    "    var=adata_cell_level.var\n",
    ")\n",
    "\n",
    "adata_pat_level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_pat_level.var['chr_band_wide'] = adata_pat_level.var['chr_band'].str.split('.').str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_pat_level.obs['Populations'] = pd.Categorical(adata_pat_level.obs['Populations'])\n",
    "adata_pat_level.obs['arch'] = pd.Categorical(adata_pat_level.obs['arch'])\n",
    "\n",
    "adata_pat_level.uns['Populations_colors'] = [pal_cell_pb[x] for x in adata_pat_level.obs['Populations'].cat.categories]\n",
    "adata_pat_level.uns['arch_colors'] = [pal_architype_renamed[x] for x in adata_pat_level.obs['arch'].cat.categories]\n",
    "\n",
    "sc.tl.pca(adata_pat_level, svd_solver='arpack', n_comps=50)\n",
    "sc.pl.pca(adata_pat_level, color=['prolif_high_0.1','arch', 'Method'], ncols = 2,\n",
    "           frameon = False,)\n",
    "sc.pp.neighbors(adata_pat_level, \n",
    "                n_neighbors=15,     \n",
    "                n_pcs=30)         \n",
    "\n",
    "sc.tl.umap(adata_pat_level, min_dist=0.8)\n",
    "\n",
    "sc.pl.umap(adata_pat_level, \n",
    "           color=['arch', 'prolif_high_0.1', 'Populations'],\n",
    "           frameon = False,\n",
    "           ncols=2)\n",
    "\n",
    "sc.pl.pca_variance_ratio(adata_pat_level, n_pcs=50)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19",
   "metadata": {},
   "source": [
    "# features extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path ='/home/projects/amit/annaku/repos/Blueprint/data/processed/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.cnv_utils import calculate_genome_wide_cnv_burden"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_pat_level = calculate_genome_wide_cnv_burden(\n",
    "    adata=adata_pat_level,\n",
    "    window_size=100,\n",
    "    window_step=25,\n",
    "    sort_genes=True, \n",
    ")\n",
    "\n",
    "adata_pat_level.var['chr_band_wide'] = adata_pat_level.var['chr_band'].str.split('.').str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23",
   "metadata": {},
   "outputs": [],
   "source": [
    "regions_todel = ['chr5_None', 'chr17_None', 'chr18_None', 'chr20_None']\n",
    "\n",
    "mask = adata_pat_level.var['chr_arm'].isin(regions_todel)\n",
    "adata_pat_level_ = adata_pat_level[:, ~mask].copy()\n",
    "\n",
    "window_size = 10 \n",
    "window_step = 5 \n",
    "min_genes = 20  \n",
    "\n",
    "adata_arms = sliding_window_cnv_region(\n",
    "    adata_pat_level_, \n",
    "    region_col='chr_arm',\n",
    "    window_size=window_size,\n",
    "    window_step=window_step,\n",
    "    min_genes=min_genes\n",
    ")\n",
    "\n",
    "def chr_arm_key(x):\n",
    "    parts = x.split('_')\n",
    "    chrom = parts[0][3:]\n",
    "    arm = parts[1] if len(parts) > 1 else 'p' \n",
    "    \n",
    "    try:\n",
    "        chrom_num = int(chrom)\n",
    "    except ValueError:\n",
    "        chrom_num = float('inf') if chrom in ['X', 'Y'] else 0\n",
    "    \n",
    "    return (chrom_num, 0 if arm == 'p' else 1)\n",
    "\n",
    "sorted_arms = sorted(adata_arms.var_names.unique(), key=chr_arm_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_pat_level_ = adata_pat_level.copy()\n",
    "\n",
    "window_size = 10  \n",
    "window_step = 5   \n",
    "min_genes = 10   \n",
    "\n",
    "adata_bands = sliding_window_cnv_region(\n",
    "    adata_pat_level_, \n",
    "    region_col='chr_band_wide',\n",
    "    window_size=window_size,\n",
    "    window_step=window_step,\n",
    "    min_genes=min_genes\n",
    ")\n",
    "\n",
    "def chr_arm_key(x):\n",
    "    parts = x.split('_')\n",
    "    chrom = parts[0][3:]\n",
    "    arm = parts[1] if len(parts) > 1 else 'p'  \n",
    "    \n",
    "    try:\n",
    "        chrom_num = int(chrom)\n",
    "    except ValueError:\n",
    "        chrom_num = float('inf') if chrom in ['X', 'Y'] else 0\n",
    "    \n",
    "    return (chrom_num, 0 if arm == 'p' else 1)\n",
    "\n",
    "sorted_bands = sorted(adata_bands.var_names.unique(), key=chr_arm_key)\n",
    "len(sorted_bands)\n",
    "\n",
    "def extract_chr_and_arm(var_name):\n",
    "    parts = var_name.split('_')\n",
    "    chr_part = parts[0]\n",
    "    arm_part = parts[1] if len(parts) > 1 else ''\n",
    "    arm = arm_part[0] if arm_part else ''\n",
    "    return chr_part, f\"{chr_part}_{arm}\" if arm else None\n",
    "\n",
    "var_names_list = list(adata_bands.var_names)\n",
    "\n",
    "extracted_data = [extract_chr_and_arm(var_name) for var_name in var_names_list]\n",
    "\n",
    "extracted_df = pd.DataFrame(extracted_data, columns=['chr', 'chr_arm'], index=adata_bands.var_names)\n",
    "\n",
    "adata_bands.var['chr'] = extracted_df['chr']\n",
    "adata_bands.var['chr_arm'] = extracted_df['chr_arm']\n",
    "\n",
    "unique_chromosomes = set(var_name.split('_')[0] for var_name in adata_bands.var_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25",
   "metadata": {},
   "source": [
    "# binarisation and HP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_cell_level.obs['arch'] = adata_cell_level.obs['arch'].astype(str).fillna('NA').astype('category')\n",
    "adata_cell_level.obs['Populations'] = adata_cell_level.obs['Populations'].astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.dendrogram(adata_cell_level, groupby=['Populations', 'arch'])\n",
    "\n",
    "sc.pl.heatmap(adata_cell_level, \n",
    "              var_names=adata_cell_level.var_names,  \n",
    "              groupby=['Populations', 'arch'],        \n",
    "              cmap='bwr',                \n",
    "              vcenter=1,                     \n",
    "              vmin=0.6,                     \n",
    "              vmax=1.4,\n",
    "              figsize = (6,4),                     \n",
    "              dendrogram=True,              \n",
    "              show_gene_labels=False,       \n",
    "              swap_axes=False)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28",
   "metadata": {},
   "outputs": [],
   "source": [
    "for var_name in adata_arms.var_names:\n",
    "    values = adata_arms[:, var_name].X.flatten()\n",
    "    \n",
    "    # initialize calls as 'neutral'\n",
    "    calls = np.full(len(values), 'neutral', dtype='object')\n",
    "    \n",
    "    # set amplifications (> 1.05)\n",
    "    calls[values > 1.05] = 'ampl'\n",
    "    \n",
    "    # set deletions (< 0.95) \n",
    "    calls[values < 0.95] = 'del'\n",
    "    \n",
    "    adata_arms.obs[f'{var_name}_call'] = calls\n",
    "\n",
    "print(f\"Added {len(adata_arms.var_names)} new call columns\")\n",
    "print(\"Example column names:\", list(filter(lambda x: x.endswith('_call'), adata_arms.obs.columns))[:3])\n",
    "\n",
    "# create combined chromosome calls for specified chromosomes\n",
    "for chrom in [3, 5, 7, 9, 11, 19]:\n",
    "    # get p and q arm calls\n",
    "    p_calls = adata_arms.obs[f'chr{chrom}_p_call']\n",
    "    q_calls = adata_arms.obs[f'chr{chrom}_q_call']\n",
    "    \n",
    "    # initialize combined calls\n",
    "    combined_calls = np.full(len(p_calls), 'mixed', dtype='object')\n",
    "    \n",
    "    # where p and q match, use that call\n",
    "    mask_match = p_calls == q_calls\n",
    "    combined_calls[mask_match] = p_calls[mask_match]\n",
    "    \n",
    "    # add to obs\n",
    "    adata_arms.obs[f'chr{chrom}_call'] = combined_calls\n",
    "\n",
    "print(\"Example values for chr3_call:\", adata_arms.obs['chr3_call'].value_counts())\n",
    "\n",
    "# HP\n",
    "chromosomes = [3, 5, 7, 9, 11, 15, 19, 21]\n",
    "\n",
    "# Count amplifications for each sample\n",
    "ampl_counts = np.zeros(len(adata_arms))\n",
    "\n",
    "for chrom in chromosomes:\n",
    "    # Add to count where status is 'ampl'\n",
    "    ampl_counts += (adata_arms.obs[f'chr{chrom}_call'] == 'ampl')\n",
    "\n",
    "# Create HP_call column - 'Yes' if â‰¥2 amplifications, 'No' otherwise\n",
    "adata_arms.obs['HP_call'] = np.where(ampl_counts >= 2, 'Yes', 'No')\n",
    "\n",
    "# Verify the results\n",
    "print(\"HP_call distribution:\", adata_arms.obs['HP_call'].value_counts())\n",
    "print(\"\\nMedian number of amplifications:\", np.median(ampl_counts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['pdf.fonttype'] = 'truetype'\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "\n",
    "save_dir = '/home/projects/amit/annaku/repos/Blueprint/figures/fig3/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats\n",
    "\n",
    "def plot_cnv_distribution(adata_arms, regions, figsize=(12, 3), threshold_del=0.95, threshold_amp=1.05):\n",
    "    \n",
    "    n = len(regions)\n",
    "    fig, axes = plt.subplots(1, n, figsize=figsize, sharey=True)\n",
    "    axes = [axes] if n == 1 else axes\n",
    "    \n",
    "    control_labels = ['nan', 'NA', 'None']\n",
    "    palette = {**pal_architype_renamed, 'Reference': 'red'}\n",
    "    \n",
    "    archetypes = sorted(adata_arms.obs['arch'].astype(str).unique())\n",
    "    stats_dict = {}\n",
    "    \n",
    "    for ax, region in zip(axes, regions):\n",
    "        for arch in archetypes:\n",
    "            scores = adata_arms[adata_arms.obs['arch'].astype(str) == arch, region].X.flatten()\n",
    "            label = 'Reference' if arch in control_labels else arch\n",
    "            sns.kdeplot(scores, label=label, color=palette.get(label, '#7f7f7f'), ax=ax)\n",
    "        \n",
    "        ax.axvline(x=threshold_del, color='r', linestyle='--', alpha=0.7)\n",
    "        ax.axvline(x=threshold_amp, color='r', linestyle='--', alpha=0.7)\n",
    "        ax.set(xlabel=f'{region} score', ylabel='Density', title=region)\n",
    "        ax.spines[['top', 'right']].set_visible(False)\n",
    "        if ax == axes[-1]:\n",
    "            ax.legend(fontsize=8)\n",
    "        \n",
    "        mm_archs = [a for a in archetypes if a not in control_labels]\n",
    "        scores_by_arch = [adata_arms[adata_arms.obs['arch'].astype(str) == a, region].X.flatten() for a in mm_archs]\n",
    "        h, p = stats.kruskal(*scores_by_arch)\n",
    "        stats_dict[region] = {'h': h, 'p': p}\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    return fig, stats_dict\n",
    "\n",
    "regions = ['chr1_p', 'chr1_q', 'chr13', 'chr17_p']\n",
    "fig, stats_dict = plot_cnv_distribution(adata_arms, regions)\n",
    "\n",
    "if save_figures:\n",
    "    fig.savefig(f'{save_dir}hist_cnv_QC_binarization_v_{version}.svg', format='svg')\n",
    "    fig.savefig(f'{save_dir}hist_cnv_QC_binarization_v_{version}.png', format='png')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "for region, s in stats_dict.items():\n",
    "    print(f\"{region}: H={s['h']:.2f}, p={s['p']:.2e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31",
   "metadata": {},
   "outputs": [],
   "source": [
    "# separated annotation\n",
    "\n",
    "cnv_call_annotation_arms = adata_arms.obs[[col for col in adata_arms.obs.columns if col.endswith('_call')]]\n",
    "cnv_call_annotation_bands = adata_bands.obs[[col for col in adata_bands.obs.columns if col.endswith('_call')]]\n",
    "cnv_call_annotation_burden = adata_pat_level.obs[['cnv_burden']]\n",
    "cnv_call_annotation = pd.concat([cnv_call_annotation_arms, cnv_call_annotation_bands, cnv_call_annotation_burden], axis = 1)\n",
    "\n",
    "cnv_call_annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save\n",
    "\n",
    "ind_malignant =[i for i in cnv_call_annotation.index if 'Malignant' in i]\n",
    "cnv_call_annotation.loc[ind_malignant].to_csv(data_path+f'cnv_ann_per_sample_v_{version}.csv')\n",
    "\n",
    "cnv_call_annotation = cnv_call_annotation.loc[ind_malignant].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33",
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = [c for c in cnv_call_annotation.columns if c not in adata_bands.obs.columns]\n",
    "print(f'adding {len(cols)} columns')\n",
    "if len(cols) > 0:\n",
    "    adata_bands.obs = pd.concat([adata_bands.obs, cnv_call_annotation[cols]], join = 'outer', axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34",
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in cnv_call_annotation.columns:\n",
    "    adata_bands.obs[col] = adata_bands.obs[col].astype(str).fillna('NA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35",
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = [c for c in cnv_call_annotation.columns if c not in adata_pat_level.obs.columns]\n",
    "print(f'adding {len(cols)} columns')\n",
    "if len(cols) > 0:\n",
    "    adata_pat_level.obs = pd.concat([adata_pat_level.obs, cnv_call_annotation[cols]], join = 'outer', axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cols = [c for c in cnv_call_annotation.columns if c not in adata_pat_level.obs.columns]\n",
    "for col in cnv_call_annotation.columns:\n",
    "    adata_pat_level.obs[col] = adata_pat_level.obs[col].astype(str).fillna('NA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37",
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = [c for c in cnv_call_annotation.columns if c not in adata_arms.obs.columns]\n",
    "print(f'adding {len(cols)} columns')\n",
    "if len(cols) > 0:\n",
    "    adata_arms.obs = pd.concat([adata_arms.obs, cnv_call_annotation[cols]], join = 'outer', axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38",
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in cnv_call_annotation.columns:\n",
    "    adata_arms.obs[col] = adata_arms.obs[col].astype(str).fillna('NA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39",
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = [c for c in cnv_call_annotation.columns if c not in adata_cell_level.obs.columns]\n",
    "\n",
    "original_index_name = adata_cell_level.obs.index.name\n",
    "original_index = adata_cell_level.obs.index\n",
    "\n",
    "merged_obs = pd.merge(\n",
    "    adata_cell_level.obs, \n",
    "    cnv_call_annotation[cols], \n",
    "    on='Sample.Code.Cell', \n",
    "    how='left'\n",
    ")\n",
    "\n",
    "merged_obs = merged_obs.set_index(original_index)\n",
    "adata_cell_level.obs = merged_obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40",
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in cols:\n",
    "    adata_cell_level.obs[col] = adata_cell_level.obs[col].astype(str).fillna('NA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_cell_level.obs['Populations'].value_counts(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# artefact from removing samples sequenced by both methods\n",
    "\n",
    "adata_cell_level = adata_cell_level[~adata_cell_level.obs['Populations'].isna()].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.heatmap(adata_cell_level, \n",
    "              var_names=adata_cell_level.var_names,  \n",
    "              groupby=['Populations', 'chr1_q_call'],        \n",
    "              cmap='bwr',                \n",
    "              vcenter=1,                     \n",
    "              vmin=0.6,                     \n",
    "              vmax=1.4,\n",
    "              figsize = (6,4),                     \n",
    "              dendrogram=False,              \n",
    "              show_gene_labels=False,       \n",
    "              swap_axes=False)               "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(adata_pat_level.obs['Sample.Code'].astype(str).unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_pat_level.obs['Populations'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.heatmap(adata_pat_level, \n",
    "              var_names=adata_pat_level.var_names,  \n",
    "              groupby=['Populations', 'chr1_q_call'],        \n",
    "              cmap='bwr',                \n",
    "              vcenter=1,                     \n",
    "              vmin=0.6,                     \n",
    "              vmax=1.4,\n",
    "              figsize = (6,4),                     \n",
    "              dendrogram=False,              \n",
    "              show_gene_labels=False,       \n",
    "              swap_axes=False)  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47",
   "metadata": {},
   "source": [
    "# saving adatas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48",
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "\n",
    "for col in ['std', 'max', 'min']:\n",
    "    try: \n",
    "        adata_bands.var[col] = adata_bands.var[col].astype(str)\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "file_path = os.path.join(data_path,f'adata_cnv_obs_pat_var_bands_v_{version}.h5ad')\n",
    "adata_bands.write(file_path)\n",
    "\n",
    "#\n",
    "\n",
    "for col in ['std', 'max', 'min']:\n",
    "    try: \n",
    "        adata_arms.var[col] = adata_arms.var[col].astype(str)\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "file_path = os.path.join(data_path,f'adata_cnv_obs_pat_var_arms_v_{version}.h5ad')\n",
    "adata_arms.write(file_path)\n",
    "\n",
    "#\n",
    "\n",
    "for col in ['std', 'max', 'min']:\n",
    "    try: \n",
    "        adata_pat_level.var[col] = adata_pat_level.var[col].astype(str)\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "file_path = os.path.join(data_path,f'adata_cnv_obs_pat_var_genes_v_{version}.h5ad')\n",
    "adata_pat_level.write(file_path)\n",
    "\n",
    "#\n",
    "\n",
    "for col in adata_cell_level.obs.columns:\n",
    "    if adata_cell_level.obs[col].dtype == 'object':\n",
    "        try:\n",
    "            adata_cell_level.obs[col] = adata_cell_level.obs[col].astype(str)\n",
    "        except:\n",
    "            continue\n",
    "\n",
    "for col in ['std', 'max', 'min']:\n",
    "    try: \n",
    "        adata_cell_level.var[col] = adata_cell_level.var[col].astype(str)\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "file_path = os.path.join(data_path,f'adata_cnv_obs_minibulk_var_genes_v_{version}.h5ad')\n",
    "adata_cell_level.write(file_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
